// rclnodejs/generated/control_msgs/control_msgs__msg__SingleDOFState.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const SingleDOFStateRefStruct = StructType({
  name: primitiveTypes.string,
  reference: primitiveTypes.float64,
  feedback: primitiveTypes.float64,
  feedback_dot: primitiveTypes.float64,
  error: primitiveTypes.float64,
  error_dot: primitiveTypes.float64,
  time_step: primitiveTypes.float64,
  output: primitiveTypes.float64,
});
const SingleDOFStateRefArray = ArrayType(SingleDOFStateRefStruct);
const SingleDOFStateRefStructArray = StructType({
  data: SingleDOFStateRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class SingleDOFStateWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new SingleDOFStateRefStruct(other._refObject.toObject());
      this._wrapperFields.name =  new StringWrapper(other._wrapperFields.name);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new SingleDOFStateRefStruct();
    this._wrapperFields.name = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new SingleDOFStateWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SingleDOFStateArrayWrapper();
  }
  static get ArrayType() {
    return SingleDOFStateArrayWrapper;
  }
  static get refObjectArrayType() {
    return SingleDOFStateRefStructArray;
  }
  static get refObjectType() {
    return SingleDOFStateRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.name.freeze(own);
    }
    this._refObject.name = this._wrapperFields.name.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.name.data = refObject.name.data;
    this._refObject.reference = refObject.reference;
    this._refObject.feedback = refObject.feedback;
    this._refObject.feedback_dot = refObject.feedback_dot;
    this._refObject.error = refObject.error;
    this._refObject.error_dot = refObject.error_dot;
    this._refObject.time_step = refObject.time_step;
    this._refObject.output = refObject.output;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.name);
  }
  static destroyRawROS(msg) {
    SingleDOFStateWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'control_msgs', subFolder: 'msg', interfaceName: 'SingleDOFState'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get name() {
    return this._wrapperFields.name.data;
  }
  set name(value) {
    this._wrapperFields.name.data = value;
  }
  get reference() {
    return this._refObject.reference;
  }
  set reference(value) {
    this._refObject.reference = value;
  }
  get feedback() {
    return this._refObject.feedback;
  }
  set feedback(value) {
    this._refObject.feedback = value;
  }
  get feedback_dot() {
    return this._refObject.feedback_dot;
  }
  set feedback_dot(value) {
    this._refObject.feedback_dot = value;
  }
  get error() {
    return this._refObject.error;
  }
  set error(value) {
    this._refObject.error = value;
  }
  get error_dot() {
    return this._refObject.error_dot;
  }
  set error_dot(value) {
    this._refObject.error_dot = value;
  }
  get time_step() {
    return this._refObject.time_step;
  }
  set time_step(value) {
    this._refObject.time_step = value;
  }
  get output() {
    return this._refObject.output;
  }
  set output(value) {
    this._refObject.output = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SingleDOFStateRefStruct(refObject.toObject());
    this._wrapperFields.name.copyRefObject(this._refObject.name);
  }
  copy(other) {
    this._refObject = new SingleDOFStateRefStruct(other._refObject.toObject());
    this._wrapperFields.name.copy(other._wrapperFields.name);
  }
  static get classType() {
    return SingleDOFStateWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reference","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"feedback","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"feedback_dot","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"error","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"error_dot","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"time_step","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"output","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"control_msgs","type":"SingleDOFState","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SingleDOFState"};
  }
  hasMember(name) {
    let memberNames = ["name","reference","feedback","feedback_dot","error","error_dot","time_step","output"];
    return memberNames.indexOf(name) !== -1;
  }
}
class SingleDOFStateArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SingleDOFStateWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SingleDOFStateWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SingleDOFStateArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SingleDOFStateArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new SingleDOFStateRefArray(size);
    this._refObject = new SingleDOFStateRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SingleDOFStateWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SingleDOFStateArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SingleDOFStateArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SingleDOFStateWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SingleDOFStateWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SingleDOFStateArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = SingleDOFStateWrapper;
