// rclnodejs/generated/control_msgs/control_msgs__msg__JointTolerance.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const JointToleranceRefStruct = StructType({
  name: primitiveTypes.string,
  position: primitiveTypes.float64,
  velocity: primitiveTypes.float64,
  acceleration: primitiveTypes.float64,
});
const JointToleranceRefArray = ArrayType(JointToleranceRefStruct);
const JointToleranceRefStructArray = StructType({
  data: JointToleranceRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class JointToleranceWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new JointToleranceRefStruct(other._refObject.toObject());
      this._wrapperFields.name =  new StringWrapper(other._wrapperFields.name);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new JointToleranceRefStruct();
    this._wrapperFields.name = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new JointToleranceWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new JointToleranceArrayWrapper();
  }
  static get ArrayType() {
    return JointToleranceArrayWrapper;
  }
  static get refObjectArrayType() {
    return JointToleranceRefStructArray;
  }
  static get refObjectType() {
    return JointToleranceRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.name.freeze(own);
    }
    this._refObject.name = this._wrapperFields.name.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.name.data = refObject.name.data;
    this._refObject.position = refObject.position;
    this._refObject.velocity = refObject.velocity;
    this._refObject.acceleration = refObject.acceleration;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.name);
  }
  static destroyRawROS(msg) {
    JointToleranceWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'control_msgs', subFolder: 'msg', interfaceName: 'JointTolerance'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get name() {
    return this._wrapperFields.name.data;
  }
  set name(value) {
    this._wrapperFields.name.data = value;
  }
  get position() {
    return this._refObject.position;
  }
  set position(value) {
    this._refObject.position = value;
  }
  get velocity() {
    return this._refObject.velocity;
  }
  set velocity(value) {
    this._refObject.velocity = value;
  }
  get acceleration() {
    return this._refObject.acceleration;
  }
  set acceleration(value) {
    this._refObject.acceleration = value;
  }
  copyRefObject(refObject) {
    this._refObject = new JointToleranceRefStruct(refObject.toObject());
    this._wrapperFields.name.copyRefObject(this._refObject.name);
  }
  copy(other) {
    this._refObject = new JointToleranceRefStruct(other._refObject.toObject());
    this._wrapperFields.name.copy(other._wrapperFields.name);
  }
  static get classType() {
    return JointToleranceWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"position","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"velocity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"acceleration","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"control_msgs","type":"JointTolerance","stringUpperBound":null,"isPrimitiveType":false},"msgName":"JointTolerance"};
  }
  hasMember(name) {
    let memberNames = ["name","position","velocity","acceleration"];
    return memberNames.indexOf(name) !== -1;
  }
}
class JointToleranceArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof JointToleranceWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new JointToleranceWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to JointToleranceArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to JointToleranceArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new JointToleranceRefArray(size);
    this._refObject = new JointToleranceRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new JointToleranceWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof JointToleranceArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "JointToleranceArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      JointToleranceWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return JointToleranceWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return JointToleranceArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = JointToleranceWrapper;
