// rclnodejs/generated/control_msgs/control_msgs__msg__DynamicInterfaceGroupValues.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const InterfaceValueWrapper = require('../../generated/control_msgs/control_msgs__msg__InterfaceValue.js');
const DynamicInterfaceGroupValuesRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  interface_groups: StringWrapper.refObjectArrayType,
  interface_values: InterfaceValueWrapper.refObjectArrayType,
});
const DynamicInterfaceGroupValuesRefArray = ArrayType(DynamicInterfaceGroupValuesRefStruct);
const DynamicInterfaceGroupValuesRefStructArray = StructType({
  data: DynamicInterfaceGroupValuesRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class DynamicInterfaceGroupValuesWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new DynamicInterfaceGroupValuesRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.interface_groups = StringWrapper.createArray();
      this._wrapperFields.interface_groups.copy(other._wrapperFields.interface_groups);
      this.interface_groups = other.interface_groups;
      this._wrapperFields.interface_values = InterfaceValueWrapper.createArray();
      this._wrapperFields.interface_values.copy(other._wrapperFields.interface_values);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new DynamicInterfaceGroupValuesRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.interface_groups = StringWrapper.createArray();
    this._interface_groupsArray = [];
    this._wrapperFields.interface_values = InterfaceValueWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new DynamicInterfaceGroupValuesWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new DynamicInterfaceGroupValuesArrayWrapper();
  }
  static get ArrayType() {
    return DynamicInterfaceGroupValuesArrayWrapper;
  }
  static get refObjectArrayType() {
    return DynamicInterfaceGroupValuesRefStructArray;
  }
  static get refObjectType() {
    return DynamicInterfaceGroupValuesRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.header.freeze(own);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.interface_groups.fill([]);
    } else {
      this._wrapperFields.interface_groups.fill(this._interface_groupsArray);
    }
    this._wrapperFields.interface_groups.freeze(own);
    this._refObject.interface_groups = this._wrapperFields.interface_groups.refObject;
    if (own) {
      this._wrapperFields.interface_values.fill([]);
    }
    this._wrapperFields.interface_values.freeze(own);
    this._refObject.interface_values = this._wrapperFields.interface_values.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    refObject.interface_groups.data.length = refObject.interface_groups.size;
    for (let index = 0; index < refObject.interface_groups.size; index++) {
      this._interface_groupsArray[index] = refObject.interface_groups.data[index].data;
    }
    this._wrapperFields.interface_values.copyRefObject(refObject.interface_values);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.interface_groups.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.interface_groups);
      if (!StringWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.interface_groups, StringWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.interface_values.size != 0) {
      InterfaceValueWrapper.ArrayType.freeArray(refObject.interface_values);
      if (!InterfaceValueWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.interface_values, InterfaceValueWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    DynamicInterfaceGroupValuesWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'control_msgs', subFolder: 'msg', interfaceName: 'DynamicInterfaceGroupValues'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get interface_groups() {
    return this._interface_groupsArray;
  }
  set interface_groups(value) {
    this._interface_groupsArray = value;
  }
  get interface_values() {
    return this._wrapperFields.interface_values;
  }
  set interface_values(value) {
    this._wrapperFields.interface_values.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new DynamicInterfaceGroupValuesRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    refObject.interface_groups.data.length = refObject.interface_groups.size;
    for (let index = 0; index < refObject.interface_groups.size; index++) {
      this._interface_groupsArray[index] = refObject.interface_groups.data[index].data;
    }
    this._wrapperFields.interface_values.copyRefObject(this._refObject.interface_values);
  }
  copy(other) {
    this._refObject = new DynamicInterfaceGroupValuesRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._interface_groupsArray = other._interface_groupsArray.slice();
    this._wrapperFields.interface_values.copy(other._wrapperFields.interface_values);
  }
  static get classType() {
    return DynamicInterfaceGroupValuesWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"interface_groups","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"interface_values","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"control_msgs","type":"InterfaceValue","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"control_msgs","type":"DynamicInterfaceGroupValues","stringUpperBound":null,"isPrimitiveType":false},"msgName":"DynamicInterfaceGroupValues"};
  }
  hasMember(name) {
    let memberNames = ["header","interface_groups","interface_values"];
    return memberNames.indexOf(name) !== -1;
  }
}
class DynamicInterfaceGroupValuesArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof DynamicInterfaceGroupValuesWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new DynamicInterfaceGroupValuesWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DynamicInterfaceGroupValuesArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DynamicInterfaceGroupValuesArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new DynamicInterfaceGroupValuesRefArray(size);
    this._refObject = new DynamicInterfaceGroupValuesRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new DynamicInterfaceGroupValuesWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof DynamicInterfaceGroupValuesArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "DynamicInterfaceGroupValuesArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      DynamicInterfaceGroupValuesWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return DynamicInterfaceGroupValuesWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return DynamicInterfaceGroupValuesArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = DynamicInterfaceGroupValuesWrapper;
