// rclnodejs/generated/moveit_task_constructor_msgs/moveit_task_constructor_msgs__action__ExecuteTaskSolution.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
class moveit_task_constructor_msgs__action__ExecuteTaskSolution {
  static get Goal() {
    return require('./moveit_task_constructor_msgs__action__ExecuteTaskSolution_Goal.js');
  }
  static get Result() {
    return require('./moveit_task_constructor_msgs__action__ExecuteTaskSolution_Result.js');
  }
  static get Feedback() {
    return require('./moveit_task_constructor_msgs__action__ExecuteTaskSolution_Feedback.js');
  }
  static get impl() {
    return {
      get SendGoalService() {
        return require('./moveit_task_constructor_msgs__action__ExecuteTaskSolution_SendGoal.js');
      },
      get GetResultService() {
        return require('./moveit_task_constructor_msgs__action__ExecuteTaskSolution_GetResult.js');
      },
      get FeedbackMessage() {
        return require('./moveit_task_constructor_msgs__action__ExecuteTaskSolution_FeedbackMessage.js');
      },
      get CancelGoal() {
        return require('../action_msgs/action_msgs__srv__CancelGoal.js');
      },
      get GoalStatusArray() {
        return require('../action_msgs/action_msgs__msg__GoalStatusArray.js');
      }
    };
  }
  static type() {
    return {pkgName: 'moveit_task_constructor_msgs', subFolder: 'action', interfaceName: 'ExecuteTaskSolution'};
  }
}
module.exports = moveit_task_constructor_msgs__action__ExecuteTaskSolution;
