// rclnodejs/generated/moveit_task_constructor_msgs/moveit_task_constructor_msgs__msg__TrajectoryExecutionInfo.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const TrajectoryExecutionInfoRefStruct = StructType({
  controller_names: StringWrapper.refObjectArrayType,
});
const TrajectoryExecutionInfoRefArray = ArrayType(TrajectoryExecutionInfoRefStruct);
const TrajectoryExecutionInfoRefStructArray = StructType({
  data: TrajectoryExecutionInfoRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class TrajectoryExecutionInfoWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new TrajectoryExecutionInfoRefStruct(other._refObject.toObject());
      this._wrapperFields.controller_names = StringWrapper.createArray();
      this._wrapperFields.controller_names.copy(other._wrapperFields.controller_names);
      this.controller_names = other.controller_names;
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new TrajectoryExecutionInfoRefStruct();
    this._wrapperFields.controller_names = StringWrapper.createArray();
    this._controller_namesArray = [];
  }
  static createFromRefObject(refObject) {
    let self = new TrajectoryExecutionInfoWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new TrajectoryExecutionInfoArrayWrapper();
  }
  static get ArrayType() {
    return TrajectoryExecutionInfoArrayWrapper;
  }
  static get refObjectArrayType() {
    return TrajectoryExecutionInfoRefStructArray;
  }
  static get refObjectType() {
    return TrajectoryExecutionInfoRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.controller_names.fill([]);
    } else {
      this._wrapperFields.controller_names.fill(this._controller_namesArray);
    }
    this._wrapperFields.controller_names.freeze(own);
    this._refObject.controller_names = this._wrapperFields.controller_names.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    refObject.controller_names.data.length = refObject.controller_names.size;
    for (let index = 0; index < refObject.controller_names.size; index++) {
      this._controller_namesArray[index] = refObject.controller_names.data[index].data;
    }
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.controller_names.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.controller_names);
      if (!StringWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.controller_names, StringWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    TrajectoryExecutionInfoWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_task_constructor_msgs', subFolder: 'msg', interfaceName: 'TrajectoryExecutionInfo'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get controller_names() {
    return this._controller_namesArray;
  }
  set controller_names(value) {
    this._controller_namesArray = value;
  }
  copyRefObject(refObject) {
    this._refObject = new TrajectoryExecutionInfoRefStruct(refObject.toObject());
    refObject.controller_names.data.length = refObject.controller_names.size;
    for (let index = 0; index < refObject.controller_names.size; index++) {
      this._controller_namesArray[index] = refObject.controller_names.data[index].data;
    }
  }
  copy(other) {
    this._refObject = new TrajectoryExecutionInfoRefStruct(other._refObject.toObject());
    this._controller_namesArray = other._controller_namesArray.slice();
  }
  static get classType() {
    return TrajectoryExecutionInfoWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"controller_names","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"moveit_task_constructor_msgs","type":"TrajectoryExecutionInfo","stringUpperBound":null,"isPrimitiveType":false},"msgName":"TrajectoryExecutionInfo"};
  }
  hasMember(name) {
    let memberNames = ["controller_names"];
    return memberNames.indexOf(name) !== -1;
  }
}
class TrajectoryExecutionInfoArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof TrajectoryExecutionInfoWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new TrajectoryExecutionInfoWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TrajectoryExecutionInfoArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TrajectoryExecutionInfoArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new TrajectoryExecutionInfoRefArray(size);
    this._refObject = new TrajectoryExecutionInfoRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new TrajectoryExecutionInfoWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof TrajectoryExecutionInfoArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "TrajectoryExecutionInfoArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      TrajectoryExecutionInfoWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return TrajectoryExecutionInfoWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return TrajectoryExecutionInfoArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = TrajectoryExecutionInfoWrapper;
