// rclnodejs/generated/moveit_task_constructor_msgs/moveit_task_constructor_msgs__srv__GetSolution_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const SolutionWrapper = require('../../generated/moveit_task_constructor_msgs/moveit_task_constructor_msgs__msg__Solution.js');
const GetSolution_ResponseRefStruct = StructType({
  solution: SolutionWrapper.refObjectType,
});
const GetSolution_ResponseRefArray = ArrayType(GetSolution_ResponseRefStruct);
const GetSolution_ResponseRefStructArray = StructType({
  data: GetSolution_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class GetSolution_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetSolution_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.solution =  new SolutionWrapper(other._wrapperFields.solution);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new GetSolution_ResponseRefStruct();
    this._wrapperFields.solution = new SolutionWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GetSolution_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetSolution_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return GetSolution_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetSolution_ResponseRefStructArray;
  }
  static get refObjectType() {
    return GetSolution_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.solution.freeze(own);
    this._refObject.solution = this._wrapperFields.solution.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.solution.copyRefObject(refObject.solution);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    SolutionWrapper.freeStruct(refObject.solution);
  }
  static destroyRawROS(msg) {
    GetSolution_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_task_constructor_msgs', subFolder: 'srv', interfaceName: 'GetSolution_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get solution() {
    return this._wrapperFields.solution;
  }
  set solution(value) {
    if (value instanceof SolutionWrapper) {
      this._wrapperFields.solution.copy(value);
    } else {
      this._wrapperFields.solution.copy(new SolutionWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new GetSolution_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.solution.copyRefObject(this._refObject.solution);
  }
  copy(other) {
    this._refObject = new GetSolution_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.solution.copy(other._wrapperFields.solution);
  }
  static get classType() {
    return GetSolution_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"solution","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_task_constructor_msgs","type":"Solution","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_task_constructor_msgs","type":"GetSolution_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetSolution_Response"};
  }
  hasMember(name) {
    let memberNames = ["solution"];
    return memberNames.indexOf(name) !== -1;
  }
}
class GetSolution_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetSolution_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetSolution_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetSolution_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetSolution_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new GetSolution_ResponseRefArray(size);
    this._refObject = new GetSolution_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetSolution_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetSolution_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetSolution_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetSolution_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetSolution_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetSolution_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = GetSolution_ResponseWrapper;
