// rclnodejs/generated/moveit_task_constructor_msgs/moveit_task_constructor_msgs__msg__SolutionInfo.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const MarkerWrapper = require('../../generated/visualization_msgs/visualization_msgs__msg__Marker.js');
const SolutionInfoRefStruct = StructType({
  id: primitiveTypes.uint32,
  cost: primitiveTypes.float32,
  comment: primitiveTypes.string,
  stage_id: primitiveTypes.uint32,
  planner_id: primitiveTypes.string,
  markers: MarkerWrapper.refObjectArrayType,
});
const SolutionInfoRefArray = ArrayType(SolutionInfoRefStruct);
const SolutionInfoRefStructArray = StructType({
  data: SolutionInfoRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class SolutionInfoWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new SolutionInfoRefStruct(other._refObject.toObject());
      this._wrapperFields.comment =  new StringWrapper(other._wrapperFields.comment);
      this._wrapperFields.planner_id =  new StringWrapper(other._wrapperFields.planner_id);
      this._wrapperFields.markers = MarkerWrapper.createArray();
      this._wrapperFields.markers.copy(other._wrapperFields.markers);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new SolutionInfoRefStruct();
    this._wrapperFields.comment = new StringWrapper();
    this._wrapperFields.planner_id = new StringWrapper();
    this._wrapperFields.markers = MarkerWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new SolutionInfoWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SolutionInfoArrayWrapper();
  }
  static get ArrayType() {
    return SolutionInfoArrayWrapper;
  }
  static get refObjectArrayType() {
    return SolutionInfoRefStructArray;
  }
  static get refObjectType() {
    return SolutionInfoRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.comment.freeze(own);
    }
    this._refObject.comment = this._wrapperFields.comment.refObject;
    if (own) {
      this._wrapperFields.planner_id.freeze(own);
    }
    this._refObject.planner_id = this._wrapperFields.planner_id.refObject;
    if (own) {
      this._wrapperFields.markers.fill([]);
    }
    this._wrapperFields.markers.freeze(own);
    this._refObject.markers = this._wrapperFields.markers.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.id = refObject.id;
    this._refObject.cost = refObject.cost;
    this._wrapperFields.comment.data = refObject.comment.data;
    this._refObject.stage_id = refObject.stage_id;
    this._wrapperFields.planner_id.data = refObject.planner_id.data;
    this._wrapperFields.markers.copyRefObject(refObject.markers);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.comment);
    StringWrapper.freeStruct(refObject.planner_id);
    if (refObject.markers.size != 0) {
      MarkerWrapper.ArrayType.freeArray(refObject.markers);
      if (!MarkerWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.markers, MarkerWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    SolutionInfoWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_task_constructor_msgs', subFolder: 'msg', interfaceName: 'SolutionInfo'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get id() {
    return this._refObject.id;
  }
  set id(value) {
    this._refObject.id = value;
  }
  get cost() {
    return this._refObject.cost;
  }
  set cost(value) {
    this._refObject.cost = value;
  }
  get comment() {
    return this._wrapperFields.comment.data;
  }
  set comment(value) {
    this._wrapperFields.comment.data = value;
  }
  get stage_id() {
    return this._refObject.stage_id;
  }
  set stage_id(value) {
    this._refObject.stage_id = value;
  }
  get planner_id() {
    return this._wrapperFields.planner_id.data;
  }
  set planner_id(value) {
    this._wrapperFields.planner_id.data = value;
  }
  get markers() {
    return this._wrapperFields.markers;
  }
  set markers(value) {
    this._wrapperFields.markers.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new SolutionInfoRefStruct(refObject.toObject());
    this._wrapperFields.comment.copyRefObject(this._refObject.comment);
    this._wrapperFields.planner_id.copyRefObject(this._refObject.planner_id);
    this._wrapperFields.markers.copyRefObject(this._refObject.markers);
  }
  copy(other) {
    this._refObject = new SolutionInfoRefStruct(other._refObject.toObject());
    this._wrapperFields.comment.copy(other._wrapperFields.comment);
    this._wrapperFields.planner_id.copy(other._wrapperFields.planner_id);
    this._wrapperFields.markers.copy(other._wrapperFields.markers);
  }
  static get classType() {
    return SolutionInfoWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cost","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"comment","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"stage_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"planner_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"markers","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"visualization_msgs","type":"Marker","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_task_constructor_msgs","type":"SolutionInfo","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SolutionInfo"};
  }
  hasMember(name) {
    let memberNames = ["id","cost","comment","stage_id","planner_id","markers"];
    return memberNames.indexOf(name) !== -1;
  }
}
class SolutionInfoArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SolutionInfoWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SolutionInfoWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SolutionInfoArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SolutionInfoArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new SolutionInfoRefArray(size);
    this._refObject = new SolutionInfoRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SolutionInfoWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SolutionInfoArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SolutionInfoArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SolutionInfoWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SolutionInfoWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SolutionInfoArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = SolutionInfoWrapper;
