// rclnodejs/generated/moveit_task_constructor_msgs/moveit_task_constructor_msgs__msg__SubSolution.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const SolutionInfoWrapper = require('../../generated/moveit_task_constructor_msgs/moveit_task_constructor_msgs__msg__SolutionInfo.js');
const UInt32Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt32.js');
const SubSolutionRefStruct = StructType({
  info: SolutionInfoWrapper.refObjectType,
  sub_solution_id: UInt32Wrapper.refObjectArrayType,
});
const SubSolutionRefArray = ArrayType(SubSolutionRefStruct);
const SubSolutionRefStructArray = StructType({
  data: SubSolutionRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class SubSolutionWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new SubSolutionRefStruct(other._refObject.toObject());
      this._wrapperFields.info =  new SolutionInfoWrapper(other._wrapperFields.info);
      this._wrapperFields.sub_solution_id = UInt32Wrapper.createArray();
      this._wrapperFields.sub_solution_id.copy(other._wrapperFields.sub_solution_id);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new SubSolutionRefStruct();
    this._wrapperFields.info = new SolutionInfoWrapper();
    this._wrapperFields.sub_solution_id = UInt32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new SubSolutionWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SubSolutionArrayWrapper();
  }
  static get ArrayType() {
    return SubSolutionArrayWrapper;
  }
  static get refObjectArrayType() {
    return SubSolutionRefStructArray;
  }
  static get refObjectType() {
    return SubSolutionRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.info.freeze(own);
    this._refObject.info = this._wrapperFields.info.refObject;
    if (own) {
      this._wrapperFields.sub_solution_id.fill(Uint32Array.from([]));
    }
    this._wrapperFields.sub_solution_id.freeze(own);
    this._refObject.sub_solution_id = this._wrapperFields.sub_solution_id.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.info.copyRefObject(refObject.info);
    this._wrapperFields.sub_solution_id.copyRefObject(refObject.sub_solution_id);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    SolutionInfoWrapper.freeStruct(refObject.info);
    if (refObject.sub_solution_id.size != 0) {
      UInt32Wrapper.ArrayType.freeArray(refObject.sub_solution_id);
      if (!UInt32Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.sub_solution_id, UInt32Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    SubSolutionWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_task_constructor_msgs', subFolder: 'msg', interfaceName: 'SubSolution'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get info() {
    return this._wrapperFields.info;
  }
  set info(value) {
    if (value instanceof SolutionInfoWrapper) {
      this._wrapperFields.info.copy(value);
    } else {
      this._wrapperFields.info.copy(new SolutionInfoWrapper(value));
    }
  }
  get sub_solution_id() {
    return this._wrapperFields['sub_solution_id'].data;
  }
  set sub_solution_id(value) {
    this._wrapperFields['sub_solution_id'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new SubSolutionRefStruct(refObject.toObject());
    this._wrapperFields.info.copyRefObject(this._refObject.info);
    this._wrapperFields.sub_solution_id.copyRefObject(this._refObject.sub_solution_id);
  }
  copy(other) {
    this._refObject = new SubSolutionRefStruct(other._refObject.toObject());
    this._wrapperFields.info.copy(other._wrapperFields.info);
    this._wrapperFields.sub_solution_id.copy(other._wrapperFields.sub_solution_id);
  }
  static get classType() {
    return SubSolutionWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"info","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_task_constructor_msgs","type":"SolutionInfo","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"sub_solution_id","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"moveit_task_constructor_msgs","type":"SubSolution","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SubSolution"};
  }
  hasMember(name) {
    let memberNames = ["info","sub_solution_id"];
    return memberNames.indexOf(name) !== -1;
  }
}
class SubSolutionArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SubSolutionWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SubSolutionWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SubSolutionArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SubSolutionArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new SubSolutionRefArray(size);
    this._refObject = new SubSolutionRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SubSolutionWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SubSolutionArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SubSolutionArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SubSolutionWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SubSolutionWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SubSolutionArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = SubSolutionWrapper;
