// rclnodejs/generated/moveit_task_constructor_msgs/moveit_task_constructor_msgs__msg__TaskStatistics.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const StageStatisticsWrapper = require('../../generated/moveit_task_constructor_msgs/moveit_task_constructor_msgs__msg__StageStatistics.js');
const TaskStatisticsRefStruct = StructType({
  task_id: primitiveTypes.string,
  stages: StageStatisticsWrapper.refObjectArrayType,
});
const TaskStatisticsRefArray = ArrayType(TaskStatisticsRefStruct);
const TaskStatisticsRefStructArray = StructType({
  data: TaskStatisticsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class TaskStatisticsWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new TaskStatisticsRefStruct(other._refObject.toObject());
      this._wrapperFields.task_id =  new StringWrapper(other._wrapperFields.task_id);
      this._wrapperFields.stages = StageStatisticsWrapper.createArray();
      this._wrapperFields.stages.copy(other._wrapperFields.stages);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new TaskStatisticsRefStruct();
    this._wrapperFields.task_id = new StringWrapper();
    this._wrapperFields.stages = StageStatisticsWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new TaskStatisticsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new TaskStatisticsArrayWrapper();
  }
  static get ArrayType() {
    return TaskStatisticsArrayWrapper;
  }
  static get refObjectArrayType() {
    return TaskStatisticsRefStructArray;
  }
  static get refObjectType() {
    return TaskStatisticsRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.task_id.freeze(own);
    }
    this._refObject.task_id = this._wrapperFields.task_id.refObject;
    if (own) {
      this._wrapperFields.stages.fill([]);
    }
    this._wrapperFields.stages.freeze(own);
    this._refObject.stages = this._wrapperFields.stages.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.task_id.data = refObject.task_id.data;
    this._wrapperFields.stages.copyRefObject(refObject.stages);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.task_id);
    if (refObject.stages.size != 0) {
      StageStatisticsWrapper.ArrayType.freeArray(refObject.stages);
      if (!StageStatisticsWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.stages, StageStatisticsWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    TaskStatisticsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_task_constructor_msgs', subFolder: 'msg', interfaceName: 'TaskStatistics'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get task_id() {
    return this._wrapperFields.task_id.data;
  }
  set task_id(value) {
    this._wrapperFields.task_id.data = value;
  }
  get stages() {
    return this._wrapperFields.stages;
  }
  set stages(value) {
    this._wrapperFields.stages.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new TaskStatisticsRefStruct(refObject.toObject());
    this._wrapperFields.task_id.copyRefObject(this._refObject.task_id);
    this._wrapperFields.stages.copyRefObject(this._refObject.stages);
  }
  copy(other) {
    this._refObject = new TaskStatisticsRefStruct(other._refObject.toObject());
    this._wrapperFields.task_id.copy(other._wrapperFields.task_id);
    this._wrapperFields.stages.copy(other._wrapperFields.stages);
  }
  static get classType() {
    return TaskStatisticsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"task_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"stages","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"moveit_task_constructor_msgs","type":"StageStatistics","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_task_constructor_msgs","type":"TaskStatistics","stringUpperBound":null,"isPrimitiveType":false},"msgName":"TaskStatistics"};
  }
  hasMember(name) {
    let memberNames = ["task_id","stages"];
    return memberNames.indexOf(name) !== -1;
  }
}
class TaskStatisticsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof TaskStatisticsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new TaskStatisticsWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TaskStatisticsArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TaskStatisticsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new TaskStatisticsRefArray(size);
    this._refObject = new TaskStatisticsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new TaskStatisticsWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof TaskStatisticsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "TaskStatisticsArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      TaskStatisticsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return TaskStatisticsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return TaskStatisticsArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = TaskStatisticsWrapper;
