// rclnodejs/generated/moveit_task_constructor_msgs/moveit_task_constructor_msgs__msg__Solution.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const PlanningSceneWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__PlanningScene.js');
const SubSolutionWrapper = require('../../generated/moveit_task_constructor_msgs/moveit_task_constructor_msgs__msg__SubSolution.js');
const SubTrajectoryWrapper = require('../../generated/moveit_task_constructor_msgs/moveit_task_constructor_msgs__msg__SubTrajectory.js');
const SolutionRefStruct = StructType({
  task_id: primitiveTypes.string,
  start_scene: PlanningSceneWrapper.refObjectType,
  sub_solution: SubSolutionWrapper.refObjectArrayType,
  sub_trajectory: SubTrajectoryWrapper.refObjectArrayType,
});
const SolutionRefArray = ArrayType(SolutionRefStruct);
const SolutionRefStructArray = StructType({
  data: SolutionRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class SolutionWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new SolutionRefStruct(other._refObject.toObject());
      this._wrapperFields.task_id =  new StringWrapper(other._wrapperFields.task_id);
      this._wrapperFields.start_scene =  new PlanningSceneWrapper(other._wrapperFields.start_scene);
      this._wrapperFields.sub_solution = SubSolutionWrapper.createArray();
      this._wrapperFields.sub_solution.copy(other._wrapperFields.sub_solution);
      this._wrapperFields.sub_trajectory = SubTrajectoryWrapper.createArray();
      this._wrapperFields.sub_trajectory.copy(other._wrapperFields.sub_trajectory);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new SolutionRefStruct();
    this._wrapperFields.task_id = new StringWrapper();
    this._wrapperFields.start_scene = new PlanningSceneWrapper();
    this._wrapperFields.sub_solution = SubSolutionWrapper.createArray();
    this._wrapperFields.sub_trajectory = SubTrajectoryWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new SolutionWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SolutionArrayWrapper();
  }
  static get ArrayType() {
    return SolutionArrayWrapper;
  }
  static get refObjectArrayType() {
    return SolutionRefStructArray;
  }
  static get refObjectType() {
    return SolutionRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.task_id.freeze(own);
    }
    this._refObject.task_id = this._wrapperFields.task_id.refObject;
    this._wrapperFields.start_scene.freeze(own);
    this._refObject.start_scene = this._wrapperFields.start_scene.refObject;
    if (own) {
      this._wrapperFields.sub_solution.fill([]);
    }
    this._wrapperFields.sub_solution.freeze(own);
    this._refObject.sub_solution = this._wrapperFields.sub_solution.refObject;
    if (own) {
      this._wrapperFields.sub_trajectory.fill([]);
    }
    this._wrapperFields.sub_trajectory.freeze(own);
    this._refObject.sub_trajectory = this._wrapperFields.sub_trajectory.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.task_id.data = refObject.task_id.data;
    this._wrapperFields.start_scene.copyRefObject(refObject.start_scene);
    this._wrapperFields.sub_solution.copyRefObject(refObject.sub_solution);
    this._wrapperFields.sub_trajectory.copyRefObject(refObject.sub_trajectory);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.task_id);
    PlanningSceneWrapper.freeStruct(refObject.start_scene);
    if (refObject.sub_solution.size != 0) {
      SubSolutionWrapper.ArrayType.freeArray(refObject.sub_solution);
      if (!SubSolutionWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.sub_solution, SubSolutionWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.sub_trajectory.size != 0) {
      SubTrajectoryWrapper.ArrayType.freeArray(refObject.sub_trajectory);
      if (!SubTrajectoryWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.sub_trajectory, SubTrajectoryWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    SolutionWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_task_constructor_msgs', subFolder: 'msg', interfaceName: 'Solution'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get task_id() {
    return this._wrapperFields.task_id.data;
  }
  set task_id(value) {
    this._wrapperFields.task_id.data = value;
  }
  get start_scene() {
    return this._wrapperFields.start_scene;
  }
  set start_scene(value) {
    if (value instanceof PlanningSceneWrapper) {
      this._wrapperFields.start_scene.copy(value);
    } else {
      this._wrapperFields.start_scene.copy(new PlanningSceneWrapper(value));
    }
  }
  get sub_solution() {
    return this._wrapperFields.sub_solution;
  }
  set sub_solution(value) {
    this._wrapperFields.sub_solution.fill(value);
  }
  get sub_trajectory() {
    return this._wrapperFields.sub_trajectory;
  }
  set sub_trajectory(value) {
    this._wrapperFields.sub_trajectory.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new SolutionRefStruct(refObject.toObject());
    this._wrapperFields.task_id.copyRefObject(this._refObject.task_id);
    this._wrapperFields.start_scene.copyRefObject(this._refObject.start_scene);
    this._wrapperFields.sub_solution.copyRefObject(this._refObject.sub_solution);
    this._wrapperFields.sub_trajectory.copyRefObject(this._refObject.sub_trajectory);
  }
  copy(other) {
    this._refObject = new SolutionRefStruct(other._refObject.toObject());
    this._wrapperFields.task_id.copy(other._wrapperFields.task_id);
    this._wrapperFields.start_scene.copy(other._wrapperFields.start_scene);
    this._wrapperFields.sub_solution.copy(other._wrapperFields.sub_solution);
    this._wrapperFields.sub_trajectory.copy(other._wrapperFields.sub_trajectory);
  }
  static get classType() {
    return SolutionWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"task_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"start_scene","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"PlanningScene","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"sub_solution","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"moveit_task_constructor_msgs","type":"SubSolution","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"sub_trajectory","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"moveit_task_constructor_msgs","type":"SubTrajectory","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_task_constructor_msgs","type":"Solution","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Solution"};
  }
  hasMember(name) {
    let memberNames = ["task_id","start_scene","sub_solution","sub_trajectory"];
    return memberNames.indexOf(name) !== -1;
  }
}
class SolutionArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SolutionWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SolutionWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SolutionArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SolutionArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new SolutionRefArray(size);
    this._refObject = new SolutionRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SolutionWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SolutionArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SolutionArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SolutionWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SolutionWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SolutionArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = SolutionWrapper;
