// rclnodejs/generated/moveit_task_constructor_msgs/moveit_task_constructor_msgs__msg__SubTrajectory.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const SolutionInfoWrapper = require('../../generated/moveit_task_constructor_msgs/moveit_task_constructor_msgs__msg__SolutionInfo.js');
const TrajectoryExecutionInfoWrapper = require('../../generated/moveit_task_constructor_msgs/moveit_task_constructor_msgs__msg__TrajectoryExecutionInfo.js');
const RobotTrajectoryWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__RobotTrajectory.js');
const PlanningSceneWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__PlanningScene.js');
const SubTrajectoryRefStruct = StructType({
  info: SolutionInfoWrapper.refObjectType,
  execution_info: TrajectoryExecutionInfoWrapper.refObjectType,
  trajectory: RobotTrajectoryWrapper.refObjectType,
  scene_diff: PlanningSceneWrapper.refObjectType,
});
const SubTrajectoryRefArray = ArrayType(SubTrajectoryRefStruct);
const SubTrajectoryRefStructArray = StructType({
  data: SubTrajectoryRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class SubTrajectoryWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new SubTrajectoryRefStruct(other._refObject.toObject());
      this._wrapperFields.info =  new SolutionInfoWrapper(other._wrapperFields.info);
      this._wrapperFields.execution_info =  new TrajectoryExecutionInfoWrapper(other._wrapperFields.execution_info);
      this._wrapperFields.trajectory =  new RobotTrajectoryWrapper(other._wrapperFields.trajectory);
      this._wrapperFields.scene_diff =  new PlanningSceneWrapper(other._wrapperFields.scene_diff);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new SubTrajectoryRefStruct();
    this._wrapperFields.info = new SolutionInfoWrapper();
    this._wrapperFields.execution_info = new TrajectoryExecutionInfoWrapper();
    this._wrapperFields.trajectory = new RobotTrajectoryWrapper();
    this._wrapperFields.scene_diff = new PlanningSceneWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new SubTrajectoryWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SubTrajectoryArrayWrapper();
  }
  static get ArrayType() {
    return SubTrajectoryArrayWrapper;
  }
  static get refObjectArrayType() {
    return SubTrajectoryRefStructArray;
  }
  static get refObjectType() {
    return SubTrajectoryRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.info.freeze(own);
    this._refObject.info = this._wrapperFields.info.refObject;
    this._wrapperFields.execution_info.freeze(own);
    this._refObject.execution_info = this._wrapperFields.execution_info.refObject;
    this._wrapperFields.trajectory.freeze(own);
    this._refObject.trajectory = this._wrapperFields.trajectory.refObject;
    this._wrapperFields.scene_diff.freeze(own);
    this._refObject.scene_diff = this._wrapperFields.scene_diff.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.info.copyRefObject(refObject.info);
    this._wrapperFields.execution_info.copyRefObject(refObject.execution_info);
    this._wrapperFields.trajectory.copyRefObject(refObject.trajectory);
    this._wrapperFields.scene_diff.copyRefObject(refObject.scene_diff);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    SolutionInfoWrapper.freeStruct(refObject.info);
    TrajectoryExecutionInfoWrapper.freeStruct(refObject.execution_info);
    RobotTrajectoryWrapper.freeStruct(refObject.trajectory);
    PlanningSceneWrapper.freeStruct(refObject.scene_diff);
  }
  static destroyRawROS(msg) {
    SubTrajectoryWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_task_constructor_msgs', subFolder: 'msg', interfaceName: 'SubTrajectory'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get info() {
    return this._wrapperFields.info;
  }
  set info(value) {
    if (value instanceof SolutionInfoWrapper) {
      this._wrapperFields.info.copy(value);
    } else {
      this._wrapperFields.info.copy(new SolutionInfoWrapper(value));
    }
  }
  get execution_info() {
    return this._wrapperFields.execution_info;
  }
  set execution_info(value) {
    if (value instanceof TrajectoryExecutionInfoWrapper) {
      this._wrapperFields.execution_info.copy(value);
    } else {
      this._wrapperFields.execution_info.copy(new TrajectoryExecutionInfoWrapper(value));
    }
  }
  get trajectory() {
    return this._wrapperFields.trajectory;
  }
  set trajectory(value) {
    if (value instanceof RobotTrajectoryWrapper) {
      this._wrapperFields.trajectory.copy(value);
    } else {
      this._wrapperFields.trajectory.copy(new RobotTrajectoryWrapper(value));
    }
  }
  get scene_diff() {
    return this._wrapperFields.scene_diff;
  }
  set scene_diff(value) {
    if (value instanceof PlanningSceneWrapper) {
      this._wrapperFields.scene_diff.copy(value);
    } else {
      this._wrapperFields.scene_diff.copy(new PlanningSceneWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new SubTrajectoryRefStruct(refObject.toObject());
    this._wrapperFields.info.copyRefObject(this._refObject.info);
    this._wrapperFields.execution_info.copyRefObject(this._refObject.execution_info);
    this._wrapperFields.trajectory.copyRefObject(this._refObject.trajectory);
    this._wrapperFields.scene_diff.copyRefObject(this._refObject.scene_diff);
  }
  copy(other) {
    this._refObject = new SubTrajectoryRefStruct(other._refObject.toObject());
    this._wrapperFields.info.copy(other._wrapperFields.info);
    this._wrapperFields.execution_info.copy(other._wrapperFields.execution_info);
    this._wrapperFields.trajectory.copy(other._wrapperFields.trajectory);
    this._wrapperFields.scene_diff.copy(other._wrapperFields.scene_diff);
  }
  static get classType() {
    return SubTrajectoryWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"info","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_task_constructor_msgs","type":"SolutionInfo","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"execution_info","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_task_constructor_msgs","type":"TrajectoryExecutionInfo","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"trajectory","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"RobotTrajectory","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"scene_diff","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"PlanningScene","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_task_constructor_msgs","type":"SubTrajectory","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SubTrajectory"};
  }
  hasMember(name) {
    let memberNames = ["info","execution_info","trajectory","scene_diff"];
    return memberNames.indexOf(name) !== -1;
  }
}
class SubTrajectoryArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SubTrajectoryWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SubTrajectoryWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SubTrajectoryArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SubTrajectoryArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new SubTrajectoryRefArray(size);
    this._refObject = new SubTrajectoryRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SubTrajectoryWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SubTrajectoryArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SubTrajectoryArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SubTrajectoryWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SubTrajectoryWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SubTrajectoryArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = SubTrajectoryWrapper;
