// rclnodejs/generated/moveit_task_constructor_msgs/moveit_task_constructor_msgs__msg__StageDescription.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const PropertyWrapper = require('../../generated/moveit_task_constructor_msgs/moveit_task_constructor_msgs__msg__Property.js');
const StageDescriptionRefStruct = StructType({
  id: primitiveTypes.uint32,
  parent_id: primitiveTypes.uint32,
  name: primitiveTypes.string,
  flags: primitiveTypes.uint32,
  properties: PropertyWrapper.refObjectArrayType,
});
const StageDescriptionRefArray = ArrayType(StageDescriptionRefStruct);
const StageDescriptionRefStructArray = StructType({
  data: StageDescriptionRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class StageDescriptionWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new StageDescriptionRefStruct(other._refObject.toObject());
      this._wrapperFields.name =  new StringWrapper(other._wrapperFields.name);
      this._wrapperFields.properties = PropertyWrapper.createArray();
      this._wrapperFields.properties.copy(other._wrapperFields.properties);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new StageDescriptionRefStruct();
    this._wrapperFields.name = new StringWrapper();
    this._wrapperFields.properties = PropertyWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new StageDescriptionWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new StageDescriptionArrayWrapper();
  }
  static get ArrayType() {
    return StageDescriptionArrayWrapper;
  }
  static get refObjectArrayType() {
    return StageDescriptionRefStructArray;
  }
  static get refObjectType() {
    return StageDescriptionRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.name.freeze(own);
    }
    this._refObject.name = this._wrapperFields.name.refObject;
    if (own) {
      this._wrapperFields.properties.fill([]);
    }
    this._wrapperFields.properties.freeze(own);
    this._refObject.properties = this._wrapperFields.properties.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.id = refObject.id;
    this._refObject.parent_id = refObject.parent_id;
    this._wrapperFields.name.data = refObject.name.data;
    this._refObject.flags = refObject.flags;
    this._wrapperFields.properties.copyRefObject(refObject.properties);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.name);
    if (refObject.properties.size != 0) {
      PropertyWrapper.ArrayType.freeArray(refObject.properties);
      if (!PropertyWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.properties, PropertyWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    StageDescriptionWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_task_constructor_msgs', subFolder: 'msg', interfaceName: 'StageDescription'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get id() {
    return this._refObject.id;
  }
  set id(value) {
    this._refObject.id = value;
  }
  get parent_id() {
    return this._refObject.parent_id;
  }
  set parent_id(value) {
    this._refObject.parent_id = value;
  }
  get name() {
    return this._wrapperFields.name.data;
  }
  set name(value) {
    this._wrapperFields.name.data = value;
  }
  get flags() {
    return this._refObject.flags;
  }
  set flags(value) {
    this._refObject.flags = value;
  }
  get properties() {
    return this._wrapperFields.properties;
  }
  set properties(value) {
    this._wrapperFields.properties.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new StageDescriptionRefStruct(refObject.toObject());
    this._wrapperFields.name.copyRefObject(this._refObject.name);
    this._wrapperFields.properties.copyRefObject(this._refObject.properties);
  }
  copy(other) {
    this._refObject = new StageDescriptionRefStruct(other._refObject.toObject());
    this._wrapperFields.name.copy(other._wrapperFields.name);
    this._wrapperFields.properties.copy(other._wrapperFields.properties);
  }
  static get classType() {
    return StageDescriptionWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"parent_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"flags","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"properties","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"moveit_task_constructor_msgs","type":"Property","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_task_constructor_msgs","type":"StageDescription","stringUpperBound":null,"isPrimitiveType":false},"msgName":"StageDescription"};
  }
  hasMember(name) {
    let memberNames = ["id","parent_id","name","flags","properties"];
    return memberNames.indexOf(name) !== -1;
  }
}
class StageDescriptionArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof StageDescriptionWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new StageDescriptionWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StageDescriptionArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StageDescriptionArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new StageDescriptionRefArray(size);
    this._refObject = new StageDescriptionRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new StageDescriptionWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof StageDescriptionArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "StageDescriptionArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      StageDescriptionWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return StageDescriptionWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return StageDescriptionArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = StageDescriptionWrapper;
