// rclnodejs/generated/moveit_task_constructor_msgs/moveit_task_constructor_msgs__msg__StageStatistics.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const UInt32Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt32.js');
const StageStatisticsRefStruct = StructType({
  id: primitiveTypes.uint32,
  solved: UInt32Wrapper.refObjectArrayType,
  failed: UInt32Wrapper.refObjectArrayType,
  num_failed: primitiveTypes.uint32,
  total_compute_time: primitiveTypes.float64,
});
const StageStatisticsRefArray = ArrayType(StageStatisticsRefStruct);
const StageStatisticsRefStructArray = StructType({
  data: StageStatisticsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class StageStatisticsWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new StageStatisticsRefStruct(other._refObject.toObject());
      this._wrapperFields.solved = UInt32Wrapper.createArray();
      this._wrapperFields.solved.copy(other._wrapperFields.solved);
      this._wrapperFields.failed = UInt32Wrapper.createArray();
      this._wrapperFields.failed.copy(other._wrapperFields.failed);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new StageStatisticsRefStruct();
    this._wrapperFields.solved = UInt32Wrapper.createArray();
    this._wrapperFields.failed = UInt32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new StageStatisticsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new StageStatisticsArrayWrapper();
  }
  static get ArrayType() {
    return StageStatisticsArrayWrapper;
  }
  static get refObjectArrayType() {
    return StageStatisticsRefStructArray;
  }
  static get refObjectType() {
    return StageStatisticsRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.solved.fill(Uint32Array.from([]));
    }
    this._wrapperFields.solved.freeze(own);
    this._refObject.solved = this._wrapperFields.solved.refObject;
    if (own) {
      this._wrapperFields.failed.fill(Uint32Array.from([]));
    }
    this._wrapperFields.failed.freeze(own);
    this._refObject.failed = this._wrapperFields.failed.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.id = refObject.id;
    this._wrapperFields.solved.copyRefObject(refObject.solved);
    this._wrapperFields.failed.copyRefObject(refObject.failed);
    this._refObject.num_failed = refObject.num_failed;
    this._refObject.total_compute_time = refObject.total_compute_time;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.solved.size != 0) {
      UInt32Wrapper.ArrayType.freeArray(refObject.solved);
      if (!UInt32Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.solved, UInt32Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.failed.size != 0) {
      UInt32Wrapper.ArrayType.freeArray(refObject.failed);
      if (!UInt32Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.failed, UInt32Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    StageStatisticsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_task_constructor_msgs', subFolder: 'msg', interfaceName: 'StageStatistics'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get id() {
    return this._refObject.id;
  }
  set id(value) {
    this._refObject.id = value;
  }
  get solved() {
    return this._wrapperFields['solved'].data;
  }
  set solved(value) {
    this._wrapperFields['solved'].fill(value);
  }
  get failed() {
    return this._wrapperFields['failed'].data;
  }
  set failed(value) {
    this._wrapperFields['failed'].fill(value);
  }
  get num_failed() {
    return this._refObject.num_failed;
  }
  set num_failed(value) {
    this._refObject.num_failed = value;
  }
  get total_compute_time() {
    return this._refObject.total_compute_time;
  }
  set total_compute_time(value) {
    this._refObject.total_compute_time = value;
  }
  copyRefObject(refObject) {
    this._refObject = new StageStatisticsRefStruct(refObject.toObject());
    this._wrapperFields.solved.copyRefObject(this._refObject.solved);
    this._wrapperFields.failed.copyRefObject(this._refObject.failed);
  }
  copy(other) {
    this._refObject = new StageStatisticsRefStruct(other._refObject.toObject());
    this._wrapperFields.solved.copy(other._wrapperFields.solved);
    this._wrapperFields.failed.copy(other._wrapperFields.failed);
  }
  static get classType() {
    return StageStatisticsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"solved","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"failed","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"num_failed","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"total_compute_time","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"moveit_task_constructor_msgs","type":"StageStatistics","stringUpperBound":null,"isPrimitiveType":false},"msgName":"StageStatistics"};
  }
  hasMember(name) {
    let memberNames = ["id","solved","failed","num_failed","total_compute_time"];
    return memberNames.indexOf(name) !== -1;
  }
}
class StageStatisticsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof StageStatisticsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new StageStatisticsWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StageStatisticsArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StageStatisticsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new StageStatisticsRefArray(size);
    this._refObject = new StageStatisticsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new StageStatisticsWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof StageStatisticsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "StageStatisticsArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      StageStatisticsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return StageStatisticsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return StageStatisticsArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = StageStatisticsWrapper;
