// rclnodejs/generated/ethercat_msgs/ethercat_msgs__srv__SetSdo_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const SetSdo_RequestRefStruct = StructType({
  master_id: primitiveTypes.int16,
  slave_position: primitiveTypes.int16,
  sdo_index: primitiveTypes.int16,
  sdo_subindex: primitiveTypes.int16,
  sdo_data_type: primitiveTypes.string,
  sdo_value: primitiveTypes.string,
});
const SetSdo_RequestRefArray = ArrayType(SetSdo_RequestRefStruct);
const SetSdo_RequestRefStructArray = StructType({
  data: SetSdo_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class SetSdo_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetSdo_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.sdo_data_type =  new StringWrapper(other._wrapperFields.sdo_data_type);
      this._wrapperFields.sdo_value =  new StringWrapper(other._wrapperFields.sdo_value);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new SetSdo_RequestRefStruct();
    this._wrapperFields.sdo_data_type = new StringWrapper();
    this._wrapperFields.sdo_value = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new SetSdo_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetSdo_RequestArrayWrapper();
  }
  static get ArrayType() {
    return SetSdo_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetSdo_RequestRefStructArray;
  }
  static get refObjectType() {
    return SetSdo_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.sdo_data_type.freeze(own);
    }
    this._refObject.sdo_data_type = this._wrapperFields.sdo_data_type.refObject;
    if (own) {
      this._wrapperFields.sdo_value.freeze(own);
    }
    this._refObject.sdo_value = this._wrapperFields.sdo_value.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.master_id = refObject.master_id;
    this._refObject.slave_position = refObject.slave_position;
    this._refObject.sdo_index = refObject.sdo_index;
    this._refObject.sdo_subindex = refObject.sdo_subindex;
    this._wrapperFields.sdo_data_type.data = refObject.sdo_data_type.data;
    this._wrapperFields.sdo_value.data = refObject.sdo_value.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.sdo_data_type);
    StringWrapper.freeStruct(refObject.sdo_value);
  }
  static destroyRawROS(msg) {
    SetSdo_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ethercat_msgs', subFolder: 'srv', interfaceName: 'SetSdo_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get master_id() {
    return this._refObject.master_id;
  }
  set master_id(value) {
    this._refObject.master_id = value;
  }
  get slave_position() {
    return this._refObject.slave_position;
  }
  set slave_position(value) {
    this._refObject.slave_position = value;
  }
  get sdo_index() {
    return this._refObject.sdo_index;
  }
  set sdo_index(value) {
    this._refObject.sdo_index = value;
  }
  get sdo_subindex() {
    return this._refObject.sdo_subindex;
  }
  set sdo_subindex(value) {
    this._refObject.sdo_subindex = value;
  }
  get sdo_data_type() {
    return this._wrapperFields.sdo_data_type.data;
  }
  set sdo_data_type(value) {
    this._wrapperFields.sdo_data_type.data = value;
  }
  get sdo_value() {
    return this._wrapperFields.sdo_value.data;
  }
  set sdo_value(value) {
    this._wrapperFields.sdo_value.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SetSdo_RequestRefStruct(refObject.toObject());
    this._wrapperFields.sdo_data_type.copyRefObject(this._refObject.sdo_data_type);
    this._wrapperFields.sdo_value.copyRefObject(this._refObject.sdo_value);
  }
  copy(other) {
    this._refObject = new SetSdo_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.sdo_data_type.copy(other._wrapperFields.sdo_data_type);
    this._wrapperFields.sdo_value.copy(other._wrapperFields.sdo_value);
  }
  static get classType() {
    return SetSdo_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"master_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"slave_position","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"sdo_index","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"sdo_subindex","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"sdo_data_type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"sdo_value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"ethercat_msgs","type":"SetSdo_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetSdo_Request"};
  }
  hasMember(name) {
    let memberNames = ["master_id","slave_position","sdo_index","sdo_subindex","sdo_data_type","sdo_value"];
    return memberNames.indexOf(name) !== -1;
  }
}
class SetSdo_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetSdo_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetSdo_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetSdo_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetSdo_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new SetSdo_RequestRefArray(size);
    this._refObject = new SetSdo_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetSdo_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetSdo_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetSdo_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetSdo_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetSdo_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetSdo_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = SetSdo_RequestWrapper;
