// rclnodejs/generated/moveit_msgs/moveit_msgs__msg__GenericTrajectory.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const JointTrajectoryWrapper = require('../../generated/trajectory_msgs/trajectory_msgs__msg__JointTrajectory.js');
const CartesianTrajectoryWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__CartesianTrajectory.js');
const GenericTrajectoryRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  joint_trajectory: JointTrajectoryWrapper.refObjectArrayType,
  cartesian_trajectory: CartesianTrajectoryWrapper.refObjectArrayType,
});
const GenericTrajectoryRefArray = ArrayType(GenericTrajectoryRefStruct);
const GenericTrajectoryRefStructArray = StructType({
  data: GenericTrajectoryRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class GenericTrajectoryWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new GenericTrajectoryRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.joint_trajectory = JointTrajectoryWrapper.createArray();
      this._wrapperFields.joint_trajectory.copy(other._wrapperFields.joint_trajectory);
      this._wrapperFields.cartesian_trajectory = CartesianTrajectoryWrapper.createArray();
      this._wrapperFields.cartesian_trajectory.copy(other._wrapperFields.cartesian_trajectory);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new GenericTrajectoryRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.joint_trajectory = JointTrajectoryWrapper.createArray();
    this._wrapperFields.cartesian_trajectory = CartesianTrajectoryWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new GenericTrajectoryWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GenericTrajectoryArrayWrapper();
  }
  static get ArrayType() {
    return GenericTrajectoryArrayWrapper;
  }
  static get refObjectArrayType() {
    return GenericTrajectoryRefStructArray;
  }
  static get refObjectType() {
    return GenericTrajectoryRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.header.freeze(own);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.joint_trajectory.fill([]);
    }
    this._wrapperFields.joint_trajectory.freeze(own);
    this._refObject.joint_trajectory = this._wrapperFields.joint_trajectory.refObject;
    if (own) {
      this._wrapperFields.cartesian_trajectory.fill([]);
    }
    this._wrapperFields.cartesian_trajectory.freeze(own);
    this._refObject.cartesian_trajectory = this._wrapperFields.cartesian_trajectory.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.joint_trajectory.copyRefObject(refObject.joint_trajectory);
    this._wrapperFields.cartesian_trajectory.copyRefObject(refObject.cartesian_trajectory);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.joint_trajectory.size != 0) {
      JointTrajectoryWrapper.ArrayType.freeArray(refObject.joint_trajectory);
      if (!JointTrajectoryWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.joint_trajectory, JointTrajectoryWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.cartesian_trajectory.size != 0) {
      CartesianTrajectoryWrapper.ArrayType.freeArray(refObject.cartesian_trajectory);
      if (!CartesianTrajectoryWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.cartesian_trajectory, CartesianTrajectoryWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    GenericTrajectoryWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'msg', interfaceName: 'GenericTrajectory'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get joint_trajectory() {
    return this._wrapperFields.joint_trajectory;
  }
  set joint_trajectory(value) {
    this._wrapperFields.joint_trajectory.fill(value);
  }
  get cartesian_trajectory() {
    return this._wrapperFields.cartesian_trajectory;
  }
  set cartesian_trajectory(value) {
    this._wrapperFields.cartesian_trajectory.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new GenericTrajectoryRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.joint_trajectory.copyRefObject(this._refObject.joint_trajectory);
    this._wrapperFields.cartesian_trajectory.copyRefObject(this._refObject.cartesian_trajectory);
  }
  copy(other) {
    this._refObject = new GenericTrajectoryRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.joint_trajectory.copy(other._wrapperFields.joint_trajectory);
    this._wrapperFields.cartesian_trajectory.copy(other._wrapperFields.cartesian_trajectory);
  }
  static get classType() {
    return GenericTrajectoryWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"joint_trajectory","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"trajectory_msgs","type":"JointTrajectory","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"cartesian_trajectory","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"moveit_msgs","type":"CartesianTrajectory","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"GenericTrajectory","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GenericTrajectory"};
  }
  hasMember(name) {
    let memberNames = ["header","joint_trajectory","cartesian_trajectory"];
    return memberNames.indexOf(name) !== -1;
  }
}
class GenericTrajectoryArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GenericTrajectoryWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GenericTrajectoryWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GenericTrajectoryArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GenericTrajectoryArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new GenericTrajectoryRefArray(size);
    this._refObject = new GenericTrajectoryRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GenericTrajectoryWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GenericTrajectoryArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GenericTrajectoryArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GenericTrajectoryWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GenericTrajectoryWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GenericTrajectoryArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = GenericTrajectoryWrapper;
