// rclnodejs/generated/moveit_msgs/moveit_msgs__action__HybridPlanner_Result.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const MoveItErrorCodesWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__MoveItErrorCodes.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const HybridPlanner_ResultRefStruct = StructType({
  error_code: MoveItErrorCodesWrapper.refObjectType,
  error_message: primitiveTypes.string,
});
const HybridPlanner_ResultRefArray = ArrayType(HybridPlanner_ResultRefStruct);
const HybridPlanner_ResultRefStructArray = StructType({
  data: HybridPlanner_ResultRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class HybridPlanner_ResultWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new HybridPlanner_ResultRefStruct(other._refObject.toObject());
      this._wrapperFields.error_code =  new MoveItErrorCodesWrapper(other._wrapperFields.error_code);
      this._wrapperFields.error_message =  new StringWrapper(other._wrapperFields.error_message);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new HybridPlanner_ResultRefStruct();
    this._wrapperFields.error_code = new MoveItErrorCodesWrapper();
    this._wrapperFields.error_message = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new HybridPlanner_ResultWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new HybridPlanner_ResultArrayWrapper();
  }
  static get ArrayType() {
    return HybridPlanner_ResultArrayWrapper;
  }
  static get refObjectArrayType() {
    return HybridPlanner_ResultRefStructArray;
  }
  static get refObjectType() {
    return HybridPlanner_ResultRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.error_code.freeze(own);
    this._refObject.error_code = this._wrapperFields.error_code.refObject;
    if (own) {
      this._wrapperFields.error_message.freeze(own);
    }
    this._refObject.error_message = this._wrapperFields.error_message.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.error_code.copyRefObject(refObject.error_code);
    this._wrapperFields.error_message.data = refObject.error_message.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    MoveItErrorCodesWrapper.freeStruct(refObject.error_code);
    StringWrapper.freeStruct(refObject.error_message);
  }
  static destroyRawROS(msg) {
    HybridPlanner_ResultWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'action', interfaceName: 'HybridPlanner_Result'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get error_code() {
    return this._wrapperFields.error_code;
  }
  set error_code(value) {
    if (value instanceof MoveItErrorCodesWrapper) {
      this._wrapperFields.error_code.copy(value);
    } else {
      this._wrapperFields.error_code.copy(new MoveItErrorCodesWrapper(value));
    }
  }
  get error_message() {
    return this._wrapperFields.error_message.data;
  }
  set error_message(value) {
    this._wrapperFields.error_message.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new HybridPlanner_ResultRefStruct(refObject.toObject());
    this._wrapperFields.error_code.copyRefObject(this._refObject.error_code);
    this._wrapperFields.error_message.copyRefObject(this._refObject.error_message);
  }
  copy(other) {
    this._refObject = new HybridPlanner_ResultRefStruct(other._refObject.toObject());
    this._wrapperFields.error_code.copy(other._wrapperFields.error_code);
    this._wrapperFields.error_message.copy(other._wrapperFields.error_message);
  }
  static get classType() {
    return HybridPlanner_ResultWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"error_code","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"MoveItErrorCodes","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"error_message","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"HybridPlanner_Result","stringUpperBound":null,"isPrimitiveType":false},"msgName":"HybridPlanner_Result"};
  }
  hasMember(name) {
    let memberNames = ["error_code","error_message"];
    return memberNames.indexOf(name) !== -1;
  }
}
class HybridPlanner_ResultArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof HybridPlanner_ResultWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new HybridPlanner_ResultWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to HybridPlanner_ResultArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to HybridPlanner_ResultArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new HybridPlanner_ResultRefArray(size);
    this._refObject = new HybridPlanner_ResultRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new HybridPlanner_ResultWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof HybridPlanner_ResultArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "HybridPlanner_ResultArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      HybridPlanner_ResultWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return HybridPlanner_ResultWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return HybridPlanner_ResultArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = HybridPlanner_ResultWrapper;
