// rclnodejs/generated/moveit_msgs/moveit_msgs__action__GlobalPlanner_Feedback.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const GlobalPlanner_FeedbackRefStruct = StructType({
  feedback: primitiveTypes.string,
});
const GlobalPlanner_FeedbackRefArray = ArrayType(GlobalPlanner_FeedbackRefStruct);
const GlobalPlanner_FeedbackRefStructArray = StructType({
  data: GlobalPlanner_FeedbackRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class GlobalPlanner_FeedbackWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new GlobalPlanner_FeedbackRefStruct(other._refObject.toObject());
      this._wrapperFields.feedback =  new StringWrapper(other._wrapperFields.feedback);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new GlobalPlanner_FeedbackRefStruct();
    this._wrapperFields.feedback = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GlobalPlanner_FeedbackWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GlobalPlanner_FeedbackArrayWrapper();
  }
  static get ArrayType() {
    return GlobalPlanner_FeedbackArrayWrapper;
  }
  static get refObjectArrayType() {
    return GlobalPlanner_FeedbackRefStructArray;
  }
  static get refObjectType() {
    return GlobalPlanner_FeedbackRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.feedback.freeze(own);
    }
    this._refObject.feedback = this._wrapperFields.feedback.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.feedback.data = refObject.feedback.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.feedback);
  }
  static destroyRawROS(msg) {
    GlobalPlanner_FeedbackWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'action', interfaceName: 'GlobalPlanner_Feedback'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get feedback() {
    return this._wrapperFields.feedback.data;
  }
  set feedback(value) {
    this._wrapperFields.feedback.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GlobalPlanner_FeedbackRefStruct(refObject.toObject());
    this._wrapperFields.feedback.copyRefObject(this._refObject.feedback);
  }
  copy(other) {
    this._refObject = new GlobalPlanner_FeedbackRefStruct(other._refObject.toObject());
    this._wrapperFields.feedback.copy(other._wrapperFields.feedback);
  }
  static get classType() {
    return GlobalPlanner_FeedbackWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"feedback","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"GlobalPlanner_Feedback","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GlobalPlanner_Feedback"};
  }
  hasMember(name) {
    let memberNames = ["feedback"];
    return memberNames.indexOf(name) !== -1;
  }
}
class GlobalPlanner_FeedbackArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GlobalPlanner_FeedbackWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GlobalPlanner_FeedbackWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GlobalPlanner_FeedbackArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GlobalPlanner_FeedbackArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new GlobalPlanner_FeedbackRefArray(size);
    this._refObject = new GlobalPlanner_FeedbackRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GlobalPlanner_FeedbackWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GlobalPlanner_FeedbackArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GlobalPlanner_FeedbackArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GlobalPlanner_FeedbackWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GlobalPlanner_FeedbackWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GlobalPlanner_FeedbackArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = GlobalPlanner_FeedbackWrapper;
