// rclnodejs/generated/moveit_msgs/moveit_msgs__msg__MotionSequenceRequest.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const MotionSequenceItemWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__MotionSequenceItem.js');
const MotionSequenceRequestRefStruct = StructType({
  items: MotionSequenceItemWrapper.refObjectArrayType,
});
const MotionSequenceRequestRefArray = ArrayType(MotionSequenceRequestRefStruct);
const MotionSequenceRequestRefStructArray = StructType({
  data: MotionSequenceRequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class MotionSequenceRequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new MotionSequenceRequestRefStruct(other._refObject.toObject());
      this._wrapperFields.items = MotionSequenceItemWrapper.createArray();
      this._wrapperFields.items.copy(other._wrapperFields.items);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new MotionSequenceRequestRefStruct();
    this._wrapperFields.items = MotionSequenceItemWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new MotionSequenceRequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MotionSequenceRequestArrayWrapper();
  }
  static get ArrayType() {
    return MotionSequenceRequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return MotionSequenceRequestRefStructArray;
  }
  static get refObjectType() {
    return MotionSequenceRequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.items.fill([]);
    }
    this._wrapperFields.items.freeze(own);
    this._refObject.items = this._wrapperFields.items.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.items.copyRefObject(refObject.items);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.items.size != 0) {
      MotionSequenceItemWrapper.ArrayType.freeArray(refObject.items);
      if (!MotionSequenceItemWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.items, MotionSequenceItemWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    MotionSequenceRequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'msg', interfaceName: 'MotionSequenceRequest'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get items() {
    return this._wrapperFields.items;
  }
  set items(value) {
    this._wrapperFields.items.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new MotionSequenceRequestRefStruct(refObject.toObject());
    this._wrapperFields.items.copyRefObject(this._refObject.items);
  }
  copy(other) {
    this._refObject = new MotionSequenceRequestRefStruct(other._refObject.toObject());
    this._wrapperFields.items.copy(other._wrapperFields.items);
  }
  static get classType() {
    return MotionSequenceRequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"items","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"moveit_msgs","type":"MotionSequenceItem","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"MotionSequenceRequest","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MotionSequenceRequest"};
  }
  hasMember(name) {
    let memberNames = ["items"];
    return memberNames.indexOf(name) !== -1;
  }
}
class MotionSequenceRequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MotionSequenceRequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MotionSequenceRequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MotionSequenceRequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MotionSequenceRequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new MotionSequenceRequestRefArray(size);
    this._refObject = new MotionSequenceRequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MotionSequenceRequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MotionSequenceRequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MotionSequenceRequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MotionSequenceRequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MotionSequenceRequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MotionSequenceRequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = MotionSequenceRequestWrapper;
