// rclnodejs/generated/moveit_msgs/moveit_msgs__action__Place_Goal.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const PlaceLocationWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__PlaceLocation.js');
const ConstraintsWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__Constraints.js');
const PlanningOptionsWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__PlanningOptions.js');
const Place_GoalRefStruct = StructType({
  group_name: primitiveTypes.string,
  attached_object_name: primitiveTypes.string,
  place_locations: PlaceLocationWrapper.refObjectArrayType,
  place_eef: primitiveTypes.bool,
  support_surface_name: primitiveTypes.string,
  allow_gripper_support_collision: primitiveTypes.bool,
  path_constraints: ConstraintsWrapper.refObjectType,
  planner_id: primitiveTypes.string,
  allowed_touch_objects: StringWrapper.refObjectArrayType,
  allowed_planning_time: primitiveTypes.float64,
  planning_options: PlanningOptionsWrapper.refObjectType,
});
const Place_GoalRefArray = ArrayType(Place_GoalRefStruct);
const Place_GoalRefStructArray = StructType({
  data: Place_GoalRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class Place_GoalWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new Place_GoalRefStruct(other._refObject.toObject());
      this._wrapperFields.group_name =  new StringWrapper(other._wrapperFields.group_name);
      this._wrapperFields.attached_object_name =  new StringWrapper(other._wrapperFields.attached_object_name);
      this._wrapperFields.place_locations = PlaceLocationWrapper.createArray();
      this._wrapperFields.place_locations.copy(other._wrapperFields.place_locations);
      this._wrapperFields.support_surface_name =  new StringWrapper(other._wrapperFields.support_surface_name);
      this._wrapperFields.path_constraints =  new ConstraintsWrapper(other._wrapperFields.path_constraints);
      this._wrapperFields.planner_id =  new StringWrapper(other._wrapperFields.planner_id);
      this._wrapperFields.allowed_touch_objects = StringWrapper.createArray();
      this._wrapperFields.allowed_touch_objects.copy(other._wrapperFields.allowed_touch_objects);
      this.allowed_touch_objects = other.allowed_touch_objects;
      this._wrapperFields.planning_options =  new PlanningOptionsWrapper(other._wrapperFields.planning_options);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new Place_GoalRefStruct();
    this._wrapperFields.group_name = new StringWrapper();
    this._wrapperFields.attached_object_name = new StringWrapper();
    this._wrapperFields.place_locations = PlaceLocationWrapper.createArray();
    this._wrapperFields.support_surface_name = new StringWrapper();
    this._wrapperFields.path_constraints = new ConstraintsWrapper();
    this._wrapperFields.planner_id = new StringWrapper();
    this._wrapperFields.allowed_touch_objects = StringWrapper.createArray();
    this._allowed_touch_objectsArray = [];
    this._wrapperFields.planning_options = new PlanningOptionsWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new Place_GoalWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new Place_GoalArrayWrapper();
  }
  static get ArrayType() {
    return Place_GoalArrayWrapper;
  }
  static get refObjectArrayType() {
    return Place_GoalRefStructArray;
  }
  static get refObjectType() {
    return Place_GoalRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.group_name.freeze(own);
    }
    this._refObject.group_name = this._wrapperFields.group_name.refObject;
    if (own) {
      this._wrapperFields.attached_object_name.freeze(own);
    }
    this._refObject.attached_object_name = this._wrapperFields.attached_object_name.refObject;
    if (own) {
      this._wrapperFields.place_locations.fill([]);
    }
    this._wrapperFields.place_locations.freeze(own);
    this._refObject.place_locations = this._wrapperFields.place_locations.refObject;
    if (own) {
      this._wrapperFields.support_surface_name.freeze(own);
    }
    this._refObject.support_surface_name = this._wrapperFields.support_surface_name.refObject;
    this._wrapperFields.path_constraints.freeze(own);
    this._refObject.path_constraints = this._wrapperFields.path_constraints.refObject;
    if (own) {
      this._wrapperFields.planner_id.freeze(own);
    }
    this._refObject.planner_id = this._wrapperFields.planner_id.refObject;
    if (own) {
      this._wrapperFields.allowed_touch_objects.fill([]);
    } else {
      this._wrapperFields.allowed_touch_objects.fill(this._allowed_touch_objectsArray);
    }
    this._wrapperFields.allowed_touch_objects.freeze(own);
    this._refObject.allowed_touch_objects = this._wrapperFields.allowed_touch_objects.refObject;
    this._wrapperFields.planning_options.freeze(own);
    this._refObject.planning_options = this._wrapperFields.planning_options.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.group_name.data = refObject.group_name.data;
    this._wrapperFields.attached_object_name.data = refObject.attached_object_name.data;
    this._wrapperFields.place_locations.copyRefObject(refObject.place_locations);
    this._refObject.place_eef = refObject.place_eef;
    this._wrapperFields.support_surface_name.data = refObject.support_surface_name.data;
    this._refObject.allow_gripper_support_collision = refObject.allow_gripper_support_collision;
    this._wrapperFields.path_constraints.copyRefObject(refObject.path_constraints);
    this._wrapperFields.planner_id.data = refObject.planner_id.data;
    refObject.allowed_touch_objects.data.length = refObject.allowed_touch_objects.size;
    for (let index = 0; index < refObject.allowed_touch_objects.size; index++) {
      this._allowed_touch_objectsArray[index] = refObject.allowed_touch_objects.data[index].data;
    }
    this._refObject.allowed_planning_time = refObject.allowed_planning_time;
    this._wrapperFields.planning_options.copyRefObject(refObject.planning_options);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.group_name);
    StringWrapper.freeStruct(refObject.attached_object_name);
    if (refObject.place_locations.size != 0) {
      PlaceLocationWrapper.ArrayType.freeArray(refObject.place_locations);
      if (!PlaceLocationWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.place_locations, PlaceLocationWrapper.refObjectArrayType, 'data');
      }
    }
    StringWrapper.freeStruct(refObject.support_surface_name);
    ConstraintsWrapper.freeStruct(refObject.path_constraints);
    StringWrapper.freeStruct(refObject.planner_id);
    if (refObject.allowed_touch_objects.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.allowed_touch_objects);
      if (!StringWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.allowed_touch_objects, StringWrapper.refObjectArrayType, 'data');
      }
    }
    PlanningOptionsWrapper.freeStruct(refObject.planning_options);
  }
  static destroyRawROS(msg) {
    Place_GoalWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'action', interfaceName: 'Place_Goal'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get group_name() {
    return this._wrapperFields.group_name.data;
  }
  set group_name(value) {
    this._wrapperFields.group_name.data = value;
  }
  get attached_object_name() {
    return this._wrapperFields.attached_object_name.data;
  }
  set attached_object_name(value) {
    this._wrapperFields.attached_object_name.data = value;
  }
  get place_locations() {
    return this._wrapperFields.place_locations;
  }
  set place_locations(value) {
    this._wrapperFields.place_locations.fill(value);
  }
  get place_eef() {
    return this._refObject.place_eef;
  }
  set place_eef(value) {
    this._refObject.place_eef = value;
  }
  get support_surface_name() {
    return this._wrapperFields.support_surface_name.data;
  }
  set support_surface_name(value) {
    this._wrapperFields.support_surface_name.data = value;
  }
  get allow_gripper_support_collision() {
    return this._refObject.allow_gripper_support_collision;
  }
  set allow_gripper_support_collision(value) {
    this._refObject.allow_gripper_support_collision = value;
  }
  get path_constraints() {
    return this._wrapperFields.path_constraints;
  }
  set path_constraints(value) {
    if (value instanceof ConstraintsWrapper) {
      this._wrapperFields.path_constraints.copy(value);
    } else {
      this._wrapperFields.path_constraints.copy(new ConstraintsWrapper(value));
    }
  }
  get planner_id() {
    return this._wrapperFields.planner_id.data;
  }
  set planner_id(value) {
    this._wrapperFields.planner_id.data = value;
  }
  get allowed_touch_objects() {
    return this._allowed_touch_objectsArray;
  }
  set allowed_touch_objects(value) {
    this._allowed_touch_objectsArray = value;
  }
  get allowed_planning_time() {
    return this._refObject.allowed_planning_time;
  }
  set allowed_planning_time(value) {
    this._refObject.allowed_planning_time = value;
  }
  get planning_options() {
    return this._wrapperFields.planning_options;
  }
  set planning_options(value) {
    if (value instanceof PlanningOptionsWrapper) {
      this._wrapperFields.planning_options.copy(value);
    } else {
      this._wrapperFields.planning_options.copy(new PlanningOptionsWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new Place_GoalRefStruct(refObject.toObject());
    this._wrapperFields.group_name.copyRefObject(this._refObject.group_name);
    this._wrapperFields.attached_object_name.copyRefObject(this._refObject.attached_object_name);
    this._wrapperFields.place_locations.copyRefObject(this._refObject.place_locations);
    this._wrapperFields.support_surface_name.copyRefObject(this._refObject.support_surface_name);
    this._wrapperFields.path_constraints.copyRefObject(this._refObject.path_constraints);
    this._wrapperFields.planner_id.copyRefObject(this._refObject.planner_id);
    refObject.allowed_touch_objects.data.length = refObject.allowed_touch_objects.size;
    for (let index = 0; index < refObject.allowed_touch_objects.size; index++) {
      this._allowed_touch_objectsArray[index] = refObject.allowed_touch_objects.data[index].data;
    }
    this._wrapperFields.planning_options.copyRefObject(this._refObject.planning_options);
  }
  copy(other) {
    this._refObject = new Place_GoalRefStruct(other._refObject.toObject());
    this._wrapperFields.group_name.copy(other._wrapperFields.group_name);
    this._wrapperFields.attached_object_name.copy(other._wrapperFields.attached_object_name);
    this._wrapperFields.place_locations.copy(other._wrapperFields.place_locations);
    this._wrapperFields.support_surface_name.copy(other._wrapperFields.support_surface_name);
    this._wrapperFields.path_constraints.copy(other._wrapperFields.path_constraints);
    this._wrapperFields.planner_id.copy(other._wrapperFields.planner_id);
    this._allowed_touch_objectsArray = other._allowed_touch_objectsArray.slice();
    this._wrapperFields.planning_options.copy(other._wrapperFields.planning_options);
  }
  static get classType() {
    return Place_GoalWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"group_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"attached_object_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"place_locations","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"moveit_msgs","type":"PlaceLocation","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"place_eef","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"support_surface_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"allow_gripper_support_collision","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"path_constraints","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"Constraints","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"planner_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"allowed_touch_objects","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"allowed_planning_time","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"planning_options","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"PlanningOptions","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"Place_Goal","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Place_Goal"};
  }
  hasMember(name) {
    let memberNames = ["group_name","attached_object_name","place_locations","place_eef","support_surface_name","allow_gripper_support_collision","path_constraints","planner_id","allowed_touch_objects","allowed_planning_time","planning_options"];
    return memberNames.indexOf(name) !== -1;
  }
}
class Place_GoalArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof Place_GoalWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new Place_GoalWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Place_GoalArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Place_GoalArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new Place_GoalRefArray(size);
    this._refObject = new Place_GoalRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new Place_GoalWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof Place_GoalArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "Place_GoalArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      Place_GoalWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return Place_GoalWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return Place_GoalArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = Place_GoalWrapper;
