// rclnodejs/generated/moveit_msgs/moveit_msgs__msg__PlanningOptions.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PlanningSceneWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__PlanningScene.js');
const PlanningOptionsRefStruct = StructType({
  planning_scene_diff: PlanningSceneWrapper.refObjectType,
  plan_only: primitiveTypes.bool,
  look_around: primitiveTypes.bool,
  look_around_attempts: primitiveTypes.int32,
  max_safe_execution_cost: primitiveTypes.float64,
  replan: primitiveTypes.bool,
  replan_attempts: primitiveTypes.int32,
  replan_delay: primitiveTypes.float64,
});
const PlanningOptionsRefArray = ArrayType(PlanningOptionsRefStruct);
const PlanningOptionsRefStructArray = StructType({
  data: PlanningOptionsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class PlanningOptionsWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new PlanningOptionsRefStruct(other._refObject.toObject());
      this._wrapperFields.planning_scene_diff =  new PlanningSceneWrapper(other._wrapperFields.planning_scene_diff);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new PlanningOptionsRefStruct();
    this._wrapperFields.planning_scene_diff = new PlanningSceneWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new PlanningOptionsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new PlanningOptionsArrayWrapper();
  }
  static get ArrayType() {
    return PlanningOptionsArrayWrapper;
  }
  static get refObjectArrayType() {
    return PlanningOptionsRefStructArray;
  }
  static get refObjectType() {
    return PlanningOptionsRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.planning_scene_diff.freeze(own);
    this._refObject.planning_scene_diff = this._wrapperFields.planning_scene_diff.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.planning_scene_diff.copyRefObject(refObject.planning_scene_diff);
    this._refObject.plan_only = refObject.plan_only;
    this._refObject.look_around = refObject.look_around;
    this._refObject.look_around_attempts = refObject.look_around_attempts;
    this._refObject.max_safe_execution_cost = refObject.max_safe_execution_cost;
    this._refObject.replan = refObject.replan;
    this._refObject.replan_attempts = refObject.replan_attempts;
    this._refObject.replan_delay = refObject.replan_delay;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PlanningSceneWrapper.freeStruct(refObject.planning_scene_diff);
  }
  static destroyRawROS(msg) {
    PlanningOptionsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'msg', interfaceName: 'PlanningOptions'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get planning_scene_diff() {
    return this._wrapperFields.planning_scene_diff;
  }
  set planning_scene_diff(value) {
    if (value instanceof PlanningSceneWrapper) {
      this._wrapperFields.planning_scene_diff.copy(value);
    } else {
      this._wrapperFields.planning_scene_diff.copy(new PlanningSceneWrapper(value));
    }
  }
  get plan_only() {
    return this._refObject.plan_only;
  }
  set plan_only(value) {
    this._refObject.plan_only = value;
  }
  get look_around() {
    return this._refObject.look_around;
  }
  set look_around(value) {
    this._refObject.look_around = value;
  }
  get look_around_attempts() {
    return this._refObject.look_around_attempts;
  }
  set look_around_attempts(value) {
    this._refObject.look_around_attempts = value;
  }
  get max_safe_execution_cost() {
    return this._refObject.max_safe_execution_cost;
  }
  set max_safe_execution_cost(value) {
    this._refObject.max_safe_execution_cost = value;
  }
  get replan() {
    return this._refObject.replan;
  }
  set replan(value) {
    this._refObject.replan = value;
  }
  get replan_attempts() {
    return this._refObject.replan_attempts;
  }
  set replan_attempts(value) {
    this._refObject.replan_attempts = value;
  }
  get replan_delay() {
    return this._refObject.replan_delay;
  }
  set replan_delay(value) {
    this._refObject.replan_delay = value;
  }
  copyRefObject(refObject) {
    this._refObject = new PlanningOptionsRefStruct(refObject.toObject());
    this._wrapperFields.planning_scene_diff.copyRefObject(this._refObject.planning_scene_diff);
  }
  copy(other) {
    this._refObject = new PlanningOptionsRefStruct(other._refObject.toObject());
    this._wrapperFields.planning_scene_diff.copy(other._wrapperFields.planning_scene_diff);
  }
  static get classType() {
    return PlanningOptionsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"planning_scene_diff","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"PlanningScene","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"plan_only","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"look_around","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"look_around_attempts","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"max_safe_execution_cost","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"replan","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"replan_attempts","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"replan_delay","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"PlanningOptions","stringUpperBound":null,"isPrimitiveType":false},"msgName":"PlanningOptions"};
  }
  hasMember(name) {
    let memberNames = ["planning_scene_diff","plan_only","look_around","look_around_attempts","max_safe_execution_cost","replan","replan_attempts","replan_delay"];
    return memberNames.indexOf(name) !== -1;
  }
}
class PlanningOptionsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof PlanningOptionsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new PlanningOptionsWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PlanningOptionsArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PlanningOptionsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new PlanningOptionsRefArray(size);
    this._refObject = new PlanningOptionsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new PlanningOptionsWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof PlanningOptionsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "PlanningOptionsArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      PlanningOptionsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return PlanningOptionsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return PlanningOptionsArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = PlanningOptionsWrapper;
