// rclnodejs/generated/moveit_msgs/moveit_msgs__msg__JointConstraint.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const JointConstraintRefStruct = StructType({
  joint_name: primitiveTypes.string,
  position: primitiveTypes.float64,
  tolerance_above: primitiveTypes.float64,
  tolerance_below: primitiveTypes.float64,
  weight: primitiveTypes.float64,
});
const JointConstraintRefArray = ArrayType(JointConstraintRefStruct);
const JointConstraintRefStructArray = StructType({
  data: JointConstraintRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class JointConstraintWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new JointConstraintRefStruct(other._refObject.toObject());
      this._wrapperFields.joint_name =  new StringWrapper(other._wrapperFields.joint_name);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new JointConstraintRefStruct();
    this._wrapperFields.joint_name = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new JointConstraintWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new JointConstraintArrayWrapper();
  }
  static get ArrayType() {
    return JointConstraintArrayWrapper;
  }
  static get refObjectArrayType() {
    return JointConstraintRefStructArray;
  }
  static get refObjectType() {
    return JointConstraintRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.joint_name.freeze(own);
    }
    this._refObject.joint_name = this._wrapperFields.joint_name.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.joint_name.data = refObject.joint_name.data;
    this._refObject.position = refObject.position;
    this._refObject.tolerance_above = refObject.tolerance_above;
    this._refObject.tolerance_below = refObject.tolerance_below;
    this._refObject.weight = refObject.weight;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.joint_name);
  }
  static destroyRawROS(msg) {
    JointConstraintWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'msg', interfaceName: 'JointConstraint'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get joint_name() {
    return this._wrapperFields.joint_name.data;
  }
  set joint_name(value) {
    this._wrapperFields.joint_name.data = value;
  }
  get position() {
    return this._refObject.position;
  }
  set position(value) {
    this._refObject.position = value;
  }
  get tolerance_above() {
    return this._refObject.tolerance_above;
  }
  set tolerance_above(value) {
    this._refObject.tolerance_above = value;
  }
  get tolerance_below() {
    return this._refObject.tolerance_below;
  }
  set tolerance_below(value) {
    this._refObject.tolerance_below = value;
  }
  get weight() {
    return this._refObject.weight;
  }
  set weight(value) {
    this._refObject.weight = value;
  }
  copyRefObject(refObject) {
    this._refObject = new JointConstraintRefStruct(refObject.toObject());
    this._wrapperFields.joint_name.copyRefObject(this._refObject.joint_name);
  }
  copy(other) {
    this._refObject = new JointConstraintRefStruct(other._refObject.toObject());
    this._wrapperFields.joint_name.copy(other._wrapperFields.joint_name);
  }
  static get classType() {
    return JointConstraintWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"joint_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"position","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"tolerance_above","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"tolerance_below","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"weight","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"JointConstraint","stringUpperBound":null,"isPrimitiveType":false},"msgName":"JointConstraint"};
  }
  hasMember(name) {
    let memberNames = ["joint_name","position","tolerance_above","tolerance_below","weight"];
    return memberNames.indexOf(name) !== -1;
  }
}
class JointConstraintArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof JointConstraintWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new JointConstraintWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to JointConstraintArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to JointConstraintArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new JointConstraintRefArray(size);
    this._refObject = new JointConstraintRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new JointConstraintWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof JointConstraintArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "JointConstraintArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      JointConstraintWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return JointConstraintWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return JointConstraintArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = JointConstraintWrapper;
