// rclnodejs/generated/moveit_msgs/moveit_msgs__msg__RobotState.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const JointStateWrapper = require('../../generated/sensor_msgs/sensor_msgs__msg__JointState.js');
const MultiDOFJointStateWrapper = require('../../generated/sensor_msgs/sensor_msgs__msg__MultiDOFJointState.js');
const AttachedCollisionObjectWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__AttachedCollisionObject.js');
const RobotStateRefStruct = StructType({
  joint_state: JointStateWrapper.refObjectType,
  multi_dof_joint_state: MultiDOFJointStateWrapper.refObjectType,
  attached_collision_objects: AttachedCollisionObjectWrapper.refObjectArrayType,
  is_diff: primitiveTypes.bool,
});
const RobotStateRefArray = ArrayType(RobotStateRefStruct);
const RobotStateRefStructArray = StructType({
  data: RobotStateRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class RobotStateWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new RobotStateRefStruct(other._refObject.toObject());
      this._wrapperFields.joint_state =  new JointStateWrapper(other._wrapperFields.joint_state);
      this._wrapperFields.multi_dof_joint_state =  new MultiDOFJointStateWrapper(other._wrapperFields.multi_dof_joint_state);
      this._wrapperFields.attached_collision_objects = AttachedCollisionObjectWrapper.createArray();
      this._wrapperFields.attached_collision_objects.copy(other._wrapperFields.attached_collision_objects);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new RobotStateRefStruct();
    this._wrapperFields.joint_state = new JointStateWrapper();
    this._wrapperFields.multi_dof_joint_state = new MultiDOFJointStateWrapper();
    this._wrapperFields.attached_collision_objects = AttachedCollisionObjectWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new RobotStateWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new RobotStateArrayWrapper();
  }
  static get ArrayType() {
    return RobotStateArrayWrapper;
  }
  static get refObjectArrayType() {
    return RobotStateRefStructArray;
  }
  static get refObjectType() {
    return RobotStateRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.joint_state.freeze(own);
    this._refObject.joint_state = this._wrapperFields.joint_state.refObject;
    this._wrapperFields.multi_dof_joint_state.freeze(own);
    this._refObject.multi_dof_joint_state = this._wrapperFields.multi_dof_joint_state.refObject;
    if (own) {
      this._wrapperFields.attached_collision_objects.fill([]);
    }
    this._wrapperFields.attached_collision_objects.freeze(own);
    this._refObject.attached_collision_objects = this._wrapperFields.attached_collision_objects.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.joint_state.copyRefObject(refObject.joint_state);
    this._wrapperFields.multi_dof_joint_state.copyRefObject(refObject.multi_dof_joint_state);
    this._wrapperFields.attached_collision_objects.copyRefObject(refObject.attached_collision_objects);
    this._refObject.is_diff = refObject.is_diff;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    JointStateWrapper.freeStruct(refObject.joint_state);
    MultiDOFJointStateWrapper.freeStruct(refObject.multi_dof_joint_state);
    if (refObject.attached_collision_objects.size != 0) {
      AttachedCollisionObjectWrapper.ArrayType.freeArray(refObject.attached_collision_objects);
      if (!AttachedCollisionObjectWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.attached_collision_objects, AttachedCollisionObjectWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    RobotStateWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'msg', interfaceName: 'RobotState'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get joint_state() {
    return this._wrapperFields.joint_state;
  }
  set joint_state(value) {
    if (value instanceof JointStateWrapper) {
      this._wrapperFields.joint_state.copy(value);
    } else {
      this._wrapperFields.joint_state.copy(new JointStateWrapper(value));
    }
  }
  get multi_dof_joint_state() {
    return this._wrapperFields.multi_dof_joint_state;
  }
  set multi_dof_joint_state(value) {
    if (value instanceof MultiDOFJointStateWrapper) {
      this._wrapperFields.multi_dof_joint_state.copy(value);
    } else {
      this._wrapperFields.multi_dof_joint_state.copy(new MultiDOFJointStateWrapper(value));
    }
  }
  get attached_collision_objects() {
    return this._wrapperFields.attached_collision_objects;
  }
  set attached_collision_objects(value) {
    this._wrapperFields.attached_collision_objects.fill(value);
  }
  get is_diff() {
    return this._refObject.is_diff;
  }
  set is_diff(value) {
    this._refObject.is_diff = value;
  }
  copyRefObject(refObject) {
    this._refObject = new RobotStateRefStruct(refObject.toObject());
    this._wrapperFields.joint_state.copyRefObject(this._refObject.joint_state);
    this._wrapperFields.multi_dof_joint_state.copyRefObject(this._refObject.multi_dof_joint_state);
    this._wrapperFields.attached_collision_objects.copyRefObject(this._refObject.attached_collision_objects);
  }
  copy(other) {
    this._refObject = new RobotStateRefStruct(other._refObject.toObject());
    this._wrapperFields.joint_state.copy(other._wrapperFields.joint_state);
    this._wrapperFields.multi_dof_joint_state.copy(other._wrapperFields.multi_dof_joint_state);
    this._wrapperFields.attached_collision_objects.copy(other._wrapperFields.attached_collision_objects);
  }
  static get classType() {
    return RobotStateWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"joint_state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"sensor_msgs","type":"JointState","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"multi_dof_joint_state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"sensor_msgs","type":"MultiDOFJointState","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"attached_collision_objects","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"moveit_msgs","type":"AttachedCollisionObject","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"is_diff","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"RobotState","stringUpperBound":null,"isPrimitiveType":false},"msgName":"RobotState"};
  }
  hasMember(name) {
    let memberNames = ["joint_state","multi_dof_joint_state","attached_collision_objects","is_diff"];
    return memberNames.indexOf(name) !== -1;
  }
}
class RobotStateArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof RobotStateWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new RobotStateWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RobotStateArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RobotStateArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new RobotStateRefArray(size);
    this._refObject = new RobotStateRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new RobotStateWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof RobotStateArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "RobotStateArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      RobotStateWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return RobotStateWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return RobotStateArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = RobotStateWrapper;
