// rclnodejs/generated/moveit_msgs/moveit_msgs__msg__OrientedBoundingBox.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PoseWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Pose.js');
const Point32Wrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Point32.js');
const OrientedBoundingBoxRefStruct = StructType({
  pose: PoseWrapper.refObjectType,
  extents: Point32Wrapper.refObjectType,
});
const OrientedBoundingBoxRefArray = ArrayType(OrientedBoundingBoxRefStruct);
const OrientedBoundingBoxRefStructArray = StructType({
  data: OrientedBoundingBoxRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class OrientedBoundingBoxWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new OrientedBoundingBoxRefStruct(other._refObject.toObject());
      this._wrapperFields.pose =  new PoseWrapper(other._wrapperFields.pose);
      this._wrapperFields.extents =  new Point32Wrapper(other._wrapperFields.extents);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new OrientedBoundingBoxRefStruct();
    this._wrapperFields.pose = new PoseWrapper();
    this._wrapperFields.extents = new Point32Wrapper();
  }
  static createFromRefObject(refObject) {
    let self = new OrientedBoundingBoxWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new OrientedBoundingBoxArrayWrapper();
  }
  static get ArrayType() {
    return OrientedBoundingBoxArrayWrapper;
  }
  static get refObjectArrayType() {
    return OrientedBoundingBoxRefStructArray;
  }
  static get refObjectType() {
    return OrientedBoundingBoxRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.pose.freeze(own);
    this._refObject.pose = this._wrapperFields.pose.refObject;
    this._wrapperFields.extents.freeze(own);
    this._refObject.extents = this._wrapperFields.extents.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.pose.copyRefObject(refObject.pose);
    this._wrapperFields.extents.copyRefObject(refObject.extents);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PoseWrapper.freeStruct(refObject.pose);
    Point32Wrapper.freeStruct(refObject.extents);
  }
  static destroyRawROS(msg) {
    OrientedBoundingBoxWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'msg', interfaceName: 'OrientedBoundingBox'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get pose() {
    return this._wrapperFields.pose;
  }
  set pose(value) {
    if (value instanceof PoseWrapper) {
      this._wrapperFields.pose.copy(value);
    } else {
      this._wrapperFields.pose.copy(new PoseWrapper(value));
    }
  }
  get extents() {
    return this._wrapperFields.extents;
  }
  set extents(value) {
    if (value instanceof Point32Wrapper) {
      this._wrapperFields.extents.copy(value);
    } else {
      this._wrapperFields.extents.copy(new Point32Wrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new OrientedBoundingBoxRefStruct(refObject.toObject());
    this._wrapperFields.pose.copyRefObject(this._refObject.pose);
    this._wrapperFields.extents.copyRefObject(this._refObject.extents);
  }
  copy(other) {
    this._refObject = new OrientedBoundingBoxRefStruct(other._refObject.toObject());
    this._wrapperFields.pose.copy(other._wrapperFields.pose);
    this._wrapperFields.extents.copy(other._wrapperFields.extents);
  }
  static get classType() {
    return OrientedBoundingBoxWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"pose","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Pose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"extents","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Point32","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"OrientedBoundingBox","stringUpperBound":null,"isPrimitiveType":false},"msgName":"OrientedBoundingBox"};
  }
  hasMember(name) {
    let memberNames = ["pose","extents"];
    return memberNames.indexOf(name) !== -1;
  }
}
class OrientedBoundingBoxArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof OrientedBoundingBoxWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new OrientedBoundingBoxWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to OrientedBoundingBoxArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to OrientedBoundingBoxArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new OrientedBoundingBoxRefArray(size);
    this._refObject = new OrientedBoundingBoxRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new OrientedBoundingBoxWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof OrientedBoundingBoxArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "OrientedBoundingBoxArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      OrientedBoundingBoxWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return OrientedBoundingBoxWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return OrientedBoundingBoxArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = OrientedBoundingBoxWrapper;
