// rclnodejs/generated/moveit_msgs/moveit_msgs__srv__GetPositionFK_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PoseStampedWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__PoseStamped.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const MoveItErrorCodesWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__MoveItErrorCodes.js');
const GetPositionFK_ResponseRefStruct = StructType({
  pose_stamped: PoseStampedWrapper.refObjectArrayType,
  fk_link_names: StringWrapper.refObjectArrayType,
  error_code: MoveItErrorCodesWrapper.refObjectType,
});
const GetPositionFK_ResponseRefArray = ArrayType(GetPositionFK_ResponseRefStruct);
const GetPositionFK_ResponseRefStructArray = StructType({
  data: GetPositionFK_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class GetPositionFK_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetPositionFK_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.pose_stamped = PoseStampedWrapper.createArray();
      this._wrapperFields.pose_stamped.copy(other._wrapperFields.pose_stamped);
      this._wrapperFields.fk_link_names = StringWrapper.createArray();
      this._wrapperFields.fk_link_names.copy(other._wrapperFields.fk_link_names);
      this.fk_link_names = other.fk_link_names;
      this._wrapperFields.error_code =  new MoveItErrorCodesWrapper(other._wrapperFields.error_code);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new GetPositionFK_ResponseRefStruct();
    this._wrapperFields.pose_stamped = PoseStampedWrapper.createArray();
    this._wrapperFields.fk_link_names = StringWrapper.createArray();
    this._fk_link_namesArray = [];
    this._wrapperFields.error_code = new MoveItErrorCodesWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GetPositionFK_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetPositionFK_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return GetPositionFK_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetPositionFK_ResponseRefStructArray;
  }
  static get refObjectType() {
    return GetPositionFK_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.pose_stamped.fill([]);
    }
    this._wrapperFields.pose_stamped.freeze(own);
    this._refObject.pose_stamped = this._wrapperFields.pose_stamped.refObject;
    if (own) {
      this._wrapperFields.fk_link_names.fill([]);
    } else {
      this._wrapperFields.fk_link_names.fill(this._fk_link_namesArray);
    }
    this._wrapperFields.fk_link_names.freeze(own);
    this._refObject.fk_link_names = this._wrapperFields.fk_link_names.refObject;
    this._wrapperFields.error_code.freeze(own);
    this._refObject.error_code = this._wrapperFields.error_code.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.pose_stamped.copyRefObject(refObject.pose_stamped);
    refObject.fk_link_names.data.length = refObject.fk_link_names.size;
    for (let index = 0; index < refObject.fk_link_names.size; index++) {
      this._fk_link_namesArray[index] = refObject.fk_link_names.data[index].data;
    }
    this._wrapperFields.error_code.copyRefObject(refObject.error_code);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.pose_stamped.size != 0) {
      PoseStampedWrapper.ArrayType.freeArray(refObject.pose_stamped);
      if (!PoseStampedWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.pose_stamped, PoseStampedWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.fk_link_names.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.fk_link_names);
      if (!StringWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.fk_link_names, StringWrapper.refObjectArrayType, 'data');
      }
    }
    MoveItErrorCodesWrapper.freeStruct(refObject.error_code);
  }
  static destroyRawROS(msg) {
    GetPositionFK_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'srv', interfaceName: 'GetPositionFK_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get pose_stamped() {
    return this._wrapperFields.pose_stamped;
  }
  set pose_stamped(value) {
    this._wrapperFields.pose_stamped.fill(value);
  }
  get fk_link_names() {
    return this._fk_link_namesArray;
  }
  set fk_link_names(value) {
    this._fk_link_namesArray = value;
  }
  get error_code() {
    return this._wrapperFields.error_code;
  }
  set error_code(value) {
    if (value instanceof MoveItErrorCodesWrapper) {
      this._wrapperFields.error_code.copy(value);
    } else {
      this._wrapperFields.error_code.copy(new MoveItErrorCodesWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new GetPositionFK_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.pose_stamped.copyRefObject(this._refObject.pose_stamped);
    refObject.fk_link_names.data.length = refObject.fk_link_names.size;
    for (let index = 0; index < refObject.fk_link_names.size; index++) {
      this._fk_link_namesArray[index] = refObject.fk_link_names.data[index].data;
    }
    this._wrapperFields.error_code.copyRefObject(this._refObject.error_code);
  }
  copy(other) {
    this._refObject = new GetPositionFK_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.pose_stamped.copy(other._wrapperFields.pose_stamped);
    this._fk_link_namesArray = other._fk_link_namesArray.slice();
    this._wrapperFields.error_code.copy(other._wrapperFields.error_code);
  }
  static get classType() {
    return GetPositionFK_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"pose_stamped","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"geometry_msgs","type":"PoseStamped","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"fk_link_names","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"error_code","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"MoveItErrorCodes","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"GetPositionFK_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetPositionFK_Response"};
  }
  hasMember(name) {
    let memberNames = ["pose_stamped","fk_link_names","error_code"];
    return memberNames.indexOf(name) !== -1;
  }
}
class GetPositionFK_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetPositionFK_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetPositionFK_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetPositionFK_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetPositionFK_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new GetPositionFK_ResponseRefArray(size);
    this._refObject = new GetPositionFK_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetPositionFK_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetPositionFK_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetPositionFK_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetPositionFK_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetPositionFK_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetPositionFK_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = GetPositionFK_ResponseWrapper;
