// rclnodejs/generated/moveit_msgs/moveit_msgs__srv__ListRobotStatesInWarehouse_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const ListRobotStatesInWarehouse_ResponseRefStruct = StructType({
  states: StringWrapper.refObjectArrayType,
});
const ListRobotStatesInWarehouse_ResponseRefArray = ArrayType(ListRobotStatesInWarehouse_ResponseRefStruct);
const ListRobotStatesInWarehouse_ResponseRefStructArray = StructType({
  data: ListRobotStatesInWarehouse_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class ListRobotStatesInWarehouse_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new ListRobotStatesInWarehouse_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.states = StringWrapper.createArray();
      this._wrapperFields.states.copy(other._wrapperFields.states);
      this.states = other.states;
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new ListRobotStatesInWarehouse_ResponseRefStruct();
    this._wrapperFields.states = StringWrapper.createArray();
    this._statesArray = [];
  }
  static createFromRefObject(refObject) {
    let self = new ListRobotStatesInWarehouse_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ListRobotStatesInWarehouse_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return ListRobotStatesInWarehouse_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return ListRobotStatesInWarehouse_ResponseRefStructArray;
  }
  static get refObjectType() {
    return ListRobotStatesInWarehouse_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.states.fill([]);
    } else {
      this._wrapperFields.states.fill(this._statesArray);
    }
    this._wrapperFields.states.freeze(own);
    this._refObject.states = this._wrapperFields.states.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    refObject.states.data.length = refObject.states.size;
    for (let index = 0; index < refObject.states.size; index++) {
      this._statesArray[index] = refObject.states.data[index].data;
    }
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.states.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.states);
      if (!StringWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.states, StringWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    ListRobotStatesInWarehouse_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'srv', interfaceName: 'ListRobotStatesInWarehouse_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get states() {
    return this._statesArray;
  }
  set states(value) {
    this._statesArray = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ListRobotStatesInWarehouse_ResponseRefStruct(refObject.toObject());
    refObject.states.data.length = refObject.states.size;
    for (let index = 0; index < refObject.states.size; index++) {
      this._statesArray[index] = refObject.states.data[index].data;
    }
  }
  copy(other) {
    this._refObject = new ListRobotStatesInWarehouse_ResponseRefStruct(other._refObject.toObject());
    this._statesArray = other._statesArray.slice();
  }
  static get classType() {
    return ListRobotStatesInWarehouse_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"states","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"ListRobotStatesInWarehouse_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ListRobotStatesInWarehouse_Response"};
  }
  hasMember(name) {
    let memberNames = ["states"];
    return memberNames.indexOf(name) !== -1;
  }
}
class ListRobotStatesInWarehouse_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ListRobotStatesInWarehouse_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ListRobotStatesInWarehouse_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ListRobotStatesInWarehouse_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ListRobotStatesInWarehouse_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new ListRobotStatesInWarehouse_ResponseRefArray(size);
    this._refObject = new ListRobotStatesInWarehouse_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ListRobotStatesInWarehouse_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ListRobotStatesInWarehouse_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ListRobotStatesInWarehouse_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ListRobotStatesInWarehouse_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ListRobotStatesInWarehouse_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ListRobotStatesInWarehouse_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = ListRobotStatesInWarehouse_ResponseWrapper;
