// rclnodejs/generated/moveit_msgs/moveit_msgs__msg__AttachedCollisionObject.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const CollisionObjectWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__CollisionObject.js');
const JointTrajectoryWrapper = require('../../generated/trajectory_msgs/trajectory_msgs__msg__JointTrajectory.js');
const AttachedCollisionObjectRefStruct = StructType({
  link_name: primitiveTypes.string,
  object: CollisionObjectWrapper.refObjectType,
  touch_links: StringWrapper.refObjectArrayType,
  detach_posture: JointTrajectoryWrapper.refObjectType,
  weight: primitiveTypes.float64,
});
const AttachedCollisionObjectRefArray = ArrayType(AttachedCollisionObjectRefStruct);
const AttachedCollisionObjectRefStructArray = StructType({
  data: AttachedCollisionObjectRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class AttachedCollisionObjectWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new AttachedCollisionObjectRefStruct(other._refObject.toObject());
      this._wrapperFields.link_name =  new StringWrapper(other._wrapperFields.link_name);
      this._wrapperFields.object =  new CollisionObjectWrapper(other._wrapperFields.object);
      this._wrapperFields.touch_links = StringWrapper.createArray();
      this._wrapperFields.touch_links.copy(other._wrapperFields.touch_links);
      this.touch_links = other.touch_links;
      this._wrapperFields.detach_posture =  new JointTrajectoryWrapper(other._wrapperFields.detach_posture);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new AttachedCollisionObjectRefStruct();
    this._wrapperFields.link_name = new StringWrapper();
    this._wrapperFields.object = new CollisionObjectWrapper();
    this._wrapperFields.touch_links = StringWrapper.createArray();
    this._touch_linksArray = [];
    this._wrapperFields.detach_posture = new JointTrajectoryWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new AttachedCollisionObjectWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new AttachedCollisionObjectArrayWrapper();
  }
  static get ArrayType() {
    return AttachedCollisionObjectArrayWrapper;
  }
  static get refObjectArrayType() {
    return AttachedCollisionObjectRefStructArray;
  }
  static get refObjectType() {
    return AttachedCollisionObjectRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.link_name.freeze(own);
    }
    this._refObject.link_name = this._wrapperFields.link_name.refObject;
    this._wrapperFields.object.freeze(own);
    this._refObject.object = this._wrapperFields.object.refObject;
    if (own) {
      this._wrapperFields.touch_links.fill([]);
    } else {
      this._wrapperFields.touch_links.fill(this._touch_linksArray);
    }
    this._wrapperFields.touch_links.freeze(own);
    this._refObject.touch_links = this._wrapperFields.touch_links.refObject;
    this._wrapperFields.detach_posture.freeze(own);
    this._refObject.detach_posture = this._wrapperFields.detach_posture.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.link_name.data = refObject.link_name.data;
    this._wrapperFields.object.copyRefObject(refObject.object);
    refObject.touch_links.data.length = refObject.touch_links.size;
    for (let index = 0; index < refObject.touch_links.size; index++) {
      this._touch_linksArray[index] = refObject.touch_links.data[index].data;
    }
    this._wrapperFields.detach_posture.copyRefObject(refObject.detach_posture);
    this._refObject.weight = refObject.weight;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.link_name);
    CollisionObjectWrapper.freeStruct(refObject.object);
    if (refObject.touch_links.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.touch_links);
      if (!StringWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.touch_links, StringWrapper.refObjectArrayType, 'data');
      }
    }
    JointTrajectoryWrapper.freeStruct(refObject.detach_posture);
  }
  static destroyRawROS(msg) {
    AttachedCollisionObjectWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'msg', interfaceName: 'AttachedCollisionObject'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get link_name() {
    return this._wrapperFields.link_name.data;
  }
  set link_name(value) {
    this._wrapperFields.link_name.data = value;
  }
  get object() {
    return this._wrapperFields.object;
  }
  set object(value) {
    if (value instanceof CollisionObjectWrapper) {
      this._wrapperFields.object.copy(value);
    } else {
      this._wrapperFields.object.copy(new CollisionObjectWrapper(value));
    }
  }
  get touch_links() {
    return this._touch_linksArray;
  }
  set touch_links(value) {
    this._touch_linksArray = value;
  }
  get detach_posture() {
    return this._wrapperFields.detach_posture;
  }
  set detach_posture(value) {
    if (value instanceof JointTrajectoryWrapper) {
      this._wrapperFields.detach_posture.copy(value);
    } else {
      this._wrapperFields.detach_posture.copy(new JointTrajectoryWrapper(value));
    }
  }
  get weight() {
    return this._refObject.weight;
  }
  set weight(value) {
    this._refObject.weight = value;
  }
  copyRefObject(refObject) {
    this._refObject = new AttachedCollisionObjectRefStruct(refObject.toObject());
    this._wrapperFields.link_name.copyRefObject(this._refObject.link_name);
    this._wrapperFields.object.copyRefObject(this._refObject.object);
    refObject.touch_links.data.length = refObject.touch_links.size;
    for (let index = 0; index < refObject.touch_links.size; index++) {
      this._touch_linksArray[index] = refObject.touch_links.data[index].data;
    }
    this._wrapperFields.detach_posture.copyRefObject(this._refObject.detach_posture);
  }
  copy(other) {
    this._refObject = new AttachedCollisionObjectRefStruct(other._refObject.toObject());
    this._wrapperFields.link_name.copy(other._wrapperFields.link_name);
    this._wrapperFields.object.copy(other._wrapperFields.object);
    this._touch_linksArray = other._touch_linksArray.slice();
    this._wrapperFields.detach_posture.copy(other._wrapperFields.detach_posture);
  }
  static get classType() {
    return AttachedCollisionObjectWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"link_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"object","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"CollisionObject","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"touch_links","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"detach_posture","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"trajectory_msgs","type":"JointTrajectory","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"weight","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"AttachedCollisionObject","stringUpperBound":null,"isPrimitiveType":false},"msgName":"AttachedCollisionObject"};
  }
  hasMember(name) {
    let memberNames = ["link_name","object","touch_links","detach_posture","weight"];
    return memberNames.indexOf(name) !== -1;
  }
}
class AttachedCollisionObjectArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof AttachedCollisionObjectWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new AttachedCollisionObjectWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AttachedCollisionObjectArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AttachedCollisionObjectArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new AttachedCollisionObjectRefArray(size);
    this._refObject = new AttachedCollisionObjectRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new AttachedCollisionObjectWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof AttachedCollisionObjectArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "AttachedCollisionObjectArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      AttachedCollisionObjectWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return AttachedCollisionObjectWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return AttachedCollisionObjectArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = AttachedCollisionObjectWrapper;
