// rclnodejs/generated/moveit_msgs/moveit_msgs__srv__GetStateValidity_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const ContactInformationWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__ContactInformation.js');
const CostSourceWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__CostSource.js');
const ConstraintEvalResultWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__ConstraintEvalResult.js');
const GetStateValidity_ResponseRefStruct = StructType({
  valid: primitiveTypes.bool,
  contacts: ContactInformationWrapper.refObjectArrayType,
  cost_sources: CostSourceWrapper.refObjectArrayType,
  constraint_result: ConstraintEvalResultWrapper.refObjectArrayType,
});
const GetStateValidity_ResponseRefArray = ArrayType(GetStateValidity_ResponseRefStruct);
const GetStateValidity_ResponseRefStructArray = StructType({
  data: GetStateValidity_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class GetStateValidity_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetStateValidity_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.contacts = ContactInformationWrapper.createArray();
      this._wrapperFields.contacts.copy(other._wrapperFields.contacts);
      this._wrapperFields.cost_sources = CostSourceWrapper.createArray();
      this._wrapperFields.cost_sources.copy(other._wrapperFields.cost_sources);
      this._wrapperFields.constraint_result = ConstraintEvalResultWrapper.createArray();
      this._wrapperFields.constraint_result.copy(other._wrapperFields.constraint_result);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new GetStateValidity_ResponseRefStruct();
    this._wrapperFields.contacts = ContactInformationWrapper.createArray();
    this._wrapperFields.cost_sources = CostSourceWrapper.createArray();
    this._wrapperFields.constraint_result = ConstraintEvalResultWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new GetStateValidity_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetStateValidity_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return GetStateValidity_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetStateValidity_ResponseRefStructArray;
  }
  static get refObjectType() {
    return GetStateValidity_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.contacts.fill([]);
    }
    this._wrapperFields.contacts.freeze(own);
    this._refObject.contacts = this._wrapperFields.contacts.refObject;
    if (own) {
      this._wrapperFields.cost_sources.fill([]);
    }
    this._wrapperFields.cost_sources.freeze(own);
    this._refObject.cost_sources = this._wrapperFields.cost_sources.refObject;
    if (own) {
      this._wrapperFields.constraint_result.fill([]);
    }
    this._wrapperFields.constraint_result.freeze(own);
    this._refObject.constraint_result = this._wrapperFields.constraint_result.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.valid = refObject.valid;
    this._wrapperFields.contacts.copyRefObject(refObject.contacts);
    this._wrapperFields.cost_sources.copyRefObject(refObject.cost_sources);
    this._wrapperFields.constraint_result.copyRefObject(refObject.constraint_result);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.contacts.size != 0) {
      ContactInformationWrapper.ArrayType.freeArray(refObject.contacts);
      if (!ContactInformationWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.contacts, ContactInformationWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.cost_sources.size != 0) {
      CostSourceWrapper.ArrayType.freeArray(refObject.cost_sources);
      if (!CostSourceWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.cost_sources, CostSourceWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.constraint_result.size != 0) {
      ConstraintEvalResultWrapper.ArrayType.freeArray(refObject.constraint_result);
      if (!ConstraintEvalResultWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.constraint_result, ConstraintEvalResultWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    GetStateValidity_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'srv', interfaceName: 'GetStateValidity_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get valid() {
    return this._refObject.valid;
  }
  set valid(value) {
    this._refObject.valid = value;
  }
  get contacts() {
    return this._wrapperFields.contacts;
  }
  set contacts(value) {
    this._wrapperFields.contacts.fill(value);
  }
  get cost_sources() {
    return this._wrapperFields.cost_sources;
  }
  set cost_sources(value) {
    this._wrapperFields.cost_sources.fill(value);
  }
  get constraint_result() {
    return this._wrapperFields.constraint_result;
  }
  set constraint_result(value) {
    this._wrapperFields.constraint_result.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new GetStateValidity_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.contacts.copyRefObject(this._refObject.contacts);
    this._wrapperFields.cost_sources.copyRefObject(this._refObject.cost_sources);
    this._wrapperFields.constraint_result.copyRefObject(this._refObject.constraint_result);
  }
  copy(other) {
    this._refObject = new GetStateValidity_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.contacts.copy(other._wrapperFields.contacts);
    this._wrapperFields.cost_sources.copy(other._wrapperFields.cost_sources);
    this._wrapperFields.constraint_result.copy(other._wrapperFields.constraint_result);
  }
  static get classType() {
    return GetStateValidity_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"valid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"contacts","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"moveit_msgs","type":"ContactInformation","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"cost_sources","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"moveit_msgs","type":"CostSource","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"constraint_result","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"moveit_msgs","type":"ConstraintEvalResult","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"GetStateValidity_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetStateValidity_Response"};
  }
  hasMember(name) {
    let memberNames = ["valid","contacts","cost_sources","constraint_result"];
    return memberNames.indexOf(name) !== -1;
  }
}
class GetStateValidity_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetStateValidity_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetStateValidity_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetStateValidity_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetStateValidity_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new GetStateValidity_ResponseRefArray(size);
    this._refObject = new GetStateValidity_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetStateValidity_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetStateValidity_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetStateValidity_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetStateValidity_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetStateValidity_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetStateValidity_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = GetStateValidity_ResponseWrapper;
