// rclnodejs/generated/moveit_msgs/moveit_msgs__msg__MotionSequenceItem.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const MotionPlanRequestWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__MotionPlanRequest.js');
const MotionSequenceItemRefStruct = StructType({
  req: MotionPlanRequestWrapper.refObjectType,
  blend_radius: primitiveTypes.float64,
});
const MotionSequenceItemRefArray = ArrayType(MotionSequenceItemRefStruct);
const MotionSequenceItemRefStructArray = StructType({
  data: MotionSequenceItemRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class MotionSequenceItemWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new MotionSequenceItemRefStruct(other._refObject.toObject());
      this._wrapperFields.req =  new MotionPlanRequestWrapper(other._wrapperFields.req);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new MotionSequenceItemRefStruct();
    this._wrapperFields.req = new MotionPlanRequestWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new MotionSequenceItemWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MotionSequenceItemArrayWrapper();
  }
  static get ArrayType() {
    return MotionSequenceItemArrayWrapper;
  }
  static get refObjectArrayType() {
    return MotionSequenceItemRefStructArray;
  }
  static get refObjectType() {
    return MotionSequenceItemRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.req.freeze(own);
    this._refObject.req = this._wrapperFields.req.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.req.copyRefObject(refObject.req);
    this._refObject.blend_radius = refObject.blend_radius;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    MotionPlanRequestWrapper.freeStruct(refObject.req);
  }
  static destroyRawROS(msg) {
    MotionSequenceItemWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'msg', interfaceName: 'MotionSequenceItem'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get req() {
    return this._wrapperFields.req;
  }
  set req(value) {
    if (value instanceof MotionPlanRequestWrapper) {
      this._wrapperFields.req.copy(value);
    } else {
      this._wrapperFields.req.copy(new MotionPlanRequestWrapper(value));
    }
  }
  get blend_radius() {
    return this._refObject.blend_radius;
  }
  set blend_radius(value) {
    this._refObject.blend_radius = value;
  }
  copyRefObject(refObject) {
    this._refObject = new MotionSequenceItemRefStruct(refObject.toObject());
    this._wrapperFields.req.copyRefObject(this._refObject.req);
  }
  copy(other) {
    this._refObject = new MotionSequenceItemRefStruct(other._refObject.toObject());
    this._wrapperFields.req.copy(other._wrapperFields.req);
  }
  static get classType() {
    return MotionSequenceItemWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"req","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"MotionPlanRequest","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"blend_radius","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"MotionSequenceItem","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MotionSequenceItem"};
  }
  hasMember(name) {
    let memberNames = ["req","blend_radius"];
    return memberNames.indexOf(name) !== -1;
  }
}
class MotionSequenceItemArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MotionSequenceItemWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MotionSequenceItemWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MotionSequenceItemArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MotionSequenceItemArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new MotionSequenceItemRefArray(size);
    this._refObject = new MotionSequenceItemRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MotionSequenceItemWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MotionSequenceItemArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MotionSequenceItemArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MotionSequenceItemWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MotionSequenceItemWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MotionSequenceItemArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = MotionSequenceItemWrapper;
