// rclnodejs/generated/moveit_msgs/moveit_msgs__srv__GetCartesianPath_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const RobotStateWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__RobotState.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const PoseWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Pose.js');
const ConstraintsWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__Constraints.js');
const GetCartesianPath_RequestRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  start_state: RobotStateWrapper.refObjectType,
  group_name: primitiveTypes.string,
  link_name: primitiveTypes.string,
  waypoints: PoseWrapper.refObjectArrayType,
  max_step: primitiveTypes.float64,
  jump_threshold: primitiveTypes.float64,
  prismatic_jump_threshold: primitiveTypes.float64,
  revolute_jump_threshold: primitiveTypes.float64,
  avoid_collisions: primitiveTypes.bool,
  path_constraints: ConstraintsWrapper.refObjectType,
});
const GetCartesianPath_RequestRefArray = ArrayType(GetCartesianPath_RequestRefStruct);
const GetCartesianPath_RequestRefStructArray = StructType({
  data: GetCartesianPath_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class GetCartesianPath_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetCartesianPath_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.start_state =  new RobotStateWrapper(other._wrapperFields.start_state);
      this._wrapperFields.group_name =  new StringWrapper(other._wrapperFields.group_name);
      this._wrapperFields.link_name =  new StringWrapper(other._wrapperFields.link_name);
      this._wrapperFields.waypoints = PoseWrapper.createArray();
      this._wrapperFields.waypoints.copy(other._wrapperFields.waypoints);
      this._wrapperFields.path_constraints =  new ConstraintsWrapper(other._wrapperFields.path_constraints);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new GetCartesianPath_RequestRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.start_state = new RobotStateWrapper();
    this._wrapperFields.group_name = new StringWrapper();
    this._wrapperFields.link_name = new StringWrapper();
    this._wrapperFields.waypoints = PoseWrapper.createArray();
    this._wrapperFields.path_constraints = new ConstraintsWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GetCartesianPath_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetCartesianPath_RequestArrayWrapper();
  }
  static get ArrayType() {
    return GetCartesianPath_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetCartesianPath_RequestRefStructArray;
  }
  static get refObjectType() {
    return GetCartesianPath_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.header.freeze(own);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.start_state.freeze(own);
    this._refObject.start_state = this._wrapperFields.start_state.refObject;
    if (own) {
      this._wrapperFields.group_name.freeze(own);
    }
    this._refObject.group_name = this._wrapperFields.group_name.refObject;
    if (own) {
      this._wrapperFields.link_name.freeze(own);
    }
    this._refObject.link_name = this._wrapperFields.link_name.refObject;
    if (own) {
      this._wrapperFields.waypoints.fill([]);
    }
    this._wrapperFields.waypoints.freeze(own);
    this._refObject.waypoints = this._wrapperFields.waypoints.refObject;
    this._wrapperFields.path_constraints.freeze(own);
    this._refObject.path_constraints = this._wrapperFields.path_constraints.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.start_state.copyRefObject(refObject.start_state);
    this._wrapperFields.group_name.data = refObject.group_name.data;
    this._wrapperFields.link_name.data = refObject.link_name.data;
    this._wrapperFields.waypoints.copyRefObject(refObject.waypoints);
    this._refObject.max_step = refObject.max_step;
    this._refObject.jump_threshold = refObject.jump_threshold;
    this._refObject.prismatic_jump_threshold = refObject.prismatic_jump_threshold;
    this._refObject.revolute_jump_threshold = refObject.revolute_jump_threshold;
    this._refObject.avoid_collisions = refObject.avoid_collisions;
    this._wrapperFields.path_constraints.copyRefObject(refObject.path_constraints);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    RobotStateWrapper.freeStruct(refObject.start_state);
    StringWrapper.freeStruct(refObject.group_name);
    StringWrapper.freeStruct(refObject.link_name);
    if (refObject.waypoints.size != 0) {
      PoseWrapper.ArrayType.freeArray(refObject.waypoints);
      if (!PoseWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.waypoints, PoseWrapper.refObjectArrayType, 'data');
      }
    }
    ConstraintsWrapper.freeStruct(refObject.path_constraints);
  }
  static destroyRawROS(msg) {
    GetCartesianPath_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'srv', interfaceName: 'GetCartesianPath_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get start_state() {
    return this._wrapperFields.start_state;
  }
  set start_state(value) {
    if (value instanceof RobotStateWrapper) {
      this._wrapperFields.start_state.copy(value);
    } else {
      this._wrapperFields.start_state.copy(new RobotStateWrapper(value));
    }
  }
  get group_name() {
    return this._wrapperFields.group_name.data;
  }
  set group_name(value) {
    this._wrapperFields.group_name.data = value;
  }
  get link_name() {
    return this._wrapperFields.link_name.data;
  }
  set link_name(value) {
    this._wrapperFields.link_name.data = value;
  }
  get waypoints() {
    return this._wrapperFields.waypoints;
  }
  set waypoints(value) {
    this._wrapperFields.waypoints.fill(value);
  }
  get max_step() {
    return this._refObject.max_step;
  }
  set max_step(value) {
    this._refObject.max_step = value;
  }
  get jump_threshold() {
    return this._refObject.jump_threshold;
  }
  set jump_threshold(value) {
    this._refObject.jump_threshold = value;
  }
  get prismatic_jump_threshold() {
    return this._refObject.prismatic_jump_threshold;
  }
  set prismatic_jump_threshold(value) {
    this._refObject.prismatic_jump_threshold = value;
  }
  get revolute_jump_threshold() {
    return this._refObject.revolute_jump_threshold;
  }
  set revolute_jump_threshold(value) {
    this._refObject.revolute_jump_threshold = value;
  }
  get avoid_collisions() {
    return this._refObject.avoid_collisions;
  }
  set avoid_collisions(value) {
    this._refObject.avoid_collisions = value;
  }
  get path_constraints() {
    return this._wrapperFields.path_constraints;
  }
  set path_constraints(value) {
    if (value instanceof ConstraintsWrapper) {
      this._wrapperFields.path_constraints.copy(value);
    } else {
      this._wrapperFields.path_constraints.copy(new ConstraintsWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new GetCartesianPath_RequestRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.start_state.copyRefObject(this._refObject.start_state);
    this._wrapperFields.group_name.copyRefObject(this._refObject.group_name);
    this._wrapperFields.link_name.copyRefObject(this._refObject.link_name);
    this._wrapperFields.waypoints.copyRefObject(this._refObject.waypoints);
    this._wrapperFields.path_constraints.copyRefObject(this._refObject.path_constraints);
  }
  copy(other) {
    this._refObject = new GetCartesianPath_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.start_state.copy(other._wrapperFields.start_state);
    this._wrapperFields.group_name.copy(other._wrapperFields.group_name);
    this._wrapperFields.link_name.copy(other._wrapperFields.link_name);
    this._wrapperFields.waypoints.copy(other._wrapperFields.waypoints);
    this._wrapperFields.path_constraints.copy(other._wrapperFields.path_constraints);
  }
  static get classType() {
    return GetCartesianPath_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"start_state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"RobotState","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"group_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"link_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"waypoints","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"geometry_msgs","type":"Pose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"max_step","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"jump_threshold","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"prismatic_jump_threshold","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"revolute_jump_threshold","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"avoid_collisions","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"path_constraints","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"Constraints","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"GetCartesianPath_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetCartesianPath_Request"};
  }
  hasMember(name) {
    let memberNames = ["header","start_state","group_name","link_name","waypoints","max_step","jump_threshold","prismatic_jump_threshold","revolute_jump_threshold","avoid_collisions","path_constraints"];
    return memberNames.indexOf(name) !== -1;
  }
}
class GetCartesianPath_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetCartesianPath_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetCartesianPath_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetCartesianPath_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetCartesianPath_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new GetCartesianPath_RequestRefArray(size);
    this._refObject = new GetCartesianPath_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetCartesianPath_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetCartesianPath_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetCartesianPath_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetCartesianPath_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetCartesianPath_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetCartesianPath_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = GetCartesianPath_RequestWrapper;
