// rclnodejs/generated/moveit_msgs/moveit_msgs__srv__QueryPlannerInterfaces_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PlannerInterfaceDescriptionWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__PlannerInterfaceDescription.js');
const QueryPlannerInterfaces_ResponseRefStruct = StructType({
  planner_interfaces: PlannerInterfaceDescriptionWrapper.refObjectArrayType,
});
const QueryPlannerInterfaces_ResponseRefArray = ArrayType(QueryPlannerInterfaces_ResponseRefStruct);
const QueryPlannerInterfaces_ResponseRefStructArray = StructType({
  data: QueryPlannerInterfaces_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class QueryPlannerInterfaces_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new QueryPlannerInterfaces_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.planner_interfaces = PlannerInterfaceDescriptionWrapper.createArray();
      this._wrapperFields.planner_interfaces.copy(other._wrapperFields.planner_interfaces);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new QueryPlannerInterfaces_ResponseRefStruct();
    this._wrapperFields.planner_interfaces = PlannerInterfaceDescriptionWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new QueryPlannerInterfaces_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new QueryPlannerInterfaces_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return QueryPlannerInterfaces_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return QueryPlannerInterfaces_ResponseRefStructArray;
  }
  static get refObjectType() {
    return QueryPlannerInterfaces_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.planner_interfaces.fill([]);
    }
    this._wrapperFields.planner_interfaces.freeze(own);
    this._refObject.planner_interfaces = this._wrapperFields.planner_interfaces.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.planner_interfaces.copyRefObject(refObject.planner_interfaces);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.planner_interfaces.size != 0) {
      PlannerInterfaceDescriptionWrapper.ArrayType.freeArray(refObject.planner_interfaces);
      if (!PlannerInterfaceDescriptionWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.planner_interfaces, PlannerInterfaceDescriptionWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    QueryPlannerInterfaces_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'srv', interfaceName: 'QueryPlannerInterfaces_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get planner_interfaces() {
    return this._wrapperFields.planner_interfaces;
  }
  set planner_interfaces(value) {
    this._wrapperFields.planner_interfaces.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new QueryPlannerInterfaces_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.planner_interfaces.copyRefObject(this._refObject.planner_interfaces);
  }
  copy(other) {
    this._refObject = new QueryPlannerInterfaces_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.planner_interfaces.copy(other._wrapperFields.planner_interfaces);
  }
  static get classType() {
    return QueryPlannerInterfaces_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"planner_interfaces","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"moveit_msgs","type":"PlannerInterfaceDescription","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"QueryPlannerInterfaces_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"QueryPlannerInterfaces_Response"};
  }
  hasMember(name) {
    let memberNames = ["planner_interfaces"];
    return memberNames.indexOf(name) !== -1;
  }
}
class QueryPlannerInterfaces_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof QueryPlannerInterfaces_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new QueryPlannerInterfaces_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to QueryPlannerInterfaces_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to QueryPlannerInterfaces_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new QueryPlannerInterfaces_ResponseRefArray(size);
    this._refObject = new QueryPlannerInterfaces_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new QueryPlannerInterfaces_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof QueryPlannerInterfaces_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "QueryPlannerInterfaces_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      QueryPlannerInterfaces_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return QueryPlannerInterfaces_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return QueryPlannerInterfaces_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = QueryPlannerInterfaces_ResponseWrapper;
