// rclnodejs/generated/moveit_msgs/moveit_msgs__srv__GraspPlanning_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const GraspWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__Grasp.js');
const MoveItErrorCodesWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__MoveItErrorCodes.js');
const GraspPlanning_ResponseRefStruct = StructType({
  grasps: GraspWrapper.refObjectArrayType,
  error_code: MoveItErrorCodesWrapper.refObjectType,
});
const GraspPlanning_ResponseRefArray = ArrayType(GraspPlanning_ResponseRefStruct);
const GraspPlanning_ResponseRefStructArray = StructType({
  data: GraspPlanning_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class GraspPlanning_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new GraspPlanning_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.grasps = GraspWrapper.createArray();
      this._wrapperFields.grasps.copy(other._wrapperFields.grasps);
      this._wrapperFields.error_code =  new MoveItErrorCodesWrapper(other._wrapperFields.error_code);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new GraspPlanning_ResponseRefStruct();
    this._wrapperFields.grasps = GraspWrapper.createArray();
    this._wrapperFields.error_code = new MoveItErrorCodesWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GraspPlanning_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GraspPlanning_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return GraspPlanning_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GraspPlanning_ResponseRefStructArray;
  }
  static get refObjectType() {
    return GraspPlanning_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.grasps.fill([]);
    }
    this._wrapperFields.grasps.freeze(own);
    this._refObject.grasps = this._wrapperFields.grasps.refObject;
    this._wrapperFields.error_code.freeze(own);
    this._refObject.error_code = this._wrapperFields.error_code.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.grasps.copyRefObject(refObject.grasps);
    this._wrapperFields.error_code.copyRefObject(refObject.error_code);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.grasps.size != 0) {
      GraspWrapper.ArrayType.freeArray(refObject.grasps);
      if (!GraspWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.grasps, GraspWrapper.refObjectArrayType, 'data');
      }
    }
    MoveItErrorCodesWrapper.freeStruct(refObject.error_code);
  }
  static destroyRawROS(msg) {
    GraspPlanning_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'srv', interfaceName: 'GraspPlanning_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get grasps() {
    return this._wrapperFields.grasps;
  }
  set grasps(value) {
    this._wrapperFields.grasps.fill(value);
  }
  get error_code() {
    return this._wrapperFields.error_code;
  }
  set error_code(value) {
    if (value instanceof MoveItErrorCodesWrapper) {
      this._wrapperFields.error_code.copy(value);
    } else {
      this._wrapperFields.error_code.copy(new MoveItErrorCodesWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new GraspPlanning_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.grasps.copyRefObject(this._refObject.grasps);
    this._wrapperFields.error_code.copyRefObject(this._refObject.error_code);
  }
  copy(other) {
    this._refObject = new GraspPlanning_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.grasps.copy(other._wrapperFields.grasps);
    this._wrapperFields.error_code.copy(other._wrapperFields.error_code);
  }
  static get classType() {
    return GraspPlanning_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"grasps","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"moveit_msgs","type":"Grasp","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"error_code","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"MoveItErrorCodes","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"GraspPlanning_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GraspPlanning_Response"};
  }
  hasMember(name) {
    let memberNames = ["grasps","error_code"];
    return memberNames.indexOf(name) !== -1;
  }
}
class GraspPlanning_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GraspPlanning_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GraspPlanning_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GraspPlanning_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GraspPlanning_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new GraspPlanning_ResponseRefArray(size);
    this._refObject = new GraspPlanning_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GraspPlanning_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GraspPlanning_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GraspPlanning_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GraspPlanning_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GraspPlanning_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GraspPlanning_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = GraspPlanning_ResponseWrapper;
