// rclnodejs/generated/moveit_msgs/moveit_msgs__srv__UpdatePointcloudOctomap_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PointCloud2Wrapper = require('../../generated/sensor_msgs/sensor_msgs__msg__PointCloud2.js');
const UpdatePointcloudOctomap_RequestRefStruct = StructType({
  cloud: PointCloud2Wrapper.refObjectType,
});
const UpdatePointcloudOctomap_RequestRefArray = ArrayType(UpdatePointcloudOctomap_RequestRefStruct);
const UpdatePointcloudOctomap_RequestRefStructArray = StructType({
  data: UpdatePointcloudOctomap_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class UpdatePointcloudOctomap_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new UpdatePointcloudOctomap_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.cloud =  new PointCloud2Wrapper(other._wrapperFields.cloud);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new UpdatePointcloudOctomap_RequestRefStruct();
    this._wrapperFields.cloud = new PointCloud2Wrapper();
  }
  static createFromRefObject(refObject) {
    let self = new UpdatePointcloudOctomap_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new UpdatePointcloudOctomap_RequestArrayWrapper();
  }
  static get ArrayType() {
    return UpdatePointcloudOctomap_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return UpdatePointcloudOctomap_RequestRefStructArray;
  }
  static get refObjectType() {
    return UpdatePointcloudOctomap_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.cloud.freeze(own);
    this._refObject.cloud = this._wrapperFields.cloud.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.cloud.copyRefObject(refObject.cloud);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PointCloud2Wrapper.freeStruct(refObject.cloud);
  }
  static destroyRawROS(msg) {
    UpdatePointcloudOctomap_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'srv', interfaceName: 'UpdatePointcloudOctomap_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get cloud() {
    return this._wrapperFields.cloud;
  }
  set cloud(value) {
    if (value instanceof PointCloud2Wrapper) {
      this._wrapperFields.cloud.copy(value);
    } else {
      this._wrapperFields.cloud.copy(new PointCloud2Wrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new UpdatePointcloudOctomap_RequestRefStruct(refObject.toObject());
    this._wrapperFields.cloud.copyRefObject(this._refObject.cloud);
  }
  copy(other) {
    this._refObject = new UpdatePointcloudOctomap_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.cloud.copy(other._wrapperFields.cloud);
  }
  static get classType() {
    return UpdatePointcloudOctomap_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"cloud","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"sensor_msgs","type":"PointCloud2","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"UpdatePointcloudOctomap_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"UpdatePointcloudOctomap_Request"};
  }
  hasMember(name) {
    let memberNames = ["cloud"];
    return memberNames.indexOf(name) !== -1;
  }
}
class UpdatePointcloudOctomap_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof UpdatePointcloudOctomap_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new UpdatePointcloudOctomap_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to UpdatePointcloudOctomap_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to UpdatePointcloudOctomap_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new UpdatePointcloudOctomap_RequestRefArray(size);
    this._refObject = new UpdatePointcloudOctomap_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new UpdatePointcloudOctomap_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof UpdatePointcloudOctomap_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "UpdatePointcloudOctomap_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      UpdatePointcloudOctomap_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return UpdatePointcloudOctomap_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return UpdatePointcloudOctomap_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = UpdatePointcloudOctomap_RequestWrapper;
