// rclnodejs/generated/moveit_msgs/moveit_msgs__srv__ChangeControlDimensions_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const ChangeControlDimensions_RequestRefStruct = StructType({
  control_x_translation: primitiveTypes.bool,
  control_y_translation: primitiveTypes.bool,
  control_z_translation: primitiveTypes.bool,
  control_x_rotation: primitiveTypes.bool,
  control_y_rotation: primitiveTypes.bool,
  control_z_rotation: primitiveTypes.bool,
});
const ChangeControlDimensions_RequestRefArray = ArrayType(ChangeControlDimensions_RequestRefStruct);
const ChangeControlDimensions_RequestRefStructArray = StructType({
  data: ChangeControlDimensions_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class ChangeControlDimensions_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new ChangeControlDimensions_RequestRefStruct(other._refObject.toObject());
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new ChangeControlDimensions_RequestRefStruct();
  }
  static createFromRefObject(refObject) {
    let self = new ChangeControlDimensions_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ChangeControlDimensions_RequestArrayWrapper();
  }
  static get ArrayType() {
    return ChangeControlDimensions_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return ChangeControlDimensions_RequestRefStructArray;
  }
  static get refObjectType() {
    return ChangeControlDimensions_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.control_x_translation = refObject.control_x_translation;
    this._refObject.control_y_translation = refObject.control_y_translation;
    this._refObject.control_z_translation = refObject.control_z_translation;
    this._refObject.control_x_rotation = refObject.control_x_rotation;
    this._refObject.control_y_rotation = refObject.control_y_rotation;
    this._refObject.control_z_rotation = refObject.control_z_rotation;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    ChangeControlDimensions_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'srv', interfaceName: 'ChangeControlDimensions_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get control_x_translation() {
    return this._refObject.control_x_translation;
  }
  set control_x_translation(value) {
    this._refObject.control_x_translation = value;
  }
  get control_y_translation() {
    return this._refObject.control_y_translation;
  }
  set control_y_translation(value) {
    this._refObject.control_y_translation = value;
  }
  get control_z_translation() {
    return this._refObject.control_z_translation;
  }
  set control_z_translation(value) {
    this._refObject.control_z_translation = value;
  }
  get control_x_rotation() {
    return this._refObject.control_x_rotation;
  }
  set control_x_rotation(value) {
    this._refObject.control_x_rotation = value;
  }
  get control_y_rotation() {
    return this._refObject.control_y_rotation;
  }
  set control_y_rotation(value) {
    this._refObject.control_y_rotation = value;
  }
  get control_z_rotation() {
    return this._refObject.control_z_rotation;
  }
  set control_z_rotation(value) {
    this._refObject.control_z_rotation = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ChangeControlDimensions_RequestRefStruct(refObject.toObject());
  }
  copy(other) {
    this._refObject = new ChangeControlDimensions_RequestRefStruct(other._refObject.toObject());
  }
  static get classType() {
    return ChangeControlDimensions_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"control_x_translation","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"control_y_translation","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"control_z_translation","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"control_x_rotation","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"control_y_rotation","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"control_z_rotation","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"ChangeControlDimensions_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ChangeControlDimensions_Request"};
  }
  hasMember(name) {
    let memberNames = ["control_x_translation","control_y_translation","control_z_translation","control_x_rotation","control_y_rotation","control_z_rotation"];
    return memberNames.indexOf(name) !== -1;
  }
}
class ChangeControlDimensions_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ChangeControlDimensions_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ChangeControlDimensions_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ChangeControlDimensions_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ChangeControlDimensions_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new ChangeControlDimensions_RequestRefArray(size);
    this._refObject = new ChangeControlDimensions_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ChangeControlDimensions_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ChangeControlDimensions_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ChangeControlDimensions_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ChangeControlDimensions_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ChangeControlDimensions_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ChangeControlDimensions_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = ChangeControlDimensions_RequestWrapper;
