// rclnodejs/generated/moveit_msgs/moveit_msgs__msg__LinkScale.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const LinkScaleRefStruct = StructType({
  link_name: primitiveTypes.string,
  scale: primitiveTypes.float64,
});
const LinkScaleRefArray = ArrayType(LinkScaleRefStruct);
const LinkScaleRefStructArray = StructType({
  data: LinkScaleRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class LinkScaleWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new LinkScaleRefStruct(other._refObject.toObject());
      this._wrapperFields.link_name =  new StringWrapper(other._wrapperFields.link_name);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new LinkScaleRefStruct();
    this._wrapperFields.link_name = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new LinkScaleWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LinkScaleArrayWrapper();
  }
  static get ArrayType() {
    return LinkScaleArrayWrapper;
  }
  static get refObjectArrayType() {
    return LinkScaleRefStructArray;
  }
  static get refObjectType() {
    return LinkScaleRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.link_name.freeze(own);
    }
    this._refObject.link_name = this._wrapperFields.link_name.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.link_name.data = refObject.link_name.data;
    this._refObject.scale = refObject.scale;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.link_name);
  }
  static destroyRawROS(msg) {
    LinkScaleWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'msg', interfaceName: 'LinkScale'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get link_name() {
    return this._wrapperFields.link_name.data;
  }
  set link_name(value) {
    this._wrapperFields.link_name.data = value;
  }
  get scale() {
    return this._refObject.scale;
  }
  set scale(value) {
    this._refObject.scale = value;
  }
  copyRefObject(refObject) {
    this._refObject = new LinkScaleRefStruct(refObject.toObject());
    this._wrapperFields.link_name.copyRefObject(this._refObject.link_name);
  }
  copy(other) {
    this._refObject = new LinkScaleRefStruct(other._refObject.toObject());
    this._wrapperFields.link_name.copy(other._wrapperFields.link_name);
  }
  static get classType() {
    return LinkScaleWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"link_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"scale","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"LinkScale","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LinkScale"};
  }
  hasMember(name) {
    let memberNames = ["link_name","scale"];
    return memberNames.indexOf(name) !== -1;
  }
}
class LinkScaleArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LinkScaleWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LinkScaleWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LinkScaleArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LinkScaleArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new LinkScaleRefArray(size);
    this._refObject = new LinkScaleRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LinkScaleWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LinkScaleArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LinkScaleArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LinkScaleWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LinkScaleWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LinkScaleArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = LinkScaleWrapper;
