// rclnodejs/generated/moveit_msgs/moveit_msgs__srv__GraspPlanning_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const CollisionObjectWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__CollisionObject.js');
const GraspWrapper = require('../../generated/moveit_msgs/moveit_msgs__msg__Grasp.js');
const GraspPlanning_RequestRefStruct = StructType({
  group_name: primitiveTypes.string,
  target: CollisionObjectWrapper.refObjectType,
  support_surfaces: StringWrapper.refObjectArrayType,
  candidate_grasps: GraspWrapper.refObjectArrayType,
  movable_obstacles: CollisionObjectWrapper.refObjectArrayType,
});
const GraspPlanning_RequestRefArray = ArrayType(GraspPlanning_RequestRefStruct);
const GraspPlanning_RequestRefStructArray = StructType({
  data: GraspPlanning_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class GraspPlanning_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new GraspPlanning_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.group_name =  new StringWrapper(other._wrapperFields.group_name);
      this._wrapperFields.target =  new CollisionObjectWrapper(other._wrapperFields.target);
      this._wrapperFields.support_surfaces = StringWrapper.createArray();
      this._wrapperFields.support_surfaces.copy(other._wrapperFields.support_surfaces);
      this.support_surfaces = other.support_surfaces;
      this._wrapperFields.candidate_grasps = GraspWrapper.createArray();
      this._wrapperFields.candidate_grasps.copy(other._wrapperFields.candidate_grasps);
      this._wrapperFields.movable_obstacles = CollisionObjectWrapper.createArray();
      this._wrapperFields.movable_obstacles.copy(other._wrapperFields.movable_obstacles);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new GraspPlanning_RequestRefStruct();
    this._wrapperFields.group_name = new StringWrapper();
    this._wrapperFields.target = new CollisionObjectWrapper();
    this._wrapperFields.support_surfaces = StringWrapper.createArray();
    this._support_surfacesArray = [];
    this._wrapperFields.candidate_grasps = GraspWrapper.createArray();
    this._wrapperFields.movable_obstacles = CollisionObjectWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new GraspPlanning_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GraspPlanning_RequestArrayWrapper();
  }
  static get ArrayType() {
    return GraspPlanning_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return GraspPlanning_RequestRefStructArray;
  }
  static get refObjectType() {
    return GraspPlanning_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.group_name.freeze(own);
    }
    this._refObject.group_name = this._wrapperFields.group_name.refObject;
    this._wrapperFields.target.freeze(own);
    this._refObject.target = this._wrapperFields.target.refObject;
    if (own) {
      this._wrapperFields.support_surfaces.fill([]);
    } else {
      this._wrapperFields.support_surfaces.fill(this._support_surfacesArray);
    }
    this._wrapperFields.support_surfaces.freeze(own);
    this._refObject.support_surfaces = this._wrapperFields.support_surfaces.refObject;
    if (own) {
      this._wrapperFields.candidate_grasps.fill([]);
    }
    this._wrapperFields.candidate_grasps.freeze(own);
    this._refObject.candidate_grasps = this._wrapperFields.candidate_grasps.refObject;
    if (own) {
      this._wrapperFields.movable_obstacles.fill([]);
    }
    this._wrapperFields.movable_obstacles.freeze(own);
    this._refObject.movable_obstacles = this._wrapperFields.movable_obstacles.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.group_name.data = refObject.group_name.data;
    this._wrapperFields.target.copyRefObject(refObject.target);
    refObject.support_surfaces.data.length = refObject.support_surfaces.size;
    for (let index = 0; index < refObject.support_surfaces.size; index++) {
      this._support_surfacesArray[index] = refObject.support_surfaces.data[index].data;
    }
    this._wrapperFields.candidate_grasps.copyRefObject(refObject.candidate_grasps);
    this._wrapperFields.movable_obstacles.copyRefObject(refObject.movable_obstacles);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.group_name);
    CollisionObjectWrapper.freeStruct(refObject.target);
    if (refObject.support_surfaces.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.support_surfaces);
      if (!StringWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.support_surfaces, StringWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.candidate_grasps.size != 0) {
      GraspWrapper.ArrayType.freeArray(refObject.candidate_grasps);
      if (!GraspWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.candidate_grasps, GraspWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.movable_obstacles.size != 0) {
      CollisionObjectWrapper.ArrayType.freeArray(refObject.movable_obstacles);
      if (!CollisionObjectWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.movable_obstacles, CollisionObjectWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    GraspPlanning_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'moveit_msgs', subFolder: 'srv', interfaceName: 'GraspPlanning_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get group_name() {
    return this._wrapperFields.group_name.data;
  }
  set group_name(value) {
    this._wrapperFields.group_name.data = value;
  }
  get target() {
    return this._wrapperFields.target;
  }
  set target(value) {
    if (value instanceof CollisionObjectWrapper) {
      this._wrapperFields.target.copy(value);
    } else {
      this._wrapperFields.target.copy(new CollisionObjectWrapper(value));
    }
  }
  get support_surfaces() {
    return this._support_surfacesArray;
  }
  set support_surfaces(value) {
    this._support_surfacesArray = value;
  }
  get candidate_grasps() {
    return this._wrapperFields.candidate_grasps;
  }
  set candidate_grasps(value) {
    this._wrapperFields.candidate_grasps.fill(value);
  }
  get movable_obstacles() {
    return this._wrapperFields.movable_obstacles;
  }
  set movable_obstacles(value) {
    this._wrapperFields.movable_obstacles.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new GraspPlanning_RequestRefStruct(refObject.toObject());
    this._wrapperFields.group_name.copyRefObject(this._refObject.group_name);
    this._wrapperFields.target.copyRefObject(this._refObject.target);
    refObject.support_surfaces.data.length = refObject.support_surfaces.size;
    for (let index = 0; index < refObject.support_surfaces.size; index++) {
      this._support_surfacesArray[index] = refObject.support_surfaces.data[index].data;
    }
    this._wrapperFields.candidate_grasps.copyRefObject(this._refObject.candidate_grasps);
    this._wrapperFields.movable_obstacles.copyRefObject(this._refObject.movable_obstacles);
  }
  copy(other) {
    this._refObject = new GraspPlanning_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.group_name.copy(other._wrapperFields.group_name);
    this._wrapperFields.target.copy(other._wrapperFields.target);
    this._support_surfacesArray = other._support_surfacesArray.slice();
    this._wrapperFields.candidate_grasps.copy(other._wrapperFields.candidate_grasps);
    this._wrapperFields.movable_obstacles.copy(other._wrapperFields.movable_obstacles);
  }
  static get classType() {
    return GraspPlanning_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"group_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"target","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"moveit_msgs","type":"CollisionObject","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"support_surfaces","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"candidate_grasps","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"moveit_msgs","type":"Grasp","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"movable_obstacles","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"moveit_msgs","type":"CollisionObject","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"moveit_msgs","type":"GraspPlanning_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GraspPlanning_Request"};
  }
  hasMember(name) {
    let memberNames = ["group_name","target","support_surfaces","candidate_grasps","movable_obstacles"];
    return memberNames.indexOf(name) !== -1;
  }
}
class GraspPlanning_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GraspPlanning_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GraspPlanning_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GraspPlanning_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GraspPlanning_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new GraspPlanning_RequestRefArray(size);
    this._refObject = new GraspPlanning_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GraspPlanning_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GraspPlanning_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GraspPlanning_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GraspPlanning_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GraspPlanning_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GraspPlanning_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = GraspPlanning_RequestWrapper;
