// rclnodejs/generated/controller_manager_msgs/controller_manager_msgs__srv__SwitchController_Request.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const DurationWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Duration.js');
const SwitchController_RequestRefStruct = StructType({
  activate_controllers: StringWrapper.refObjectArrayType,
  deactivate_controllers: StringWrapper.refObjectArrayType,
  start_controllers: StringWrapper.refObjectArrayType,
  stop_controllers: StringWrapper.refObjectArrayType,
  strictness: primitiveTypes.int32,
  start_asap: primitiveTypes.bool,
  activate_asap: primitiveTypes.bool,
  timeout: DurationWrapper.refObjectType,
});
const SwitchController_RequestRefArray = ArrayType(SwitchController_RequestRefStruct);
const SwitchController_RequestRefStructArray = StructType({
  data: SwitchController_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class SwitchController_RequestWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new SwitchController_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.activate_controllers = StringWrapper.createArray();
      this._wrapperFields.activate_controllers.copy(other._wrapperFields.activate_controllers);
      this.activate_controllers = other.activate_controllers;
      this._wrapperFields.deactivate_controllers = StringWrapper.createArray();
      this._wrapperFields.deactivate_controllers.copy(other._wrapperFields.deactivate_controllers);
      this.deactivate_controllers = other.deactivate_controllers;
      this._wrapperFields.start_controllers = StringWrapper.createArray();
      this._wrapperFields.start_controllers.copy(other._wrapperFields.start_controllers);
      this.start_controllers = other.start_controllers;
      this._wrapperFields.stop_controllers = StringWrapper.createArray();
      this._wrapperFields.stop_controllers.copy(other._wrapperFields.stop_controllers);
      this.stop_controllers = other.stop_controllers;
      this._wrapperFields.timeout =  new DurationWrapper(other._wrapperFields.timeout);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new SwitchController_RequestRefStruct();
    this._wrapperFields.activate_controllers = StringWrapper.createArray();
    this._activate_controllersArray = [];
    this._wrapperFields.deactivate_controllers = StringWrapper.createArray();
    this._deactivate_controllersArray = [];
    this._wrapperFields.start_controllers = StringWrapper.createArray();
    this._start_controllersArray = [];
    this._wrapperFields.stop_controllers = StringWrapper.createArray();
    this._stop_controllersArray = [];
    this._wrapperFields.timeout = new DurationWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new SwitchController_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SwitchController_RequestArrayWrapper();
  }
  static get ArrayType() {
    return SwitchController_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return SwitchController_RequestRefStructArray;
  }
  static get refObjectType() {
    return SwitchController_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.activate_controllers.fill([]);
    } else {
      this._wrapperFields.activate_controllers.fill(this._activate_controllersArray);
    }
    this._wrapperFields.activate_controllers.freeze(own);
    this._refObject.activate_controllers = this._wrapperFields.activate_controllers.refObject;
    if (own) {
      this._wrapperFields.deactivate_controllers.fill([]);
    } else {
      this._wrapperFields.deactivate_controllers.fill(this._deactivate_controllersArray);
    }
    this._wrapperFields.deactivate_controllers.freeze(own);
    this._refObject.deactivate_controllers = this._wrapperFields.deactivate_controllers.refObject;
    if (own) {
      this._wrapperFields.start_controllers.fill([]);
    } else {
      this._wrapperFields.start_controllers.fill(this._start_controllersArray);
    }
    this._wrapperFields.start_controllers.freeze(own);
    this._refObject.start_controllers = this._wrapperFields.start_controllers.refObject;
    if (own) {
      this._wrapperFields.stop_controllers.fill([]);
    } else {
      this._wrapperFields.stop_controllers.fill(this._stop_controllersArray);
    }
    this._wrapperFields.stop_controllers.freeze(own);
    this._refObject.stop_controllers = this._wrapperFields.stop_controllers.refObject;
    this._wrapperFields.timeout.freeze(own);
    this._refObject.timeout = this._wrapperFields.timeout.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    refObject.activate_controllers.data.length = refObject.activate_controllers.size;
    for (let index = 0; index < refObject.activate_controllers.size; index++) {
      this._activate_controllersArray[index] = refObject.activate_controllers.data[index].data;
    }
    refObject.deactivate_controllers.data.length = refObject.deactivate_controllers.size;
    for (let index = 0; index < refObject.deactivate_controllers.size; index++) {
      this._deactivate_controllersArray[index] = refObject.deactivate_controllers.data[index].data;
    }
    refObject.start_controllers.data.length = refObject.start_controllers.size;
    for (let index = 0; index < refObject.start_controllers.size; index++) {
      this._start_controllersArray[index] = refObject.start_controllers.data[index].data;
    }
    refObject.stop_controllers.data.length = refObject.stop_controllers.size;
    for (let index = 0; index < refObject.stop_controllers.size; index++) {
      this._stop_controllersArray[index] = refObject.stop_controllers.data[index].data;
    }
    this._refObject.strictness = refObject.strictness;
    this._refObject.start_asap = refObject.start_asap;
    this._refObject.activate_asap = refObject.activate_asap;
    this._wrapperFields.timeout.copyRefObject(refObject.timeout);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.activate_controllers.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.activate_controllers);
      if (!StringWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.activate_controllers, StringWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.deactivate_controllers.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.deactivate_controllers);
      if (!StringWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.deactivate_controllers, StringWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.start_controllers.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.start_controllers);
      if (!StringWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.start_controllers, StringWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.stop_controllers.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.stop_controllers);
      if (!StringWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.stop_controllers, StringWrapper.refObjectArrayType, 'data');
      }
    }
    DurationWrapper.freeStruct(refObject.timeout);
  }
  static destroyRawROS(msg) {
    SwitchController_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'controller_manager_msgs', subFolder: 'srv', interfaceName: 'SwitchController_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get activate_controllers() {
    return this._activate_controllersArray;
  }
  set activate_controllers(value) {
    this._activate_controllersArray = value;
  }
  get deactivate_controllers() {
    return this._deactivate_controllersArray;
  }
  set deactivate_controllers(value) {
    this._deactivate_controllersArray = value;
  }
  get start_controllers() {
    return this._start_controllersArray;
  }
  set start_controllers(value) {
    this._start_controllersArray = value;
  }
  get stop_controllers() {
    return this._stop_controllersArray;
  }
  set stop_controllers(value) {
    this._stop_controllersArray = value;
  }
  get strictness() {
    return this._refObject.strictness;
  }
  set strictness(value) {
    this._refObject.strictness = value;
  }
  get start_asap() {
    return this._refObject.start_asap;
  }
  set start_asap(value) {
    this._refObject.start_asap = value;
  }
  get activate_asap() {
    return this._refObject.activate_asap;
  }
  set activate_asap(value) {
    this._refObject.activate_asap = value;
  }
  get timeout() {
    return this._wrapperFields.timeout;
  }
  set timeout(value) {
    if (value instanceof DurationWrapper) {
      this._wrapperFields.timeout.copy(value);
    } else {
      this._wrapperFields.timeout.copy(new DurationWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new SwitchController_RequestRefStruct(refObject.toObject());
    refObject.activate_controllers.data.length = refObject.activate_controllers.size;
    for (let index = 0; index < refObject.activate_controllers.size; index++) {
      this._activate_controllersArray[index] = refObject.activate_controllers.data[index].data;
    }
    refObject.deactivate_controllers.data.length = refObject.deactivate_controllers.size;
    for (let index = 0; index < refObject.deactivate_controllers.size; index++) {
      this._deactivate_controllersArray[index] = refObject.deactivate_controllers.data[index].data;
    }
    refObject.start_controllers.data.length = refObject.start_controllers.size;
    for (let index = 0; index < refObject.start_controllers.size; index++) {
      this._start_controllersArray[index] = refObject.start_controllers.data[index].data;
    }
    refObject.stop_controllers.data.length = refObject.stop_controllers.size;
    for (let index = 0; index < refObject.stop_controllers.size; index++) {
      this._stop_controllersArray[index] = refObject.stop_controllers.data[index].data;
    }
    this._wrapperFields.timeout.copyRefObject(this._refObject.timeout);
  }
  copy(other) {
    this._refObject = new SwitchController_RequestRefStruct(other._refObject.toObject());
    this._activate_controllersArray = other._activate_controllersArray.slice();
    this._deactivate_controllersArray = other._deactivate_controllersArray.slice();
    this._start_controllersArray = other._start_controllersArray.slice();
    this._stop_controllersArray = other._stop_controllersArray.slice();
    this._wrapperFields.timeout.copy(other._wrapperFields.timeout);
  }
  static get classType() {
    return SwitchController_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"int32","name":"BEST_EFFORT","value":1},{"type":"int32","name":"STRICT","value":2}],"fields":[{"name":"activate_controllers","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"deactivate_controllers","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"start_controllers","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"stop_controllers","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"strictness","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"start_asap","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"activate_asap","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"timeout","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Duration","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"controller_manager_msgs","type":"SwitchController_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SwitchController_Request"};
  }
  hasMember(name) {
    let memberNames = ["activate_controllers","deactivate_controllers","start_controllers","stop_controllers","strictness","start_asap","activate_asap","timeout"];
    return memberNames.indexOf(name) !== -1;
  }
}
class SwitchController_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SwitchController_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SwitchController_RequestWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SwitchController_RequestArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SwitchController_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new SwitchController_RequestRefArray(size);
    this._refObject = new SwitchController_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SwitchController_RequestWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SwitchController_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SwitchController_RequestArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SwitchController_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SwitchController_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SwitchController_RequestArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
Object.defineProperty(SwitchController_RequestWrapper, "BEST_EFFORT", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SwitchController_RequestWrapper, "STRICT", {value: 2, writable: false, enumerable: true, configurable: true});
module.exports = SwitchController_RequestWrapper;
