// rclnodejs/generated/controller_manager_msgs/controller_manager_msgs__msg__ChainConnection.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const ChainConnectionRefStruct = StructType({
  name: primitiveTypes.string,
  reference_interfaces: StringWrapper.refObjectArrayType,
});
const ChainConnectionRefArray = ArrayType(ChainConnectionRefStruct);
const ChainConnectionRefStructArray = StructType({
  data: ChainConnectionRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class ChainConnectionWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new ChainConnectionRefStruct(other._refObject.toObject());
      this._wrapperFields.name =  new StringWrapper(other._wrapperFields.name);
      this._wrapperFields.reference_interfaces = StringWrapper.createArray();
      this._wrapperFields.reference_interfaces.copy(other._wrapperFields.reference_interfaces);
      this.reference_interfaces = other.reference_interfaces;
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new ChainConnectionRefStruct();
    this._wrapperFields.name = new StringWrapper();
    this._wrapperFields.reference_interfaces = StringWrapper.createArray();
    this._reference_interfacesArray = [];
  }
  static createFromRefObject(refObject) {
    let self = new ChainConnectionWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ChainConnectionArrayWrapper();
  }
  static get ArrayType() {
    return ChainConnectionArrayWrapper;
  }
  static get refObjectArrayType() {
    return ChainConnectionRefStructArray;
  }
  static get refObjectType() {
    return ChainConnectionRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.name.freeze(own);
    }
    this._refObject.name = this._wrapperFields.name.refObject;
    if (own) {
      this._wrapperFields.reference_interfaces.fill([]);
    } else {
      this._wrapperFields.reference_interfaces.fill(this._reference_interfacesArray);
    }
    this._wrapperFields.reference_interfaces.freeze(own);
    this._refObject.reference_interfaces = this._wrapperFields.reference_interfaces.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.name.data = refObject.name.data;
    refObject.reference_interfaces.data.length = refObject.reference_interfaces.size;
    for (let index = 0; index < refObject.reference_interfaces.size; index++) {
      this._reference_interfacesArray[index] = refObject.reference_interfaces.data[index].data;
    }
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.name);
    if (refObject.reference_interfaces.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.reference_interfaces);
      if (!StringWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.reference_interfaces, StringWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    ChainConnectionWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'controller_manager_msgs', subFolder: 'msg', interfaceName: 'ChainConnection'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get name() {
    return this._wrapperFields.name.data;
  }
  set name(value) {
    this._wrapperFields.name.data = value;
  }
  get reference_interfaces() {
    return this._reference_interfacesArray;
  }
  set reference_interfaces(value) {
    this._reference_interfacesArray = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ChainConnectionRefStruct(refObject.toObject());
    this._wrapperFields.name.copyRefObject(this._refObject.name);
    refObject.reference_interfaces.data.length = refObject.reference_interfaces.size;
    for (let index = 0; index < refObject.reference_interfaces.size; index++) {
      this._reference_interfacesArray[index] = refObject.reference_interfaces.data[index].data;
    }
  }
  copy(other) {
    this._refObject = new ChainConnectionRefStruct(other._refObject.toObject());
    this._wrapperFields.name.copy(other._wrapperFields.name);
    this._reference_interfacesArray = other._reference_interfacesArray.slice();
  }
  static get classType() {
    return ChainConnectionWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reference_interfaces","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"controller_manager_msgs","type":"ChainConnection","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ChainConnection"};
  }
  hasMember(name) {
    let memberNames = ["name","reference_interfaces"];
    return memberNames.indexOf(name) !== -1;
  }
}
class ChainConnectionArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ChainConnectionWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ChainConnectionWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ChainConnectionArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ChainConnectionArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new ChainConnectionRefArray(size);
    this._refObject = new ChainConnectionRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ChainConnectionWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ChainConnectionArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ChainConnectionArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ChainConnectionWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ChainConnectionWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ChainConnectionArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = ChainConnectionWrapper;
