// rclnodejs/generated/object_recognition_msgs/object_recognition_msgs__action__ObjectRecognition_Feedback.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const ObjectRecognition_FeedbackRefStruct = StructType({
  _dummy: primitiveTypes.bool,
});
const ObjectRecognition_FeedbackRefArray = ArrayType(ObjectRecognition_FeedbackRefStruct);
const ObjectRecognition_FeedbackRefStructArray = StructType({
  data: ObjectRecognition_FeedbackRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class ObjectRecognition_FeedbackWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new ObjectRecognition_FeedbackRefStruct(other._refObject.toObject());
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new ObjectRecognition_FeedbackRefStruct();
  }
  static createFromRefObject(refObject) {
    let self = new ObjectRecognition_FeedbackWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ObjectRecognition_FeedbackArrayWrapper();
  }
  static get ArrayType() {
    return ObjectRecognition_FeedbackArrayWrapper;
  }
  static get refObjectArrayType() {
    return ObjectRecognition_FeedbackRefStructArray;
  }
  static get refObjectType() {
    return ObjectRecognition_FeedbackRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject._dummy = refObject._dummy;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    ObjectRecognition_FeedbackWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'object_recognition_msgs', subFolder: 'action', interfaceName: 'ObjectRecognition_Feedback'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get _dummy() {
    return this._refObject._dummy;
  }
  set _dummy(value) {
    this._refObject._dummy = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ObjectRecognition_FeedbackRefStruct(refObject.toObject());
  }
  copy(other) {
    this._refObject = new ObjectRecognition_FeedbackRefStruct(other._refObject.toObject());
  }
  static get classType() {
    return ObjectRecognition_FeedbackWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[],"baseType":{"pkgName":"object_recognition_msgs","type":"ObjectRecognition_Feedback","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ObjectRecognition_Feedback"};
  }
  hasMember(name) {
    let memberNames = ["_dummy"];
    return memberNames.indexOf(name) !== -1;
  }
}
class ObjectRecognition_FeedbackArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ObjectRecognition_FeedbackWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ObjectRecognition_FeedbackWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ObjectRecognition_FeedbackArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ObjectRecognition_FeedbackArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new ObjectRecognition_FeedbackRefArray(size);
    this._refObject = new ObjectRecognition_FeedbackRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ObjectRecognition_FeedbackWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ObjectRecognition_FeedbackArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ObjectRecognition_FeedbackArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ObjectRecognition_FeedbackWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ObjectRecognition_FeedbackWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ObjectRecognition_FeedbackArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = ObjectRecognition_FeedbackWrapper;
