// rclnodejs/generated/object_recognition_msgs/object_recognition_msgs__msg__ObjectInformation.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const MeshWrapper = require('../../generated/shape_msgs/shape_msgs__msg__Mesh.js');
const PointCloud2Wrapper = require('../../generated/sensor_msgs/sensor_msgs__msg__PointCloud2.js');
const ObjectInformationRefStruct = StructType({
  name: primitiveTypes.string,
  ground_truth_mesh: MeshWrapper.refObjectType,
  ground_truth_point_cloud: PointCloud2Wrapper.refObjectType,
});
const ObjectInformationRefArray = ArrayType(ObjectInformationRefStruct);
const ObjectInformationRefStructArray = StructType({
  data: ObjectInformationRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class ObjectInformationWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new ObjectInformationRefStruct(other._refObject.toObject());
      this._wrapperFields.name =  new StringWrapper(other._wrapperFields.name);
      this._wrapperFields.ground_truth_mesh =  new MeshWrapper(other._wrapperFields.ground_truth_mesh);
      this._wrapperFields.ground_truth_point_cloud =  new PointCloud2Wrapper(other._wrapperFields.ground_truth_point_cloud);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new ObjectInformationRefStruct();
    this._wrapperFields.name = new StringWrapper();
    this._wrapperFields.ground_truth_mesh = new MeshWrapper();
    this._wrapperFields.ground_truth_point_cloud = new PointCloud2Wrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ObjectInformationWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ObjectInformationArrayWrapper();
  }
  static get ArrayType() {
    return ObjectInformationArrayWrapper;
  }
  static get refObjectArrayType() {
    return ObjectInformationRefStructArray;
  }
  static get refObjectType() {
    return ObjectInformationRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.name.freeze(own);
    }
    this._refObject.name = this._wrapperFields.name.refObject;
    this._wrapperFields.ground_truth_mesh.freeze(own);
    this._refObject.ground_truth_mesh = this._wrapperFields.ground_truth_mesh.refObject;
    this._wrapperFields.ground_truth_point_cloud.freeze(own);
    this._refObject.ground_truth_point_cloud = this._wrapperFields.ground_truth_point_cloud.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.name.data = refObject.name.data;
    this._wrapperFields.ground_truth_mesh.copyRefObject(refObject.ground_truth_mesh);
    this._wrapperFields.ground_truth_point_cloud.copyRefObject(refObject.ground_truth_point_cloud);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.name);
    MeshWrapper.freeStruct(refObject.ground_truth_mesh);
    PointCloud2Wrapper.freeStruct(refObject.ground_truth_point_cloud);
  }
  static destroyRawROS(msg) {
    ObjectInformationWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'object_recognition_msgs', subFolder: 'msg', interfaceName: 'ObjectInformation'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get name() {
    return this._wrapperFields.name.data;
  }
  set name(value) {
    this._wrapperFields.name.data = value;
  }
  get ground_truth_mesh() {
    return this._wrapperFields.ground_truth_mesh;
  }
  set ground_truth_mesh(value) {
    if (value instanceof MeshWrapper) {
      this._wrapperFields.ground_truth_mesh.copy(value);
    } else {
      this._wrapperFields.ground_truth_mesh.copy(new MeshWrapper(value));
    }
  }
  get ground_truth_point_cloud() {
    return this._wrapperFields.ground_truth_point_cloud;
  }
  set ground_truth_point_cloud(value) {
    if (value instanceof PointCloud2Wrapper) {
      this._wrapperFields.ground_truth_point_cloud.copy(value);
    } else {
      this._wrapperFields.ground_truth_point_cloud.copy(new PointCloud2Wrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new ObjectInformationRefStruct(refObject.toObject());
    this._wrapperFields.name.copyRefObject(this._refObject.name);
    this._wrapperFields.ground_truth_mesh.copyRefObject(this._refObject.ground_truth_mesh);
    this._wrapperFields.ground_truth_point_cloud.copyRefObject(this._refObject.ground_truth_point_cloud);
  }
  copy(other) {
    this._refObject = new ObjectInformationRefStruct(other._refObject.toObject());
    this._wrapperFields.name.copy(other._wrapperFields.name);
    this._wrapperFields.ground_truth_mesh.copy(other._wrapperFields.ground_truth_mesh);
    this._wrapperFields.ground_truth_point_cloud.copy(other._wrapperFields.ground_truth_point_cloud);
  }
  static get classType() {
    return ObjectInformationWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ground_truth_mesh","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"shape_msgs","type":"Mesh","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"ground_truth_point_cloud","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"sensor_msgs","type":"PointCloud2","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"object_recognition_msgs","type":"ObjectInformation","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ObjectInformation"};
  }
  hasMember(name) {
    let memberNames = ["name","ground_truth_mesh","ground_truth_point_cloud"];
    return memberNames.indexOf(name) !== -1;
  }
}
class ObjectInformationArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ObjectInformationWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ObjectInformationWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ObjectInformationArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ObjectInformationArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new ObjectInformationRefArray(size);
    this._refObject = new ObjectInformationRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ObjectInformationWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ObjectInformationArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ObjectInformationArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ObjectInformationWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ObjectInformationWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ObjectInformationArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = ObjectInformationWrapper;
