// rclnodejs/generated/object_recognition_msgs/object_recognition_msgs__msg__TableArray.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const TableWrapper = require('../../generated/object_recognition_msgs/object_recognition_msgs__msg__Table.js');
const TableArrayRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  tables: TableWrapper.refObjectArrayType,
});
const TableArrayRefArray = ArrayType(TableArrayRefStruct);
const TableArrayRefStructArray = StructType({
  data: TableArrayRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class TableArrayWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new TableArrayRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.tables = TableWrapper.createArray();
      this._wrapperFields.tables.copy(other._wrapperFields.tables);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new TableArrayRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.tables = TableWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new TableArrayWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new TableArrayArrayWrapper();
  }
  static get ArrayType() {
    return TableArrayArrayWrapper;
  }
  static get refObjectArrayType() {
    return TableArrayRefStructArray;
  }
  static get refObjectType() {
    return TableArrayRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.header.freeze(own);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.tables.fill([]);
    }
    this._wrapperFields.tables.freeze(own);
    this._refObject.tables = this._wrapperFields.tables.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.tables.copyRefObject(refObject.tables);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.tables.size != 0) {
      TableWrapper.ArrayType.freeArray(refObject.tables);
      if (!TableWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.tables, TableWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    TableArrayWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'object_recognition_msgs', subFolder: 'msg', interfaceName: 'TableArray'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get tables() {
    return this._wrapperFields.tables;
  }
  set tables(value) {
    this._wrapperFields.tables.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new TableArrayRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.tables.copyRefObject(this._refObject.tables);
  }
  copy(other) {
    this._refObject = new TableArrayRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.tables.copy(other._wrapperFields.tables);
  }
  static get classType() {
    return TableArrayWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"tables","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"object_recognition_msgs","type":"Table","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"object_recognition_msgs","type":"TableArray","stringUpperBound":null,"isPrimitiveType":false},"msgName":"TableArray"};
  }
  hasMember(name) {
    let memberNames = ["header","tables"];
    return memberNames.indexOf(name) !== -1;
  }
}
class TableArrayArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof TableArrayWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new TableArrayWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TableArrayArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TableArrayArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new TableArrayRefArray(size);
    this._refObject = new TableArrayRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new TableArrayWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof TableArrayArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "TableArrayArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      TableArrayWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return TableArrayWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return TableArrayArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = TableArrayWrapper;
