// rclnodejs/generated/ethercat_msgs/ethercat_msgs__srv__SetSdo_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const SetSdo_ResponseRefStruct = StructType({
  success: primitiveTypes.bool,
  sdo_return_message: primitiveTypes.string,
});
const SetSdo_ResponseRefArray = ArrayType(SetSdo_ResponseRefStruct);
const SetSdo_ResponseRefStructArray = StructType({
  data: SetSdo_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class SetSdo_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetSdo_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.sdo_return_message =  new StringWrapper(other._wrapperFields.sdo_return_message);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new SetSdo_ResponseRefStruct();
    this._wrapperFields.sdo_return_message = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new SetSdo_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetSdo_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return SetSdo_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetSdo_ResponseRefStructArray;
  }
  static get refObjectType() {
    return SetSdo_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.sdo_return_message.freeze(own);
    }
    this._refObject.sdo_return_message = this._wrapperFields.sdo_return_message.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.success = refObject.success;
    this._wrapperFields.sdo_return_message.data = refObject.sdo_return_message.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.sdo_return_message);
  }
  static destroyRawROS(msg) {
    SetSdo_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ethercat_msgs', subFolder: 'srv', interfaceName: 'SetSdo_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get success() {
    return this._refObject.success;
  }
  set success(value) {
    this._refObject.success = value;
  }
  get sdo_return_message() {
    return this._wrapperFields.sdo_return_message.data;
  }
  set sdo_return_message(value) {
    this._wrapperFields.sdo_return_message.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SetSdo_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.sdo_return_message.copyRefObject(this._refObject.sdo_return_message);
  }
  copy(other) {
    this._refObject = new SetSdo_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.sdo_return_message.copy(other._wrapperFields.sdo_return_message);
  }
  static get classType() {
    return SetSdo_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"success","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"sdo_return_message","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"ethercat_msgs","type":"SetSdo_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetSdo_Response"};
  }
  hasMember(name) {
    let memberNames = ["success","sdo_return_message"];
    return memberNames.indexOf(name) !== -1;
  }
}
class SetSdo_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetSdo_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetSdo_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetSdo_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetSdo_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new SetSdo_ResponseRefArray(size);
    this._refObject = new SetSdo_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetSdo_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetSdo_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetSdo_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetSdo_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetSdo_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetSdo_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = SetSdo_ResponseWrapper;
