// rclnodejs/generated/control_msgs/control_msgs__srv__QueryTrajectoryState_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const Float64Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float64.js');
const QueryTrajectoryState_ResponseRefStruct = StructType({
  success: primitiveTypes.bool,
  message: primitiveTypes.string,
  name: StringWrapper.refObjectArrayType,
  position: Float64Wrapper.refObjectArrayType,
  velocity: Float64Wrapper.refObjectArrayType,
  acceleration: Float64Wrapper.refObjectArrayType,
});
const QueryTrajectoryState_ResponseRefArray = ArrayType(QueryTrajectoryState_ResponseRefStruct);
const QueryTrajectoryState_ResponseRefStructArray = StructType({
  data: QueryTrajectoryState_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class QueryTrajectoryState_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new QueryTrajectoryState_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.message =  new StringWrapper(other._wrapperFields.message);
      this._wrapperFields.name = StringWrapper.createArray();
      this._wrapperFields.name.copy(other._wrapperFields.name);
      this.name = other.name;
      this._wrapperFields.position = Float64Wrapper.createArray();
      this._wrapperFields.position.copy(other._wrapperFields.position);
      this._wrapperFields.velocity = Float64Wrapper.createArray();
      this._wrapperFields.velocity.copy(other._wrapperFields.velocity);
      this._wrapperFields.acceleration = Float64Wrapper.createArray();
      this._wrapperFields.acceleration.copy(other._wrapperFields.acceleration);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new QueryTrajectoryState_ResponseRefStruct();
    this._wrapperFields.message = new StringWrapper();
    this._wrapperFields.name = StringWrapper.createArray();
    this._nameArray = [];
    this._wrapperFields.position = Float64Wrapper.createArray();
    this._wrapperFields.velocity = Float64Wrapper.createArray();
    this._wrapperFields.acceleration = Float64Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new QueryTrajectoryState_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new QueryTrajectoryState_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return QueryTrajectoryState_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return QueryTrajectoryState_ResponseRefStructArray;
  }
  static get refObjectType() {
    return QueryTrajectoryState_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.message.freeze(own);
    }
    this._refObject.message = this._wrapperFields.message.refObject;
    if (own) {
      this._wrapperFields.name.fill([]);
    } else {
      this._wrapperFields.name.fill(this._nameArray);
    }
    this._wrapperFields.name.freeze(own);
    this._refObject.name = this._wrapperFields.name.refObject;
    if (own) {
      this._wrapperFields.position.fill(Float64Array.from([]));
    }
    this._wrapperFields.position.freeze(own);
    this._refObject.position = this._wrapperFields.position.refObject;
    if (own) {
      this._wrapperFields.velocity.fill(Float64Array.from([]));
    }
    this._wrapperFields.velocity.freeze(own);
    this._refObject.velocity = this._wrapperFields.velocity.refObject;
    if (own) {
      this._wrapperFields.acceleration.fill(Float64Array.from([]));
    }
    this._wrapperFields.acceleration.freeze(own);
    this._refObject.acceleration = this._wrapperFields.acceleration.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.success = refObject.success;
    this._wrapperFields.message.data = refObject.message.data;
    refObject.name.data.length = refObject.name.size;
    for (let index = 0; index < refObject.name.size; index++) {
      this._nameArray[index] = refObject.name.data[index].data;
    }
    this._wrapperFields.position.copyRefObject(refObject.position);
    this._wrapperFields.velocity.copyRefObject(refObject.velocity);
    this._wrapperFields.acceleration.copyRefObject(refObject.acceleration);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.message);
    if (refObject.name.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.name);
      if (!StringWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.name, StringWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.position.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.position);
      if (!Float64Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.position, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.velocity.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.velocity);
      if (!Float64Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.velocity, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.acceleration.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.acceleration);
      if (!Float64Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.acceleration, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    QueryTrajectoryState_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'control_msgs', subFolder: 'srv', interfaceName: 'QueryTrajectoryState_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get success() {
    return this._refObject.success;
  }
  set success(value) {
    this._refObject.success = value;
  }
  get message() {
    return this._wrapperFields.message.data;
  }
  set message(value) {
    this._wrapperFields.message.data = value;
  }
  get name() {
    return this._nameArray;
  }
  set name(value) {
    this._nameArray = value;
  }
  get position() {
    return this._wrapperFields['position'].data;
  }
  set position(value) {
    this._wrapperFields['position'].fill(value);
  }
  get velocity() {
    return this._wrapperFields['velocity'].data;
  }
  set velocity(value) {
    this._wrapperFields['velocity'].fill(value);
  }
  get acceleration() {
    return this._wrapperFields['acceleration'].data;
  }
  set acceleration(value) {
    this._wrapperFields['acceleration'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new QueryTrajectoryState_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.message.copyRefObject(this._refObject.message);
    refObject.name.data.length = refObject.name.size;
    for (let index = 0; index < refObject.name.size; index++) {
      this._nameArray[index] = refObject.name.data[index].data;
    }
    this._wrapperFields.position.copyRefObject(this._refObject.position);
    this._wrapperFields.velocity.copyRefObject(this._refObject.velocity);
    this._wrapperFields.acceleration.copyRefObject(this._refObject.acceleration);
  }
  copy(other) {
    this._refObject = new QueryTrajectoryState_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.message.copy(other._wrapperFields.message);
    this._nameArray = other._nameArray.slice();
    this._wrapperFields.position.copy(other._wrapperFields.position);
    this._wrapperFields.velocity.copy(other._wrapperFields.velocity);
    this._wrapperFields.acceleration.copy(other._wrapperFields.acceleration);
  }
  static get classType() {
    return QueryTrajectoryState_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"success","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"message","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"name","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"position","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"velocity","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"acceleration","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"control_msgs","type":"QueryTrajectoryState_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"QueryTrajectoryState_Response"};
  }
  hasMember(name) {
    let memberNames = ["success","message","name","position","velocity","acceleration"];
    return memberNames.indexOf(name) !== -1;
  }
}
class QueryTrajectoryState_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof QueryTrajectoryState_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new QueryTrajectoryState_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to QueryTrajectoryState_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to QueryTrajectoryState_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new QueryTrajectoryState_ResponseRefArray(size);
    this._refObject = new QueryTrajectoryState_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new QueryTrajectoryState_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof QueryTrajectoryState_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "QueryTrajectoryState_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      QueryTrajectoryState_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return QueryTrajectoryState_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return QueryTrajectoryState_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = QueryTrajectoryState_ResponseWrapper;
