// rclnodejs/generated/control_msgs/control_msgs__msg__MecanumDriveControllerState.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const TwistWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Twist.js');
const MecanumDriveControllerStateRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  front_left_wheel_velocity: primitiveTypes.float64,
  back_left_wheel_velocity: primitiveTypes.float64,
  back_right_wheel_velocity: primitiveTypes.float64,
  front_right_wheel_velocity: primitiveTypes.float64,
  reference_velocity: TwistWrapper.refObjectType,
});
const MecanumDriveControllerStateRefArray = ArrayType(MecanumDriveControllerStateRefStruct);
const MecanumDriveControllerStateRefStructArray = StructType({
  data: MecanumDriveControllerStateRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class MecanumDriveControllerStateWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new MecanumDriveControllerStateRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.reference_velocity =  new TwistWrapper(other._wrapperFields.reference_velocity);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new MecanumDriveControllerStateRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.reference_velocity = new TwistWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new MecanumDriveControllerStateWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MecanumDriveControllerStateArrayWrapper();
  }
  static get ArrayType() {
    return MecanumDriveControllerStateArrayWrapper;
  }
  static get refObjectArrayType() {
    return MecanumDriveControllerStateRefStructArray;
  }
  static get refObjectType() {
    return MecanumDriveControllerStateRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.header.freeze(own);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.reference_velocity.freeze(own);
    this._refObject.reference_velocity = this._wrapperFields.reference_velocity.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._refObject.front_left_wheel_velocity = refObject.front_left_wheel_velocity;
    this._refObject.back_left_wheel_velocity = refObject.back_left_wheel_velocity;
    this._refObject.back_right_wheel_velocity = refObject.back_right_wheel_velocity;
    this._refObject.front_right_wheel_velocity = refObject.front_right_wheel_velocity;
    this._wrapperFields.reference_velocity.copyRefObject(refObject.reference_velocity);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    TwistWrapper.freeStruct(refObject.reference_velocity);
  }
  static destroyRawROS(msg) {
    MecanumDriveControllerStateWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'control_msgs', subFolder: 'msg', interfaceName: 'MecanumDriveControllerState'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get front_left_wheel_velocity() {
    return this._refObject.front_left_wheel_velocity;
  }
  set front_left_wheel_velocity(value) {
    this._refObject.front_left_wheel_velocity = value;
  }
  get back_left_wheel_velocity() {
    return this._refObject.back_left_wheel_velocity;
  }
  set back_left_wheel_velocity(value) {
    this._refObject.back_left_wheel_velocity = value;
  }
  get back_right_wheel_velocity() {
    return this._refObject.back_right_wheel_velocity;
  }
  set back_right_wheel_velocity(value) {
    this._refObject.back_right_wheel_velocity = value;
  }
  get front_right_wheel_velocity() {
    return this._refObject.front_right_wheel_velocity;
  }
  set front_right_wheel_velocity(value) {
    this._refObject.front_right_wheel_velocity = value;
  }
  get reference_velocity() {
    return this._wrapperFields.reference_velocity;
  }
  set reference_velocity(value) {
    if (value instanceof TwistWrapper) {
      this._wrapperFields.reference_velocity.copy(value);
    } else {
      this._wrapperFields.reference_velocity.copy(new TwistWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new MecanumDriveControllerStateRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.reference_velocity.copyRefObject(this._refObject.reference_velocity);
  }
  copy(other) {
    this._refObject = new MecanumDriveControllerStateRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.reference_velocity.copy(other._wrapperFields.reference_velocity);
  }
  static get classType() {
    return MecanumDriveControllerStateWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"front_left_wheel_velocity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"back_left_wheel_velocity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"back_right_wheel_velocity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"front_right_wheel_velocity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reference_velocity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Twist","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"control_msgs","type":"MecanumDriveControllerState","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MecanumDriveControllerState"};
  }
  hasMember(name) {
    let memberNames = ["header","front_left_wheel_velocity","back_left_wheel_velocity","back_right_wheel_velocity","front_right_wheel_velocity","reference_velocity"];
    return memberNames.indexOf(name) !== -1;
  }
}
class MecanumDriveControllerStateArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MecanumDriveControllerStateWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MecanumDriveControllerStateWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MecanumDriveControllerStateArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MecanumDriveControllerStateArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new MecanumDriveControllerStateRefArray(size);
    this._refObject = new MecanumDriveControllerStateRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MecanumDriveControllerStateWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MecanumDriveControllerStateArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MecanumDriveControllerStateArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MecanumDriveControllerStateWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MecanumDriveControllerStateWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MecanumDriveControllerStateArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = MecanumDriveControllerStateWrapper;
