// rclnodejs/generated/control_msgs/control_msgs__action__SingleJointPosition_Result.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const SingleJointPosition_ResultRefStruct = StructType({
  _dummy: primitiveTypes.bool,
});
const SingleJointPosition_ResultRefArray = ArrayType(SingleJointPosition_ResultRefStruct);
const SingleJointPosition_ResultRefStructArray = StructType({
  data: SingleJointPosition_ResultRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class SingleJointPosition_ResultWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new SingleJointPosition_ResultRefStruct(other._refObject.toObject());
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new SingleJointPosition_ResultRefStruct();
  }
  static createFromRefObject(refObject) {
    let self = new SingleJointPosition_ResultWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SingleJointPosition_ResultArrayWrapper();
  }
  static get ArrayType() {
    return SingleJointPosition_ResultArrayWrapper;
  }
  static get refObjectArrayType() {
    return SingleJointPosition_ResultRefStructArray;
  }
  static get refObjectType() {
    return SingleJointPosition_ResultRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject._dummy = refObject._dummy;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    SingleJointPosition_ResultWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'control_msgs', subFolder: 'action', interfaceName: 'SingleJointPosition_Result'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get _dummy() {
    return this._refObject._dummy;
  }
  set _dummy(value) {
    this._refObject._dummy = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SingleJointPosition_ResultRefStruct(refObject.toObject());
  }
  copy(other) {
    this._refObject = new SingleJointPosition_ResultRefStruct(other._refObject.toObject());
  }
  static get classType() {
    return SingleJointPosition_ResultWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[],"baseType":{"pkgName":"control_msgs","type":"SingleJointPosition_Result","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SingleJointPosition_Result"};
  }
  hasMember(name) {
    let memberNames = ["_dummy"];
    return memberNames.indexOf(name) !== -1;
  }
}
class SingleJointPosition_ResultArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SingleJointPosition_ResultWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SingleJointPosition_ResultWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SingleJointPosition_ResultArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SingleJointPosition_ResultArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new SingleJointPosition_ResultRefArray(size);
    this._refObject = new SingleJointPosition_ResultRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SingleJointPosition_ResultWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SingleJointPosition_ResultArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SingleJointPosition_ResultArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SingleJointPosition_ResultWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SingleJointPosition_ResultWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SingleJointPosition_ResultArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = SingleJointPosition_ResultWrapper;
