// rclnodejs/generated/control_msgs/control_msgs__msg__MultiDOFCommand.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const Float64Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float64.js');
const MultiDOFCommandRefStruct = StructType({
  dof_names: StringWrapper.refObjectArrayType,
  values: Float64Wrapper.refObjectArrayType,
  values_dot: Float64Wrapper.refObjectArrayType,
});
const MultiDOFCommandRefArray = ArrayType(MultiDOFCommandRefStruct);
const MultiDOFCommandRefStructArray = StructType({
  data: MultiDOFCommandRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class MultiDOFCommandWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new MultiDOFCommandRefStruct(other._refObject.toObject());
      this._wrapperFields.dof_names = StringWrapper.createArray();
      this._wrapperFields.dof_names.copy(other._wrapperFields.dof_names);
      this.dof_names = other.dof_names;
      this._wrapperFields.values = Float64Wrapper.createArray();
      this._wrapperFields.values.copy(other._wrapperFields.values);
      this._wrapperFields.values_dot = Float64Wrapper.createArray();
      this._wrapperFields.values_dot.copy(other._wrapperFields.values_dot);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new MultiDOFCommandRefStruct();
    this._wrapperFields.dof_names = StringWrapper.createArray();
    this._dof_namesArray = [];
    this._wrapperFields.values = Float64Wrapper.createArray();
    this._wrapperFields.values_dot = Float64Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new MultiDOFCommandWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MultiDOFCommandArrayWrapper();
  }
  static get ArrayType() {
    return MultiDOFCommandArrayWrapper;
  }
  static get refObjectArrayType() {
    return MultiDOFCommandRefStructArray;
  }
  static get refObjectType() {
    return MultiDOFCommandRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.dof_names.fill([]);
    } else {
      this._wrapperFields.dof_names.fill(this._dof_namesArray);
    }
    this._wrapperFields.dof_names.freeze(own);
    this._refObject.dof_names = this._wrapperFields.dof_names.refObject;
    if (own) {
      this._wrapperFields.values.fill(Float64Array.from([]));
    }
    this._wrapperFields.values.freeze(own);
    this._refObject.values = this._wrapperFields.values.refObject;
    if (own) {
      this._wrapperFields.values_dot.fill(Float64Array.from([]));
    }
    this._wrapperFields.values_dot.freeze(own);
    this._refObject.values_dot = this._wrapperFields.values_dot.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    refObject.dof_names.data.length = refObject.dof_names.size;
    for (let index = 0; index < refObject.dof_names.size; index++) {
      this._dof_namesArray[index] = refObject.dof_names.data[index].data;
    }
    this._wrapperFields.values.copyRefObject(refObject.values);
    this._wrapperFields.values_dot.copyRefObject(refObject.values_dot);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.dof_names.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.dof_names);
      if (!StringWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.dof_names, StringWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.values.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.values);
      if (!Float64Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.values, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.values_dot.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.values_dot);
      if (!Float64Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.values_dot, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    MultiDOFCommandWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'control_msgs', subFolder: 'msg', interfaceName: 'MultiDOFCommand'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get dof_names() {
    return this._dof_namesArray;
  }
  set dof_names(value) {
    this._dof_namesArray = value;
  }
  get values() {
    return this._wrapperFields['values'].data;
  }
  set values(value) {
    this._wrapperFields['values'].fill(value);
  }
  get values_dot() {
    return this._wrapperFields['values_dot'].data;
  }
  set values_dot(value) {
    this._wrapperFields['values_dot'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new MultiDOFCommandRefStruct(refObject.toObject());
    refObject.dof_names.data.length = refObject.dof_names.size;
    for (let index = 0; index < refObject.dof_names.size; index++) {
      this._dof_namesArray[index] = refObject.dof_names.data[index].data;
    }
    this._wrapperFields.values.copyRefObject(this._refObject.values);
    this._wrapperFields.values_dot.copyRefObject(this._refObject.values_dot);
  }
  copy(other) {
    this._refObject = new MultiDOFCommandRefStruct(other._refObject.toObject());
    this._dof_namesArray = other._dof_namesArray.slice();
    this._wrapperFields.values.copy(other._wrapperFields.values);
    this._wrapperFields.values_dot.copy(other._wrapperFields.values_dot);
  }
  static get classType() {
    return MultiDOFCommandWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"dof_names","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"values","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"values_dot","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"control_msgs","type":"MultiDOFCommand","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MultiDOFCommand"};
  }
  hasMember(name) {
    let memberNames = ["dof_names","values","values_dot"];
    return memberNames.indexOf(name) !== -1;
  }
}
class MultiDOFCommandArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MultiDOFCommandWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MultiDOFCommandWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MultiDOFCommandArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MultiDOFCommandArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new MultiDOFCommandRefArray(size);
    this._refObject = new MultiDOFCommandRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MultiDOFCommandWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MultiDOFCommandArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MultiDOFCommandArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MultiDOFCommandWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MultiDOFCommandWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MultiDOFCommandArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = MultiDOFCommandWrapper;
