// rclnodejs/generated/control_msgs/control_msgs__msg__GripperCommand.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const GripperCommandRefStruct = StructType({
  position: primitiveTypes.float64,
  max_effort: primitiveTypes.float64,
});
const GripperCommandRefArray = ArrayType(GripperCommandRefStruct);
const GripperCommandRefStructArray = StructType({
  data: GripperCommandRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class GripperCommandWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new GripperCommandRefStruct(other._refObject.toObject());
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new GripperCommandRefStruct();
  }
  static createFromRefObject(refObject) {
    let self = new GripperCommandWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GripperCommandArrayWrapper();
  }
  static get ArrayType() {
    return GripperCommandArrayWrapper;
  }
  static get refObjectArrayType() {
    return GripperCommandRefStructArray;
  }
  static get refObjectType() {
    return GripperCommandRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.position = refObject.position;
    this._refObject.max_effort = refObject.max_effort;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    GripperCommandWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'control_msgs', subFolder: 'msg', interfaceName: 'GripperCommand'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get position() {
    return this._refObject.position;
  }
  set position(value) {
    this._refObject.position = value;
  }
  get max_effort() {
    return this._refObject.max_effort;
  }
  set max_effort(value) {
    this._refObject.max_effort = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GripperCommandRefStruct(refObject.toObject());
  }
  copy(other) {
    this._refObject = new GripperCommandRefStruct(other._refObject.toObject());
  }
  static get classType() {
    return GripperCommandWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"position","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"max_effort","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"control_msgs","type":"GripperCommand","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GripperCommand"};
  }
  hasMember(name) {
    let memberNames = ["position","max_effort"];
    return memberNames.indexOf(name) !== -1;
  }
}
class GripperCommandArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GripperCommandWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GripperCommandWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GripperCommandArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GripperCommandArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new GripperCommandRefArray(size);
    this._refObject = new GripperCommandRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GripperCommandWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GripperCommandArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GripperCommandArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GripperCommandWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GripperCommandWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GripperCommandArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = GripperCommandWrapper;
