// rclnodejs/generated/control_msgs/control_msgs__action__ParallelGripperCommand_Result.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const JointStateWrapper = require('../../generated/sensor_msgs/sensor_msgs__msg__JointState.js');
const ParallelGripperCommand_ResultRefStruct = StructType({
  state: JointStateWrapper.refObjectType,
  stalled: primitiveTypes.bool,
  reached_goal: primitiveTypes.bool,
});
const ParallelGripperCommand_ResultRefArray = ArrayType(ParallelGripperCommand_ResultRefStruct);
const ParallelGripperCommand_ResultRefStructArray = StructType({
  data: ParallelGripperCommand_ResultRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class ParallelGripperCommand_ResultWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new ParallelGripperCommand_ResultRefStruct(other._refObject.toObject());
      this._wrapperFields.state =  new JointStateWrapper(other._wrapperFields.state);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new ParallelGripperCommand_ResultRefStruct();
    this._wrapperFields.state = new JointStateWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ParallelGripperCommand_ResultWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ParallelGripperCommand_ResultArrayWrapper();
  }
  static get ArrayType() {
    return ParallelGripperCommand_ResultArrayWrapper;
  }
  static get refObjectArrayType() {
    return ParallelGripperCommand_ResultRefStructArray;
  }
  static get refObjectType() {
    return ParallelGripperCommand_ResultRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.state.freeze(own);
    this._refObject.state = this._wrapperFields.state.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.state.copyRefObject(refObject.state);
    this._refObject.stalled = refObject.stalled;
    this._refObject.reached_goal = refObject.reached_goal;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    JointStateWrapper.freeStruct(refObject.state);
  }
  static destroyRawROS(msg) {
    ParallelGripperCommand_ResultWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'control_msgs', subFolder: 'action', interfaceName: 'ParallelGripperCommand_Result'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get state() {
    return this._wrapperFields.state;
  }
  set state(value) {
    if (value instanceof JointStateWrapper) {
      this._wrapperFields.state.copy(value);
    } else {
      this._wrapperFields.state.copy(new JointStateWrapper(value));
    }
  }
  get stalled() {
    return this._refObject.stalled;
  }
  set stalled(value) {
    this._refObject.stalled = value;
  }
  get reached_goal() {
    return this._refObject.reached_goal;
  }
  set reached_goal(value) {
    this._refObject.reached_goal = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ParallelGripperCommand_ResultRefStruct(refObject.toObject());
    this._wrapperFields.state.copyRefObject(this._refObject.state);
  }
  copy(other) {
    this._refObject = new ParallelGripperCommand_ResultRefStruct(other._refObject.toObject());
    this._wrapperFields.state.copy(other._wrapperFields.state);
  }
  static get classType() {
    return ParallelGripperCommand_ResultWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"sensor_msgs","type":"JointState","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"stalled","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reached_goal","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"control_msgs","type":"ParallelGripperCommand_Result","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ParallelGripperCommand_Result"};
  }
  hasMember(name) {
    let memberNames = ["state","stalled","reached_goal"];
    return memberNames.indexOf(name) !== -1;
  }
}
class ParallelGripperCommand_ResultArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ParallelGripperCommand_ResultWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ParallelGripperCommand_ResultWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ParallelGripperCommand_ResultArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ParallelGripperCommand_ResultArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new ParallelGripperCommand_ResultRefArray(size);
    this._refObject = new ParallelGripperCommand_ResultRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ParallelGripperCommand_ResultWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ParallelGripperCommand_ResultArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ParallelGripperCommand_ResultArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ParallelGripperCommand_ResultWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ParallelGripperCommand_ResultWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ParallelGripperCommand_ResultArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = ParallelGripperCommand_ResultWrapper;
