// rclnodejs/generated/control_msgs/control_msgs__msg__JointComponentTolerance.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const JointComponentToleranceRefStruct = StructType({
  joint_name: primitiveTypes.string,
  component: primitiveTypes.uint16,
  position: primitiveTypes.float64,
  velocity: primitiveTypes.float64,
  acceleration: primitiveTypes.float64,
});
const JointComponentToleranceRefArray = ArrayType(JointComponentToleranceRefStruct);
const JointComponentToleranceRefStructArray = StructType({
  data: JointComponentToleranceRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class JointComponentToleranceWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new JointComponentToleranceRefStruct(other._refObject.toObject());
      this._wrapperFields.joint_name =  new StringWrapper(other._wrapperFields.joint_name);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new JointComponentToleranceRefStruct();
    this._wrapperFields.joint_name = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new JointComponentToleranceWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new JointComponentToleranceArrayWrapper();
  }
  static get ArrayType() {
    return JointComponentToleranceArrayWrapper;
  }
  static get refObjectArrayType() {
    return JointComponentToleranceRefStructArray;
  }
  static get refObjectType() {
    return JointComponentToleranceRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.joint_name.freeze(own);
    }
    this._refObject.joint_name = this._wrapperFields.joint_name.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.joint_name.data = refObject.joint_name.data;
    this._refObject.component = refObject.component;
    this._refObject.position = refObject.position;
    this._refObject.velocity = refObject.velocity;
    this._refObject.acceleration = refObject.acceleration;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.joint_name);
  }
  static destroyRawROS(msg) {
    JointComponentToleranceWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'control_msgs', subFolder: 'msg', interfaceName: 'JointComponentTolerance'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get joint_name() {
    return this._wrapperFields.joint_name.data;
  }
  set joint_name(value) {
    this._wrapperFields.joint_name.data = value;
  }
  get component() {
    return this._refObject.component;
  }
  set component(value) {
    this._refObject.component = value;
  }
  get position() {
    return this._refObject.position;
  }
  set position(value) {
    this._refObject.position = value;
  }
  get velocity() {
    return this._refObject.velocity;
  }
  set velocity(value) {
    this._refObject.velocity = value;
  }
  get acceleration() {
    return this._refObject.acceleration;
  }
  set acceleration(value) {
    this._refObject.acceleration = value;
  }
  copyRefObject(refObject) {
    this._refObject = new JointComponentToleranceRefStruct(refObject.toObject());
    this._wrapperFields.joint_name.copyRefObject(this._refObject.joint_name);
  }
  copy(other) {
    this._refObject = new JointComponentToleranceRefStruct(other._refObject.toObject());
    this._wrapperFields.joint_name.copy(other._wrapperFields.joint_name);
  }
  static get classType() {
    return JointComponentToleranceWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint16","name":"X_AXIS","value":1},{"type":"uint16","name":"Y_AXIS","value":2},{"type":"uint16","name":"Z_AXIS","value":3},{"type":"uint16","name":"TRANSLATION","value":4},{"type":"uint16","name":"ROTATION","value":5}],"fields":[{"name":"joint_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"component","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"position","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"velocity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"acceleration","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"control_msgs","type":"JointComponentTolerance","stringUpperBound":null,"isPrimitiveType":false},"msgName":"JointComponentTolerance"};
  }
  hasMember(name) {
    let memberNames = ["joint_name","component","position","velocity","acceleration"];
    return memberNames.indexOf(name) !== -1;
  }
}
class JointComponentToleranceArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof JointComponentToleranceWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new JointComponentToleranceWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to JointComponentToleranceArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to JointComponentToleranceArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new JointComponentToleranceRefArray(size);
    this._refObject = new JointComponentToleranceRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new JointComponentToleranceWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof JointComponentToleranceArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "JointComponentToleranceArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      JointComponentToleranceWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return JointComponentToleranceWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return JointComponentToleranceArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
Object.defineProperty(JointComponentToleranceWrapper, "X_AXIS", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(JointComponentToleranceWrapper, "Y_AXIS", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(JointComponentToleranceWrapper, "Z_AXIS", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(JointComponentToleranceWrapper, "TRANSLATION", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(JointComponentToleranceWrapper, "ROTATION", {value: 5, writable: false, enumerable: true, configurable: true});
module.exports = JointComponentToleranceWrapper;
