// rclnodejs/generated/nextup_joint_interfaces/nextup_joint_interfaces__msg__NextupDigitalInputs.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const BoolWrapper = require('../../generated/std_msgs/std_msgs__msg__Bool.js');
const UInt32Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt32.js');
const NextupDigitalInputsRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  name: StringWrapper.refObjectArrayType,
  di1: BoolWrapper.refObjectArrayType,
  di2: BoolWrapper.refObjectArrayType,
  di3: BoolWrapper.refObjectArrayType,
  di4: BoolWrapper.refObjectArrayType,
  di5: BoolWrapper.refObjectArrayType,
  sto1: BoolWrapper.refObjectArrayType,
  sto2: BoolWrapper.refObjectArrayType,
  edm: BoolWrapper.refObjectArrayType,
  digital_input_raw: UInt32Wrapper.refObjectArrayType,
});
const NextupDigitalInputsRefArray = ArrayType(NextupDigitalInputsRefStruct);
const NextupDigitalInputsRefStructArray = StructType({
  data: NextupDigitalInputsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class NextupDigitalInputsWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new NextupDigitalInputsRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.name = StringWrapper.createArray();
      this._wrapperFields.name.copy(other._wrapperFields.name);
      this.name = other.name;
      this._wrapperFields.di1 = BoolWrapper.createArray();
      this._wrapperFields.di1.copy(other._wrapperFields.di1);
      this.di1 = other.di1;
      this._wrapperFields.di2 = BoolWrapper.createArray();
      this._wrapperFields.di2.copy(other._wrapperFields.di2);
      this.di2 = other.di2;
      this._wrapperFields.di3 = BoolWrapper.createArray();
      this._wrapperFields.di3.copy(other._wrapperFields.di3);
      this.di3 = other.di3;
      this._wrapperFields.di4 = BoolWrapper.createArray();
      this._wrapperFields.di4.copy(other._wrapperFields.di4);
      this.di4 = other.di4;
      this._wrapperFields.di5 = BoolWrapper.createArray();
      this._wrapperFields.di5.copy(other._wrapperFields.di5);
      this.di5 = other.di5;
      this._wrapperFields.sto1 = BoolWrapper.createArray();
      this._wrapperFields.sto1.copy(other._wrapperFields.sto1);
      this.sto1 = other.sto1;
      this._wrapperFields.sto2 = BoolWrapper.createArray();
      this._wrapperFields.sto2.copy(other._wrapperFields.sto2);
      this.sto2 = other.sto2;
      this._wrapperFields.edm = BoolWrapper.createArray();
      this._wrapperFields.edm.copy(other._wrapperFields.edm);
      this.edm = other.edm;
      this._wrapperFields.digital_input_raw = UInt32Wrapper.createArray();
      this._wrapperFields.digital_input_raw.copy(other._wrapperFields.digital_input_raw);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new NextupDigitalInputsRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.name = StringWrapper.createArray();
    this._nameArray = [];
    this._wrapperFields.di1 = BoolWrapper.createArray();
    this._di1Array = [];
    this._wrapperFields.di2 = BoolWrapper.createArray();
    this._di2Array = [];
    this._wrapperFields.di3 = BoolWrapper.createArray();
    this._di3Array = [];
    this._wrapperFields.di4 = BoolWrapper.createArray();
    this._di4Array = [];
    this._wrapperFields.di5 = BoolWrapper.createArray();
    this._di5Array = [];
    this._wrapperFields.sto1 = BoolWrapper.createArray();
    this._sto1Array = [];
    this._wrapperFields.sto2 = BoolWrapper.createArray();
    this._sto2Array = [];
    this._wrapperFields.edm = BoolWrapper.createArray();
    this._edmArray = [];
    this._wrapperFields.digital_input_raw = UInt32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new NextupDigitalInputsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new NextupDigitalInputsArrayWrapper();
  }
  static get ArrayType() {
    return NextupDigitalInputsArrayWrapper;
  }
  static get refObjectArrayType() {
    return NextupDigitalInputsRefStructArray;
  }
  static get refObjectType() {
    return NextupDigitalInputsRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.header.freeze(own);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.name.fill([]);
    } else {
      this._wrapperFields.name.fill(this._nameArray);
    }
    this._wrapperFields.name.freeze(own);
    this._refObject.name = this._wrapperFields.name.refObject;
    if (own) {
      this._wrapperFields.di1.fill([]);
    } else {
      this._wrapperFields.di1.fill(this._di1Array);
    }
    this._wrapperFields.di1.freeze(own);
    this._refObject.di1 = this._wrapperFields.di1.refObject;
    if (own) {
      this._wrapperFields.di2.fill([]);
    } else {
      this._wrapperFields.di2.fill(this._di2Array);
    }
    this._wrapperFields.di2.freeze(own);
    this._refObject.di2 = this._wrapperFields.di2.refObject;
    if (own) {
      this._wrapperFields.di3.fill([]);
    } else {
      this._wrapperFields.di3.fill(this._di3Array);
    }
    this._wrapperFields.di3.freeze(own);
    this._refObject.di3 = this._wrapperFields.di3.refObject;
    if (own) {
      this._wrapperFields.di4.fill([]);
    } else {
      this._wrapperFields.di4.fill(this._di4Array);
    }
    this._wrapperFields.di4.freeze(own);
    this._refObject.di4 = this._wrapperFields.di4.refObject;
    if (own) {
      this._wrapperFields.di5.fill([]);
    } else {
      this._wrapperFields.di5.fill(this._di5Array);
    }
    this._wrapperFields.di5.freeze(own);
    this._refObject.di5 = this._wrapperFields.di5.refObject;
    if (own) {
      this._wrapperFields.sto1.fill([]);
    } else {
      this._wrapperFields.sto1.fill(this._sto1Array);
    }
    this._wrapperFields.sto1.freeze(own);
    this._refObject.sto1 = this._wrapperFields.sto1.refObject;
    if (own) {
      this._wrapperFields.sto2.fill([]);
    } else {
      this._wrapperFields.sto2.fill(this._sto2Array);
    }
    this._wrapperFields.sto2.freeze(own);
    this._refObject.sto2 = this._wrapperFields.sto2.refObject;
    if (own) {
      this._wrapperFields.edm.fill([]);
    } else {
      this._wrapperFields.edm.fill(this._edmArray);
    }
    this._wrapperFields.edm.freeze(own);
    this._refObject.edm = this._wrapperFields.edm.refObject;
    if (own) {
      this._wrapperFields.digital_input_raw.fill(Uint32Array.from([]));
    }
    this._wrapperFields.digital_input_raw.freeze(own);
    this._refObject.digital_input_raw = this._wrapperFields.digital_input_raw.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    refObject.name.data.length = refObject.name.size;
    for (let index = 0; index < refObject.name.size; index++) {
      this._nameArray[index] = refObject.name.data[index].data;
    }
    refObject.di1.data.length = refObject.di1.size;
    for (let index = 0; index < refObject.di1.size; index++) {
      this._di1Array[index] = refObject.di1.data[index].data;
    }
    refObject.di2.data.length = refObject.di2.size;
    for (let index = 0; index < refObject.di2.size; index++) {
      this._di2Array[index] = refObject.di2.data[index].data;
    }
    refObject.di3.data.length = refObject.di3.size;
    for (let index = 0; index < refObject.di3.size; index++) {
      this._di3Array[index] = refObject.di3.data[index].data;
    }
    refObject.di4.data.length = refObject.di4.size;
    for (let index = 0; index < refObject.di4.size; index++) {
      this._di4Array[index] = refObject.di4.data[index].data;
    }
    refObject.di5.data.length = refObject.di5.size;
    for (let index = 0; index < refObject.di5.size; index++) {
      this._di5Array[index] = refObject.di5.data[index].data;
    }
    refObject.sto1.data.length = refObject.sto1.size;
    for (let index = 0; index < refObject.sto1.size; index++) {
      this._sto1Array[index] = refObject.sto1.data[index].data;
    }
    refObject.sto2.data.length = refObject.sto2.size;
    for (let index = 0; index < refObject.sto2.size; index++) {
      this._sto2Array[index] = refObject.sto2.data[index].data;
    }
    refObject.edm.data.length = refObject.edm.size;
    for (let index = 0; index < refObject.edm.size; index++) {
      this._edmArray[index] = refObject.edm.data[index].data;
    }
    this._wrapperFields.digital_input_raw.copyRefObject(refObject.digital_input_raw);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.name.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.name);
      if (!StringWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.name, StringWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.di1.size != 0) {
      BoolWrapper.ArrayType.freeArray(refObject.di1);
      if (!BoolWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.di1, BoolWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.di2.size != 0) {
      BoolWrapper.ArrayType.freeArray(refObject.di2);
      if (!BoolWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.di2, BoolWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.di3.size != 0) {
      BoolWrapper.ArrayType.freeArray(refObject.di3);
      if (!BoolWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.di3, BoolWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.di4.size != 0) {
      BoolWrapper.ArrayType.freeArray(refObject.di4);
      if (!BoolWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.di4, BoolWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.di5.size != 0) {
      BoolWrapper.ArrayType.freeArray(refObject.di5);
      if (!BoolWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.di5, BoolWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.sto1.size != 0) {
      BoolWrapper.ArrayType.freeArray(refObject.sto1);
      if (!BoolWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.sto1, BoolWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.sto2.size != 0) {
      BoolWrapper.ArrayType.freeArray(refObject.sto2);
      if (!BoolWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.sto2, BoolWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.edm.size != 0) {
      BoolWrapper.ArrayType.freeArray(refObject.edm);
      if (!BoolWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.edm, BoolWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.digital_input_raw.size != 0) {
      UInt32Wrapper.ArrayType.freeArray(refObject.digital_input_raw);
      if (!UInt32Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.digital_input_raw, UInt32Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    NextupDigitalInputsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nextup_joint_interfaces', subFolder: 'msg', interfaceName: 'NextupDigitalInputs'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get name() {
    return this._nameArray;
  }
  set name(value) {
    this._nameArray = value;
  }
  get di1() {
    return this._di1Array;
  }
  set di1(value) {
    this._di1Array = value;
  }
  get di2() {
    return this._di2Array;
  }
  set di2(value) {
    this._di2Array = value;
  }
  get di3() {
    return this._di3Array;
  }
  set di3(value) {
    this._di3Array = value;
  }
  get di4() {
    return this._di4Array;
  }
  set di4(value) {
    this._di4Array = value;
  }
  get di5() {
    return this._di5Array;
  }
  set di5(value) {
    this._di5Array = value;
  }
  get sto1() {
    return this._sto1Array;
  }
  set sto1(value) {
    this._sto1Array = value;
  }
  get sto2() {
    return this._sto2Array;
  }
  set sto2(value) {
    this._sto2Array = value;
  }
  get edm() {
    return this._edmArray;
  }
  set edm(value) {
    this._edmArray = value;
  }
  get digital_input_raw() {
    return this._wrapperFields['digital_input_raw'].data;
  }
  set digital_input_raw(value) {
    this._wrapperFields['digital_input_raw'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new NextupDigitalInputsRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    refObject.name.data.length = refObject.name.size;
    for (let index = 0; index < refObject.name.size; index++) {
      this._nameArray[index] = refObject.name.data[index].data;
    }
    refObject.di1.data.length = refObject.di1.size;
    for (let index = 0; index < refObject.di1.size; index++) {
      this._di1Array[index] = refObject.di1.data[index].data;
    }
    refObject.di2.data.length = refObject.di2.size;
    for (let index = 0; index < refObject.di2.size; index++) {
      this._di2Array[index] = refObject.di2.data[index].data;
    }
    refObject.di3.data.length = refObject.di3.size;
    for (let index = 0; index < refObject.di3.size; index++) {
      this._di3Array[index] = refObject.di3.data[index].data;
    }
    refObject.di4.data.length = refObject.di4.size;
    for (let index = 0; index < refObject.di4.size; index++) {
      this._di4Array[index] = refObject.di4.data[index].data;
    }
    refObject.di5.data.length = refObject.di5.size;
    for (let index = 0; index < refObject.di5.size; index++) {
      this._di5Array[index] = refObject.di5.data[index].data;
    }
    refObject.sto1.data.length = refObject.sto1.size;
    for (let index = 0; index < refObject.sto1.size; index++) {
      this._sto1Array[index] = refObject.sto1.data[index].data;
    }
    refObject.sto2.data.length = refObject.sto2.size;
    for (let index = 0; index < refObject.sto2.size; index++) {
      this._sto2Array[index] = refObject.sto2.data[index].data;
    }
    refObject.edm.data.length = refObject.edm.size;
    for (let index = 0; index < refObject.edm.size; index++) {
      this._edmArray[index] = refObject.edm.data[index].data;
    }
    this._wrapperFields.digital_input_raw.copyRefObject(this._refObject.digital_input_raw);
  }
  copy(other) {
    this._refObject = new NextupDigitalInputsRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._nameArray = other._nameArray.slice();
    this._di1Array = other._di1Array.slice();
    this._di2Array = other._di2Array.slice();
    this._di3Array = other._di3Array.slice();
    this._di4Array = other._di4Array.slice();
    this._di5Array = other._di5Array.slice();
    this._sto1Array = other._sto1Array.slice();
    this._sto2Array = other._sto2Array.slice();
    this._edmArray = other._edmArray.slice();
    this._wrapperFields.digital_input_raw.copy(other._wrapperFields.digital_input_raw);
  }
  static get classType() {
    return NextupDigitalInputsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"name","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"di1","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"di2","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"di3","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"di4","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"di5","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"sto1","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"sto2","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"edm","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"digital_input_raw","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"nextup_joint_interfaces","type":"NextupDigitalInputs","stringUpperBound":null,"isPrimitiveType":false},"msgName":"NextupDigitalInputs"};
  }
  hasMember(name) {
    let memberNames = ["header","name","di1","di2","di3","di4","di5","sto1","sto2","edm","digital_input_raw"];
    return memberNames.indexOf(name) !== -1;
  }
}
class NextupDigitalInputsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof NextupDigitalInputsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new NextupDigitalInputsWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NextupDigitalInputsArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NextupDigitalInputsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new NextupDigitalInputsRefArray(size);
    this._refObject = new NextupDigitalInputsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new NextupDigitalInputsWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof NextupDigitalInputsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "NextupDigitalInputsArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      NextupDigitalInputsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return NextupDigitalInputsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return NextupDigitalInputsArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = NextupDigitalInputsWrapper;
