// rclnodejs/generated/nextup_joint_interfaces/nextup_joint_interfaces__msg__NextupResetFault.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const NextupResetFaultRefStruct = StructType({
  resetfault: primitiveTypes.bool,
});
const NextupResetFaultRefArray = ArrayType(NextupResetFaultRefStruct);
const NextupResetFaultRefStructArray = StructType({
  data: NextupResetFaultRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class NextupResetFaultWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new NextupResetFaultRefStruct(other._refObject.toObject());
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new NextupResetFaultRefStruct();
  }
  static createFromRefObject(refObject) {
    let self = new NextupResetFaultWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new NextupResetFaultArrayWrapper();
  }
  static get ArrayType() {
    return NextupResetFaultArrayWrapper;
  }
  static get refObjectArrayType() {
    return NextupResetFaultRefStructArray;
  }
  static get refObjectType() {
    return NextupResetFaultRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.resetfault = refObject.resetfault;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destroyRawROS(msg) {
    NextupResetFaultWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nextup_joint_interfaces', subFolder: 'msg', interfaceName: 'NextupResetFault'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get resetfault() {
    return this._refObject.resetfault;
  }
  set resetfault(value) {
    this._refObject.resetfault = value;
  }
  copyRefObject(refObject) {
    this._refObject = new NextupResetFaultRefStruct(refObject.toObject());
  }
  copy(other) {
    this._refObject = new NextupResetFaultRefStruct(other._refObject.toObject());
  }
  static get classType() {
    return NextupResetFaultWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"resetfault","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"nextup_joint_interfaces","type":"NextupResetFault","stringUpperBound":null,"isPrimitiveType":false},"msgName":"NextupResetFault"};
  }
  hasMember(name) {
    let memberNames = ["resetfault"];
    return memberNames.indexOf(name) !== -1;
  }
}
class NextupResetFaultArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof NextupResetFaultWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new NextupResetFaultWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NextupResetFaultArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NextupResetFaultArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new NextupResetFaultRefArray(size);
    this._refObject = new NextupResetFaultRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new NextupResetFaultWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof NextupResetFaultArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "NextupResetFaultArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      NextupResetFaultWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return NextupResetFaultWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return NextupResetFaultArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = NextupResetFaultWrapper;
