// rclnodejs/generated/nextup_joint_interfaces/nextup_joint_interfaces__msg__NextupModeResetHandler.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const NextupModeResetHandlerRefStruct = StructType({
  reset: primitiveTypes.bool,
  mode: primitiveTypes.string,
});
const NextupModeResetHandlerRefArray = ArrayType(NextupModeResetHandlerRefStruct);
const NextupModeResetHandlerRefStructArray = StructType({
  data: NextupModeResetHandlerRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class NextupModeResetHandlerWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new NextupModeResetHandlerRefStruct(other._refObject.toObject());
      this._wrapperFields.mode =  new StringWrapper(other._wrapperFields.mode);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new NextupModeResetHandlerRefStruct();
    this._wrapperFields.mode = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new NextupModeResetHandlerWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new NextupModeResetHandlerArrayWrapper();
  }
  static get ArrayType() {
    return NextupModeResetHandlerArrayWrapper;
  }
  static get refObjectArrayType() {
    return NextupModeResetHandlerRefStructArray;
  }
  static get refObjectType() {
    return NextupModeResetHandlerRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.mode.freeze(own);
    }
    this._refObject.mode = this._wrapperFields.mode.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.reset = refObject.reset;
    this._wrapperFields.mode.data = refObject.mode.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.mode);
  }
  static destroyRawROS(msg) {
    NextupModeResetHandlerWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nextup_joint_interfaces', subFolder: 'msg', interfaceName: 'NextupModeResetHandler'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get reset() {
    return this._refObject.reset;
  }
  set reset(value) {
    this._refObject.reset = value;
  }
  get mode() {
    return this._wrapperFields.mode.data;
  }
  set mode(value) {
    this._wrapperFields.mode.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new NextupModeResetHandlerRefStruct(refObject.toObject());
    this._wrapperFields.mode.copyRefObject(this._refObject.mode);
  }
  copy(other) {
    this._refObject = new NextupModeResetHandlerRefStruct(other._refObject.toObject());
    this._wrapperFields.mode.copy(other._wrapperFields.mode);
  }
  static get classType() {
    return NextupModeResetHandlerWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"reset","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"nextup_joint_interfaces","type":"NextupModeResetHandler","stringUpperBound":null,"isPrimitiveType":false},"msgName":"NextupModeResetHandler"};
  }
  hasMember(name) {
    let memberNames = ["reset","mode"];
    return memberNames.indexOf(name) !== -1;
  }
}
class NextupModeResetHandlerArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof NextupModeResetHandlerWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new NextupModeResetHandlerWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NextupModeResetHandlerArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NextupModeResetHandlerArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new NextupModeResetHandlerRefArray(size);
    this._refObject = new NextupModeResetHandlerRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new NextupModeResetHandlerWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof NextupModeResetHandlerArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "NextupModeResetHandlerArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      NextupModeResetHandlerWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return NextupModeResetHandlerWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return NextupModeResetHandlerArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = NextupModeResetHandlerWrapper;
