// rclnodejs/generated/nextup_joint_interfaces/nextup_joint_interfaces__msg__NextupDigitalOutputs.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const BoolWrapper = require('../../generated/std_msgs/std_msgs__msg__Bool.js');
const NextupDigitalOutputsRefStruct = StructType({
  do1: BoolWrapper.refObjectArrayType,
  do2: BoolWrapper.refObjectArrayType,
  do3: BoolWrapper.refObjectArrayType,
  pi_p: BoolWrapper.refObjectArrayType,
});
const NextupDigitalOutputsRefArray = ArrayType(NextupDigitalOutputsRefStruct);
const NextupDigitalOutputsRefStructArray = StructType({
  data: NextupDigitalOutputsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class NextupDigitalOutputsWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new NextupDigitalOutputsRefStruct(other._refObject.toObject());
      this._wrapperFields.do1 = BoolWrapper.createArray();
      this._wrapperFields.do1.copy(other._wrapperFields.do1);
      this.do1 = other.do1;
      this._wrapperFields.do2 = BoolWrapper.createArray();
      this._wrapperFields.do2.copy(other._wrapperFields.do2);
      this.do2 = other.do2;
      this._wrapperFields.do3 = BoolWrapper.createArray();
      this._wrapperFields.do3.copy(other._wrapperFields.do3);
      this.do3 = other.do3;
      this._wrapperFields.pi_p = BoolWrapper.createArray();
      this._wrapperFields.pi_p.copy(other._wrapperFields.pi_p);
      this.pi_p = other.pi_p;
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new NextupDigitalOutputsRefStruct();
    this._wrapperFields.do1 = BoolWrapper.createArray();
    this._do1Array = [];
    this._wrapperFields.do2 = BoolWrapper.createArray();
    this._do2Array = [];
    this._wrapperFields.do3 = BoolWrapper.createArray();
    this._do3Array = [];
    this._wrapperFields.pi_p = BoolWrapper.createArray();
    this._pi_pArray = [];
  }
  static createFromRefObject(refObject) {
    let self = new NextupDigitalOutputsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new NextupDigitalOutputsArrayWrapper();
  }
  static get ArrayType() {
    return NextupDigitalOutputsArrayWrapper;
  }
  static get refObjectArrayType() {
    return NextupDigitalOutputsRefStructArray;
  }
  static get refObjectType() {
    return NextupDigitalOutputsRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.do1.fill([]);
    } else {
      this._wrapperFields.do1.fill(this._do1Array);
    }
    this._wrapperFields.do1.freeze(own);
    this._refObject.do1 = this._wrapperFields.do1.refObject;
    if (own) {
      this._wrapperFields.do2.fill([]);
    } else {
      this._wrapperFields.do2.fill(this._do2Array);
    }
    this._wrapperFields.do2.freeze(own);
    this._refObject.do2 = this._wrapperFields.do2.refObject;
    if (own) {
      this._wrapperFields.do3.fill([]);
    } else {
      this._wrapperFields.do3.fill(this._do3Array);
    }
    this._wrapperFields.do3.freeze(own);
    this._refObject.do3 = this._wrapperFields.do3.refObject;
    if (own) {
      this._wrapperFields.pi_p.fill([]);
    } else {
      this._wrapperFields.pi_p.fill(this._pi_pArray);
    }
    this._wrapperFields.pi_p.freeze(own);
    this._refObject.pi_p = this._wrapperFields.pi_p.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    refObject.do1.data.length = refObject.do1.size;
    for (let index = 0; index < refObject.do1.size; index++) {
      this._do1Array[index] = refObject.do1.data[index].data;
    }
    refObject.do2.data.length = refObject.do2.size;
    for (let index = 0; index < refObject.do2.size; index++) {
      this._do2Array[index] = refObject.do2.data[index].data;
    }
    refObject.do3.data.length = refObject.do3.size;
    for (let index = 0; index < refObject.do3.size; index++) {
      this._do3Array[index] = refObject.do3.data[index].data;
    }
    refObject.pi_p.data.length = refObject.pi_p.size;
    for (let index = 0; index < refObject.pi_p.size; index++) {
      this._pi_pArray[index] = refObject.pi_p.data[index].data;
    }
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.do1.size != 0) {
      BoolWrapper.ArrayType.freeArray(refObject.do1);
      if (!BoolWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.do1, BoolWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.do2.size != 0) {
      BoolWrapper.ArrayType.freeArray(refObject.do2);
      if (!BoolWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.do2, BoolWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.do3.size != 0) {
      BoolWrapper.ArrayType.freeArray(refObject.do3);
      if (!BoolWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.do3, BoolWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.pi_p.size != 0) {
      BoolWrapper.ArrayType.freeArray(refObject.pi_p);
      if (!BoolWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.pi_p, BoolWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    NextupDigitalOutputsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nextup_joint_interfaces', subFolder: 'msg', interfaceName: 'NextupDigitalOutputs'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get do1() {
    return this._do1Array;
  }
  set do1(value) {
    this._do1Array = value;
  }
  get do2() {
    return this._do2Array;
  }
  set do2(value) {
    this._do2Array = value;
  }
  get do3() {
    return this._do3Array;
  }
  set do3(value) {
    this._do3Array = value;
  }
  get pi_p() {
    return this._pi_pArray;
  }
  set pi_p(value) {
    this._pi_pArray = value;
  }
  copyRefObject(refObject) {
    this._refObject = new NextupDigitalOutputsRefStruct(refObject.toObject());
    refObject.do1.data.length = refObject.do1.size;
    for (let index = 0; index < refObject.do1.size; index++) {
      this._do1Array[index] = refObject.do1.data[index].data;
    }
    refObject.do2.data.length = refObject.do2.size;
    for (let index = 0; index < refObject.do2.size; index++) {
      this._do2Array[index] = refObject.do2.data[index].data;
    }
    refObject.do3.data.length = refObject.do3.size;
    for (let index = 0; index < refObject.do3.size; index++) {
      this._do3Array[index] = refObject.do3.data[index].data;
    }
    refObject.pi_p.data.length = refObject.pi_p.size;
    for (let index = 0; index < refObject.pi_p.size; index++) {
      this._pi_pArray[index] = refObject.pi_p.data[index].data;
    }
  }
  copy(other) {
    this._refObject = new NextupDigitalOutputsRefStruct(other._refObject.toObject());
    this._do1Array = other._do1Array.slice();
    this._do2Array = other._do2Array.slice();
    this._do3Array = other._do3Array.slice();
    this._pi_pArray = other._pi_pArray.slice();
  }
  static get classType() {
    return NextupDigitalOutputsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"do1","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"do2","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"do3","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pi_p","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"nextup_joint_interfaces","type":"NextupDigitalOutputs","stringUpperBound":null,"isPrimitiveType":false},"msgName":"NextupDigitalOutputs"};
  }
  hasMember(name) {
    let memberNames = ["do1","do2","do3","pi_p"];
    return memberNames.indexOf(name) !== -1;
  }
}
class NextupDigitalOutputsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof NextupDigitalOutputsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new NextupDigitalOutputsWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NextupDigitalOutputsArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NextupDigitalOutputsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new NextupDigitalOutputsRefArray(size);
    this._refObject = new NextupDigitalOutputsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new NextupDigitalOutputsWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof NextupDigitalOutputsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "NextupDigitalOutputsArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      NextupDigitalOutputsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return NextupDigitalOutputsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return NextupDigitalOutputsArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = NextupDigitalOutputsWrapper;
