// rclnodejs/generated/nextup_joint_interfaces/nextup_joint_interfaces__msg__NextupDriverStatus.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const BoolWrapper = require('../../generated/std_msgs/std_msgs__msg__Bool.js');
const UInt16Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt16.js');
const NextupDriverStatusRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  name: StringWrapper.refObjectArrayType,
  op_status: BoolWrapper.refObjectArrayType,
  fault: BoolWrapper.refObjectArrayType,
  state: StringWrapper.refObjectArrayType,
  statusword: UInt16Wrapper.refObjectArrayType,
});
const NextupDriverStatusRefArray = ArrayType(NextupDriverStatusRefStruct);
const NextupDriverStatusRefStructArray = StructType({
  data: NextupDriverStatusRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class NextupDriverStatusWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new NextupDriverStatusRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.name = StringWrapper.createArray();
      this._wrapperFields.name.copy(other._wrapperFields.name);
      this.name = other.name;
      this._wrapperFields.op_status = BoolWrapper.createArray();
      this._wrapperFields.op_status.copy(other._wrapperFields.op_status);
      this.op_status = other.op_status;
      this._wrapperFields.fault = BoolWrapper.createArray();
      this._wrapperFields.fault.copy(other._wrapperFields.fault);
      this.fault = other.fault;
      this._wrapperFields.state = StringWrapper.createArray();
      this._wrapperFields.state.copy(other._wrapperFields.state);
      this.state = other.state;
      this._wrapperFields.statusword = UInt16Wrapper.createArray();
      this._wrapperFields.statusword.copy(other._wrapperFields.statusword);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new NextupDriverStatusRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.name = StringWrapper.createArray();
    this._nameArray = [];
    this._wrapperFields.op_status = BoolWrapper.createArray();
    this._op_statusArray = [];
    this._wrapperFields.fault = BoolWrapper.createArray();
    this._faultArray = [];
    this._wrapperFields.state = StringWrapper.createArray();
    this._stateArray = [];
    this._wrapperFields.statusword = UInt16Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new NextupDriverStatusWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new NextupDriverStatusArrayWrapper();
  }
  static get ArrayType() {
    return NextupDriverStatusArrayWrapper;
  }
  static get refObjectArrayType() {
    return NextupDriverStatusRefStructArray;
  }
  static get refObjectType() {
    return NextupDriverStatusRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.header.freeze(own);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.name.fill([]);
    } else {
      this._wrapperFields.name.fill(this._nameArray);
    }
    this._wrapperFields.name.freeze(own);
    this._refObject.name = this._wrapperFields.name.refObject;
    if (own) {
      this._wrapperFields.op_status.fill([]);
    } else {
      this._wrapperFields.op_status.fill(this._op_statusArray);
    }
    this._wrapperFields.op_status.freeze(own);
    this._refObject.op_status = this._wrapperFields.op_status.refObject;
    if (own) {
      this._wrapperFields.fault.fill([]);
    } else {
      this._wrapperFields.fault.fill(this._faultArray);
    }
    this._wrapperFields.fault.freeze(own);
    this._refObject.fault = this._wrapperFields.fault.refObject;
    if (own) {
      this._wrapperFields.state.fill([]);
    } else {
      this._wrapperFields.state.fill(this._stateArray);
    }
    this._wrapperFields.state.freeze(own);
    this._refObject.state = this._wrapperFields.state.refObject;
    if (own) {
      this._wrapperFields.statusword.fill(Uint16Array.from([]));
    }
    this._wrapperFields.statusword.freeze(own);
    this._refObject.statusword = this._wrapperFields.statusword.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    refObject.name.data.length = refObject.name.size;
    for (let index = 0; index < refObject.name.size; index++) {
      this._nameArray[index] = refObject.name.data[index].data;
    }
    refObject.op_status.data.length = refObject.op_status.size;
    for (let index = 0; index < refObject.op_status.size; index++) {
      this._op_statusArray[index] = refObject.op_status.data[index].data;
    }
    refObject.fault.data.length = refObject.fault.size;
    for (let index = 0; index < refObject.fault.size; index++) {
      this._faultArray[index] = refObject.fault.data[index].data;
    }
    refObject.state.data.length = refObject.state.size;
    for (let index = 0; index < refObject.state.size; index++) {
      this._stateArray[index] = refObject.state.data[index].data;
    }
    this._wrapperFields.statusword.copyRefObject(refObject.statusword);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.name.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.name);
      if (!StringWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.name, StringWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.op_status.size != 0) {
      BoolWrapper.ArrayType.freeArray(refObject.op_status);
      if (!BoolWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.op_status, BoolWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.fault.size != 0) {
      BoolWrapper.ArrayType.freeArray(refObject.fault);
      if (!BoolWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.fault, BoolWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.state.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.state);
      if (!StringWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.state, StringWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.statusword.size != 0) {
      UInt16Wrapper.ArrayType.freeArray(refObject.statusword);
      if (!UInt16Wrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.statusword, UInt16Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    NextupDriverStatusWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nextup_joint_interfaces', subFolder: 'msg', interfaceName: 'NextupDriverStatus'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get name() {
    return this._nameArray;
  }
  set name(value) {
    this._nameArray = value;
  }
  get op_status() {
    return this._op_statusArray;
  }
  set op_status(value) {
    this._op_statusArray = value;
  }
  get fault() {
    return this._faultArray;
  }
  set fault(value) {
    this._faultArray = value;
  }
  get state() {
    return this._stateArray;
  }
  set state(value) {
    this._stateArray = value;
  }
  get statusword() {
    return this._wrapperFields['statusword'].data;
  }
  set statusword(value) {
    this._wrapperFields['statusword'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new NextupDriverStatusRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    refObject.name.data.length = refObject.name.size;
    for (let index = 0; index < refObject.name.size; index++) {
      this._nameArray[index] = refObject.name.data[index].data;
    }
    refObject.op_status.data.length = refObject.op_status.size;
    for (let index = 0; index < refObject.op_status.size; index++) {
      this._op_statusArray[index] = refObject.op_status.data[index].data;
    }
    refObject.fault.data.length = refObject.fault.size;
    for (let index = 0; index < refObject.fault.size; index++) {
      this._faultArray[index] = refObject.fault.data[index].data;
    }
    refObject.state.data.length = refObject.state.size;
    for (let index = 0; index < refObject.state.size; index++) {
      this._stateArray[index] = refObject.state.data[index].data;
    }
    this._wrapperFields.statusword.copyRefObject(this._refObject.statusword);
  }
  copy(other) {
    this._refObject = new NextupDriverStatusRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._nameArray = other._nameArray.slice();
    this._op_statusArray = other._op_statusArray.slice();
    this._faultArray = other._faultArray.slice();
    this._stateArray = other._stateArray.slice();
    this._wrapperFields.statusword.copy(other._wrapperFields.statusword);
  }
  static get classType() {
    return NextupDriverStatusWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"name","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"op_status","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fault","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"state","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"statusword","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"nextup_joint_interfaces","type":"NextupDriverStatus","stringUpperBound":null,"isPrimitiveType":false},"msgName":"NextupDriverStatus"};
  }
  hasMember(name) {
    let memberNames = ["header","name","op_status","fault","state","statusword"];
    return memberNames.indexOf(name) !== -1;
  }
}
class NextupDriverStatusArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof NextupDriverStatusWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new NextupDriverStatusWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NextupDriverStatusArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NextupDriverStatusArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new NextupDriverStatusRefArray(size);
    this._refObject = new NextupDriverStatusRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new NextupDriverStatusWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof NextupDriverStatusArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "NextupDriverStatusArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      NextupDriverStatusWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return NextupDriverStatusWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return NextupDriverStatusArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = NextupDriverStatusWrapper;
