// rclnodejs/generated/controller_manager_msgs/controller_manager_msgs__srv__ListHardwareComponents_Response.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HardwareComponentStateWrapper = require('../../generated/controller_manager_msgs/controller_manager_msgs__msg__HardwareComponentState.js');
const ListHardwareComponents_ResponseRefStruct = StructType({
  component: HardwareComponentStateWrapper.refObjectArrayType,
});
const ListHardwareComponents_ResponseRefArray = ArrayType(ListHardwareComponents_ResponseRefStruct);
const ListHardwareComponents_ResponseRefStructArray = StructType({
  data: ListHardwareComponents_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class ListHardwareComponents_ResponseWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new ListHardwareComponents_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.component = HardwareComponentStateWrapper.createArray();
      this._wrapperFields.component.copy(other._wrapperFields.component);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new ListHardwareComponents_ResponseRefStruct();
    this._wrapperFields.component = HardwareComponentStateWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new ListHardwareComponents_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ListHardwareComponents_ResponseArrayWrapper();
  }
  static get ArrayType() {
    return ListHardwareComponents_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return ListHardwareComponents_ResponseRefStructArray;
  }
  static get refObjectType() {
    return ListHardwareComponents_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    if (own) {
      this._wrapperFields.component.fill([]);
    }
    this._wrapperFields.component.freeze(own);
    this._refObject.component = this._wrapperFields.component.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.component.copyRefObject(refObject.component);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.component.size != 0) {
      HardwareComponentStateWrapper.ArrayType.freeArray(refObject.component);
      if (!HardwareComponentStateWrapper.ArrayType.useTypedArray) {
        deallocator.freeStructMember(refObject.component, HardwareComponentStateWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destroyRawROS(msg) {
    ListHardwareComponents_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'controller_manager_msgs', subFolder: 'srv', interfaceName: 'ListHardwareComponents_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get component() {
    return this._wrapperFields.component;
  }
  set component(value) {
    this._wrapperFields.component.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new ListHardwareComponents_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.component.copyRefObject(this._refObject.component);
  }
  copy(other) {
    this._refObject = new ListHardwareComponents_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.component.copy(other._wrapperFields.component);
  }
  static get classType() {
    return ListHardwareComponents_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"component","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"controller_manager_msgs","type":"HardwareComponentState","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"controller_manager_msgs","type":"ListHardwareComponents_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ListHardwareComponents_Response"};
  }
  hasMember(name) {
    let memberNames = ["component"];
    return memberNames.indexOf(name) !== -1;
  }
}
class ListHardwareComponents_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ListHardwareComponents_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ListHardwareComponents_ResponseWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ListHardwareComponents_ResponseArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ListHardwareComponents_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new ListHardwareComponents_ResponseRefArray(size);
    this._refObject = new ListHardwareComponents_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ListHardwareComponents_ResponseWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ListHardwareComponents_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ListHardwareComponents_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ListHardwareComponents_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ListHardwareComponents_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ListHardwareComponents_ResponseArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = ListHardwareComponents_ResponseWrapper;
