// rclnodejs/generated/object_recognition_msgs/object_recognition_msgs__action__ObjectRecognition_Result.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('../../third_party/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const RecognizedObjectArrayWrapper = require('../../generated/object_recognition_msgs/object_recognition_msgs__msg__RecognizedObjectArray.js');
const ObjectRecognition_ResultRefStruct = StructType({
  recognized_objects: RecognizedObjectArrayWrapper.refObjectType,
});
const ObjectRecognition_ResultRefArray = ArrayType(ObjectRecognition_ResultRefStruct);
const ObjectRecognition_ResultRefStructArray = StructType({
  data: ObjectRecognition_ResultRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
class ObjectRecognition_ResultWrapper {
  constructor(other) {
    this._initialize();
    this._setDefaults();
     if (typeof other === 'object' && other._refObject) {
      this._refObject = new ObjectRecognition_ResultRefStruct(other._refObject.toObject());
      this._wrapperFields.recognized_objects =  new RecognizedObjectArrayWrapper(other._wrapperFields.recognized_objects);
    } else if (typeof other !== 'undefined') {
      translator.constructFromPlanObject(this, other);
    }
    this.freeze(/*own=*/false);
  }
  _setDefaults() {
  }
  _initialize() {
    this._wrapperFields = {};
    this._refObject = new ObjectRecognition_ResultRefStruct();
    this._wrapperFields.recognized_objects = new RecognizedObjectArrayWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ObjectRecognition_ResultWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ObjectRecognition_ResultArrayWrapper();
  }
  static get ArrayType() {
    return ObjectRecognition_ResultArrayWrapper;
  }
  static get refObjectArrayType() {
    return ObjectRecognition_ResultRefStructArray;
  }
  static get refObjectType() {
    return ObjectRecognition_ResultRefStruct;
  }
  toRawROS() {
    this.freeze(/*own=*/true);
    return this._refObject.ref();
  }
  freeze(own) {
    this._wrapperFields.recognized_objects.freeze(own);
    this._refObject.recognized_objects = this._wrapperFields.recognized_objects.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.recognized_objects.copyRefObject(refObject.recognized_objects);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    RecognizedObjectArrayWrapper.freeStruct(refObject.recognized_objects);
  }
  static destroyRawROS(msg) {
    ObjectRecognition_ResultWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'object_recognition_msgs', subFolder: 'action', interfaceName: 'ObjectRecognition_Result'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get recognized_objects() {
    return this._wrapperFields.recognized_objects;
  }
  set recognized_objects(value) {
    if (value instanceof RecognizedObjectArrayWrapper) {
      this._wrapperFields.recognized_objects.copy(value);
    } else {
      this._wrapperFields.recognized_objects.copy(new RecognizedObjectArrayWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new ObjectRecognition_ResultRefStruct(refObject.toObject());
    this._wrapperFields.recognized_objects.copyRefObject(this._refObject.recognized_objects);
  }
  copy(other) {
    this._refObject = new ObjectRecognition_ResultRefStruct(other._refObject.toObject());
    this._wrapperFields.recognized_objects.copy(other._wrapperFields.recognized_objects);
  }
  static get classType() {
    return ObjectRecognition_ResultWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"recognized_objects","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"object_recognition_msgs","type":"RecognizedObjectArray","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"object_recognition_msgs","type":"ObjectRecognition_Result","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ObjectRecognition_Result"};
  }
  hasMember(name) {
    let memberNames = ["recognized_objects"];
    return memberNames.indexOf(name) !== -1;
  }
}
class ObjectRecognition_ResultArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ObjectRecognition_ResultWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ObjectRecognition_ResultWrapper(value);
      }
    });
  }
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
      this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ObjectRecognition_ResultArrayWrapper.size setter');
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ObjectRecognition_ResultArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
    }
    this._refArray = new ObjectRecognition_ResultRefArray(size);
    this._refObject = new ObjectRecognition_ResultRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ObjectRecognition_ResultWrapper());
    }
  }
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ObjectRecognition_ResultArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ObjectRecognition_ResultArrayWrapper".');
    }
    this._resize(other.size);
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ObjectRecognition_ResultWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ObjectRecognition_ResultWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ObjectRecognition_ResultArrayWrapper;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
}
module.exports = ObjectRecognition_ResultWrapper;
